{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 基于LLM bagging + 可解释\n",
    "> zhenkun.shi@tib.cas.cn   \n",
    "> 2025-01-14 \n",
    "\n",
    "## 1. 导入必要的包"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 192 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "import sys,os\n",
    "sys.path.insert(0, os.path.dirname(os.path.realpath('__file__')))\n",
    "sys.path.insert(1,'../../')\n",
    "from config import conf as cfg\n",
    "import pandas as pd\n",
    "import json\n",
    "import plotly.graph_objects as go\n",
    "from tools import btools\n",
    "from tqdm import tqdm\n",
    "import rxnrecer as production\n",
    "\n",
    "from modules.llm import Chat as llmchat\n",
    "import tools.bioFunctionLib as bfl\n",
    "from IPython.display import HTML\n",
    "from pandarallel import pandarallel # 导入pandaralle\n",
    "\n",
    "pandarallel.initialize(progress_bar=False)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. 加载数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>functionCounts</th>\n",
       "      <th>ec_specific_level</th>\n",
       "      <th>isenzyme</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A9JLI2</td>\n",
       "      <td>MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A9JLI3</td>\n",
       "      <td>MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A9JLI5</td>\n",
       "      <td>MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id                                                seq reaction_id  \\\n",
       "0     A9JLI2  MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...           -   \n",
       "1     A9JLI3  MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...           -   \n",
       "2     A9JLI5  MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...           -   \n",
       "\n",
       "  ec_number  functionCounts  ec_specific_level  isenzyme  \\\n",
       "0         -               0                  0     False   \n",
       "1         -               0                  0     False   \n",
       "2         -               0                  0     False   \n",
       "\n",
       "                                               label  \n",
       "0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_test = pd.read_feather(cfg.FILE_DS_TEST)\n",
    "ds_test.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. RXNRECer EXP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step 1: Preparing input data\n",
      "Step 2: Loading predictive model\n",
      "Use device: cuda\n",
      "Step 3: Running prediction on 100 proteins\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Predicting reactions: 100%|██████████| 50/50 [00:03<00:00, 13.79it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transferred model to GPU\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [00:03<00:00, 27.43it/s]\n",
      "/hpcfs/fhome/shizhenkun/miniconda3/envs/rxnrecer/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884\n",
      "  warnings.warn(\n",
      "Processing Sequences: 100%|██████████| 100/100 [00:01<00:00, 68.00it/s]\n",
      "/hpcfs/fhome/shizhenkun/miniconda3/envs/rxnrecer/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884\n",
      "  warnings.warn(\n",
      "Processing Sequences: 100%|██████████| 100/100 [00:01<00:00, 67.94it/s]\n",
      "100%|██████████| 1/1 [02:00<00:00, 120.85s/it]\n"
     ]
    }
   ],
   "source": [
    "expdata = ds_test.head(100)\n",
    "res_ecrecer = production.step_by_step_prediction(input_data=expdata[['uniprot_id', 'seq']], Ensemble=True, batch_size=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A9JLI2</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A9JLI3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A9JLI5</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A9JLI7</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B5KVH4</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Q572D3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>D0NPN8</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>E2DWQ7</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>P0DOO9</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>P0DOP0</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    input_id RXNRECer RXNRECer_with_prob\n",
       "0     A9JLI2        -    {'-': 0.999997}\n",
       "1     A9JLI3        -    {'-': 0.999998}\n",
       "2     A9JLI5        -         {'-': 1.0}\n",
       "3     A9JLI7        -    {'-': 0.999997}\n",
       "4     B5KVH4        -    {'-': 0.999998}\n",
       "..       ...      ...                ...\n",
       "96    Q572D3        -    {'-': 0.999996}\n",
       "97    D0NPN8        -    {'-': 0.999996}\n",
       "98    E2DWQ7        -    {'-': 0.999996}\n",
       "99    P0DOO9        -         {'-': 1.0}\n",
       "100   P0DOP0        -         {'-': 1.0}\n",
       "\n",
       "[101 rows x 3 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A9JLI2</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A9JLI3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A9JLI5</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "      <td>MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A9JLI7</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MLVIILGVIGLLANQVLGLPTQAGGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B5KVH4</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MAKPILLSIYLCLIIVALFNGCLAQSGGRQQHKFGQCQLNRLDALE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Q572D3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>MRLAIMLSATAVAINFATSSAIDQTKVLVYGTPAHYIHDSAGRRLL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>D0NPN8</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>MRLAIMLSATAVAINFATSSAIDQTKVLVYGTPAHYIHDSAGRRLL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>E2DWQ7</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>MRLAIMLSATAVAINFATCSAIDQTKVLVYGTPAHYIHDSAGRRLL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>P0DOO9</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "      <td>MGNKNIKPSKENRLSILYKDRMDSFKRGSWATSSFREKSHATIQRF...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>P0DOP0</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "      <td>MGNKNIKPSKENRLSILYKDRMDSFKRGSWATSSFREKSHATIQRF...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    uniprot_id RXNRECer RXNRECer_with_prob  \\\n",
       "0       A9JLI2        -    {'-': 0.999997}   \n",
       "1       A9JLI3        -    {'-': 0.999998}   \n",
       "2       A9JLI5        -         {'-': 1.0}   \n",
       "3       A9JLI7        -    {'-': 0.999997}   \n",
       "4       B5KVH4        -    {'-': 0.999998}   \n",
       "..         ...      ...                ...   \n",
       "96      Q572D3        -    {'-': 0.999996}   \n",
       "97      D0NPN8        -    {'-': 0.999996}   \n",
       "98      E2DWQ7        -    {'-': 0.999996}   \n",
       "99      P0DOO9        -         {'-': 1.0}   \n",
       "100     P0DOP0        -         {'-': 1.0}   \n",
       "\n",
       "                                                   seq rxn_groundtruth  \n",
       "0    MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...               -  \n",
       "1    MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...               -  \n",
       "2    MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...               -  \n",
       "3    MLVIILGVIGLLANQVLGLPTQAGGHLRSTDNPPQEELGYWCTYME...               -  \n",
       "4    MAKPILLSIYLCLIIVALFNGCLAQSGGRQQHKFGQCQLNRLDALE...               -  \n",
       "..                                                 ...             ...  \n",
       "96   MRLAIMLSATAVAINFATSSAIDQTKVLVYGTPAHYIHDSAGRRLL...               -  \n",
       "97   MRLAIMLSATAVAINFATSSAIDQTKVLVYGTPAHYIHDSAGRRLL...               -  \n",
       "98   MRLAIMLSATAVAINFATCSAIDQTKVLVYGTPAHYIHDSAGRRLL...               -  \n",
       "99   MGNKNIKPSKENRLSILYKDRMDSFKRGSWATSSFREKSHATIQRF...               -  \n",
       "100  MGNKNIKPSKENRLSILYKDRMDSFKRGSWATSSFREKSHATIQRF...               -  \n",
       "\n",
       "[101 rows x 5 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer = res_ecrecer.rename(columns={'input_id':'uniprot_id'})\n",
    "res_ecrecer = res_ecrecer.merge(expdata[['uniprot_id', 'seq', 'reaction_id']], on='uniprot_id', how='left').rename(columns={'reaction_id':'rxn_groundtruth'})\n",
    "res_ecrecer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Load RXN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>equation</th>\n",
       "      <th>chebi_id</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>equation_chebi</th>\n",
       "      <th>equation_smiles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHEA:22636</td>\n",
       "      <td>dCTP + H2O = dCMP + diphosphate + H(+)</td>\n",
       "      <td>CHEBI:61481;CHEBI:15377;CHEBI:57566;CHEBI:3301...</td>\n",
       "      <td>EC:3.6.1.9;EC:3.6.1.12;EC:3.6.1.65</td>\n",
       "      <td>CHEBI:61481 + CHEBI:15377 = CHEBI:57566 + CHEB...</td>\n",
       "      <td>Nc1ccn([C@H]2C[C@H](O)[C@@H](COP([O-])(=O)OP([...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RHEA:22640</td>\n",
       "      <td>NADP(+) + sphinganine = 3-oxosphinganine + H(+...</td>\n",
       "      <td>CHEBI:58349;CHEBI:57817;CHEBI:58299;CHEBI:1537...</td>\n",
       "      <td>EC:1.1.1.102</td>\n",
       "      <td>CHEBI:58349 + CHEBI:57817 = CHEBI:58299 + CHEB...</td>\n",
       "      <td>NC(=O)c1ccc[n+](c1)[C@@H]1O[C@H](COP([O-])(=O)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RHEA:22644</td>\n",
       "      <td>O2 + protopine + reduced [NADPH--hemoprotein r...</td>\n",
       "      <td>CHEBI:15379;CHEBI:16415;CHEBI:57618;CHEBI:1710...</td>\n",
       "      <td>EC:1.14.14.98</td>\n",
       "      <td>CHEBI:15379 + CHEBI:16415 + CHEBI:57618 = CHEB...</td>\n",
       "      <td>O=O.CN1CCc2cc3OCOc3cc2C(=O)Cc2ccc3OCOc3c2C1.Cc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id                                           equation  \\\n",
       "0  RHEA:22636             dCTP + H2O = dCMP + diphosphate + H(+)   \n",
       "1  RHEA:22640  NADP(+) + sphinganine = 3-oxosphinganine + H(+...   \n",
       "2  RHEA:22644  O2 + protopine + reduced [NADPH--hemoprotein r...   \n",
       "\n",
       "                                            chebi_id  \\\n",
       "0  CHEBI:61481;CHEBI:15377;CHEBI:57566;CHEBI:3301...   \n",
       "1  CHEBI:58349;CHEBI:57817;CHEBI:58299;CHEBI:1537...   \n",
       "2  CHEBI:15379;CHEBI:16415;CHEBI:57618;CHEBI:1710...   \n",
       "\n",
       "                            ec_number  \\\n",
       "0  EC:3.6.1.9;EC:3.6.1.12;EC:3.6.1.65   \n",
       "1                        EC:1.1.1.102   \n",
       "2                       EC:1.14.14.98   \n",
       "\n",
       "                                      equation_chebi  \\\n",
       "0  CHEBI:61481 + CHEBI:15377 = CHEBI:57566 + CHEB...   \n",
       "1  CHEBI:58349 + CHEBI:57817 = CHEBI:58299 + CHEB...   \n",
       "2  CHEBI:15379 + CHEBI:16415 + CHEBI:57618 = CHEB...   \n",
       "\n",
       "                                     equation_smiles  \n",
       "0  Nc1ccn([C@H]2C[C@H](O)[C@@H](COP([O-])(=O)OP([...  \n",
       "1  NC(=O)c1ccc[n+](c1)[C@@H]1O[C@H](COP([O-])(=O)...  \n",
       "2  O=O.CN1CCc2cc3OCOc3cc2C(=O)Cc2ccc3OCOc3c2C1.Cc...  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxn = pd.read_feather(cfg.FILE_RHEA_REACTION)\n",
    "rxn.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P0DTH6</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MKKIFLTMTPLPYMNHLPENHYFLSLTVVISIIHLFTTCVPQHNHS...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Q7XQ85</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.997593, '-': 0.985379}</td>\n",
       "      <td>MAYQGIDLLSTKAAGDDHGENSSYFDGWKAYDTNPFDLRHNRGGVI...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A0A0P0WIY3</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.990111, '-': 0.7777}</td>\n",
       "      <td>MVGRMLSSPEPTLSTMAMSAAHGEDSPYFAGWRAYDEDPYDPITNP...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Q5W6F9</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.989836, '-': 0.7777}</td>\n",
       "      <td>MGVKLLADGCAGASSSPALSRVATSAAHGEGSPYFAGWKAYDEDPY...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A0P0UZP7</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.979187, '-': 0.7777}</td>\n",
       "      <td>MIMSGFHIGIYTSICLYIPLPHLEPWIISSHTPKNLNLLDCLLYCS...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Q9SNN8</td>\n",
       "      <td>RHEA:22104;-;RHEA:21744</td>\n",
       "      <td>{'RHEA:22104': 0.980896, '-': 0.980291, 'RHEA:...</td>\n",
       "      <td>MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRG...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>L0TBY6</td>\n",
       "      <td>-;RHEA:20049</td>\n",
       "      <td>{'-': 0.99999, 'RHEA:20049': 0.989305}</td>\n",
       "      <td>MTHREELLPPMKWDAWGDPAAAKPLSDGVRSLLKQVVGLADSEQPE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M5AWY0</td>\n",
       "      <td>RHEA:34415;-;RHEA:56596</td>\n",
       "      <td>{'RHEA:34415': 0.999452, '-': 0.7777, 'RHEA:56...</td>\n",
       "      <td>MQAGFFHTPYNLPTRTARQMFDWSLKLAQVCDEAGFADFMIGEHST...</td>\n",
       "      <td>RHEA:34415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Q6I621</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MWKGFLSKLPRKTSASGRGADLDSGQCSNGAGNGNPIQRTSSCGSI...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Q88FI7</td>\n",
       "      <td>RHEA:18049;RHEA:23352;RHEA:10212;RHEA:12268</td>\n",
       "      <td>{'RHEA:18049': 0.99144, 'RHEA:23352': 0.988573...</td>\n",
       "      <td>MNQESISQSIAIVHPITLSHGRNAEVWDTDGKRYIDFVGGIGVLNL...</td>\n",
       "      <td>RHEA:12601;RHEA:10212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A5CAL1</td>\n",
       "      <td>RHEA:10992;RHEA:17905;RHEA:18657;RHEA:57408;RH...</td>\n",
       "      <td>{'RHEA:10992': 0.998241, 'RHEA:17905': 0.99824...</td>\n",
       "      <td>MESIGVLLTCPMNPYLEQELDKRFKLFRFWDFPSANDLFREHSNSI...</td>\n",
       "      <td>RHEA:16369;RHEA:18657;RHEA:18229;RHEA:10992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Q2PS07</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MKLFIILATATLLIAATQATYPRDEGFDLGETQMSSKCMRQVKMNE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>E9NX73</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>MKLFIILATATLLIAATQAKYLRDEGFDLGETQMSSKCTRQVKMME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P93198</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999994}</td>\n",
       "      <td>AALLVALLFVANAAAFRTTITTMEIDEDIDNPRRRGEGCREQIQRQ...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    uniprot_id                                           RXNRECer  \\\n",
       "11      P0DTH6                                                  -   \n",
       "12      Q7XQ85                                       RHEA:21744;-   \n",
       "13  A0A0P0WIY3                                       RHEA:21744;-   \n",
       "14      Q5W6F9                                       RHEA:21744;-   \n",
       "15  A0A0P0UZP7                                       RHEA:21744;-   \n",
       "16      Q9SNN8                            RHEA:22104;-;RHEA:21744   \n",
       "17      L0TBY6                                       -;RHEA:20049   \n",
       "18      M5AWY0                            RHEA:34415;-;RHEA:56596   \n",
       "19      Q6I621                                                  -   \n",
       "20      Q88FI7        RHEA:18049;RHEA:23352;RHEA:10212;RHEA:12268   \n",
       "21      A5CAL1  RHEA:10992;RHEA:17905;RHEA:18657;RHEA:57408;RH...   \n",
       "22      Q2PS07                                                  -   \n",
       "23      E9NX73                                                  -   \n",
       "24      P93198                                                  -   \n",
       "\n",
       "                                   RXNRECer_with_prob  \\\n",
       "11                                    {'-': 0.999998}   \n",
       "12            {'RHEA:21744': 0.997593, '-': 0.985379}   \n",
       "13              {'RHEA:21744': 0.990111, '-': 0.7777}   \n",
       "14              {'RHEA:21744': 0.989836, '-': 0.7777}   \n",
       "15              {'RHEA:21744': 0.979187, '-': 0.7777}   \n",
       "16  {'RHEA:22104': 0.980896, '-': 0.980291, 'RHEA:...   \n",
       "17             {'-': 0.99999, 'RHEA:20049': 0.989305}   \n",
       "18  {'RHEA:34415': 0.999452, '-': 0.7777, 'RHEA:56...   \n",
       "19                                    {'-': 0.999998}   \n",
       "20  {'RHEA:18049': 0.99144, 'RHEA:23352': 0.988573...   \n",
       "21  {'RHEA:10992': 0.998241, 'RHEA:17905': 0.99824...   \n",
       "22                                    {'-': 0.999997}   \n",
       "23                                    {'-': 0.999996}   \n",
       "24                                    {'-': 0.999994}   \n",
       "\n",
       "                                                  seq  \\\n",
       "11  MKKIFLTMTPLPYMNHLPENHYFLSLTVVISIIHLFTTCVPQHNHS...   \n",
       "12  MAYQGIDLLSTKAAGDDHGENSSYFDGWKAYDTNPFDLRHNRGGVI...   \n",
       "13  MVGRMLSSPEPTLSTMAMSAAHGEDSPYFAGWRAYDEDPYDPITNP...   \n",
       "14  MGVKLLADGCAGASSSPALSRVATSAAHGEGSPYFAGWKAYDEDPY...   \n",
       "15  MIMSGFHIGIYTSICLYIPLPHLEPWIISSHTPKNLNLLDCLLYCS...   \n",
       "16  MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRG...   \n",
       "17  MTHREELLPPMKWDAWGDPAAAKPLSDGVRSLLKQVVGLADSEQPE...   \n",
       "18  MQAGFFHTPYNLPTRTARQMFDWSLKLAQVCDEAGFADFMIGEHST...   \n",
       "19  MWKGFLSKLPRKTSASGRGADLDSGQCSNGAGNGNPIQRTSSCGSI...   \n",
       "20  MNQESISQSIAIVHPITLSHGRNAEVWDTDGKRYIDFVGGIGVLNL...   \n",
       "21  MESIGVLLTCPMNPYLEQELDKRFKLFRFWDFPSANDLFREHSNSI...   \n",
       "22  MKLFIILATATLLIAATQATYPRDEGFDLGETQMSSKCMRQVKMNE...   \n",
       "23  MKLFIILATATLLIAATQAKYLRDEGFDLGETQMSSKCTRQVKMME...   \n",
       "24  AALLVALLFVANAAAFRTTITTMEIDEDIDNPRRRGEGCREQIQRQ...   \n",
       "\n",
       "                                rxn_groundtruth  \n",
       "11                                            -  \n",
       "12                                   RHEA:21744  \n",
       "13                                   RHEA:21744  \n",
       "14                                   RHEA:21744  \n",
       "15                                   RHEA:21744  \n",
       "16                                   RHEA:21744  \n",
       "17                                            -  \n",
       "18                                   RHEA:34415  \n",
       "19                                            -  \n",
       "20                        RHEA:12601;RHEA:10212  \n",
       "21  RHEA:16369;RHEA:18657;RHEA:18229;RHEA:10992  \n",
       "22                                            -  \n",
       "23                                            -  \n",
       "24                                            -  "
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer.iloc[11:25]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def get_rxn_detail(rxn_id, rxn_bank):\n",
    "    if rxn_id == '-':\n",
    "        return {\n",
    "            'reaction id': '-',\n",
    "            'reaction equation': '-'\n",
    "        }\n",
    "    \n",
    "    rxn_record = rxn_bank[rxn_bank.reaction_id == rxn_id]\n",
    "    \n",
    "    if rxn_record.empty:\n",
    "        return {}  # Avoid errors if the record is missing\n",
    "\n",
    "    rxn_record = rxn_record.iloc[0]  # Take the first match\n",
    "\n",
    "    return {\n",
    "        'reaction id': rxn_record.reaction_id,\n",
    "        'reaction equation': rxn_record.equation,\n",
    "        'reaction equation in ChEBI format': rxn_record.equation_chebi,\n",
    "        'reaction equation in SMILES format': rxn_record.equation_smiles,\n",
    "        'reaction associated Enzyme Commission Number': rxn_record.ec_number\n",
    "    }\n",
    "\n",
    "def make_make_query(uniprot_id, seq, rxn_id, rxn_bank, prompt):\n",
    "    if not rxn_id or cfg.SPLITER not in rxn_id:\n",
    "        return {}\n",
    "\n",
    "    rxns = rxn_id.split(cfg.SPLITER)\n",
    "\n",
    "\n",
    "\n",
    "    query_info = {\n",
    "        'protein information': {\n",
    "            'uniprot id': uniprot_id,\n",
    "            'protein amino acid sequence': seq\n",
    "        },\n",
    "        'reaction information': {\n",
    "            f'candidate reaction {i+1}': get_rxn_detail(rxn, rxn_bank)\n",
    "            for i, rxn in enumerate(rxns)\n",
    "        }\n",
    "    }\n",
    "\n",
    "    llm_query = prompt  + json.dumps(query_info, indent=4, ensure_ascii=False)\n",
    "\n",
    "    return llm_query\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt0 = \"\"\"You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data.\n",
    "\n",
    "### Guidelines:\n",
    "1. **Input Data:**\n",
    "- **Protein Information:** Includes the Uniprot ID and the protein's amino acid sequence.\n",
    "- **Reaction Information:** A list of candidate reactions. Each reaction is provided in:\n",
    "    - Standard reaction equation\n",
    "    - Reaction equation in ChEBI format\n",
    "    - Reaction equation in SMILES format\n",
    "\n",
    "2. **Sequence Analysis & Functional Inference:**\n",
    "- **Basic Properties:** Consider the protein's length, molecular weight, and amino acid composition to see if they match known enzyme classes.\n",
    "- **Domain & Active Site Identification:** Look for conserved domains and key catalytic residues (or cofactor-binding motifs) that are critical for catalysis.\n",
    "- **Structural Considerations:** Evaluate whether the predicted secondary/tertiary structure supports the formation of an active site suitable for the reaction’s substrate.\n",
    "- **Database & Homology Evidence:** Use similarities to annotated proteins and known enzyme families to infer likely catalytic activity.\n",
    "\n",
    "3. **Handling Non-Catalytic Cases:**\n",
    "- A candidate reaction denoted by a single dash (\"-\") indicates that the protein is predicted not to catalyze that particular reaction.\n",
    "- **Important:** If all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in the output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "\n",
    "4. **Multiple Reactions:**\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "\n",
    "5. **Ranking:**\n",
    "- Rank the selected reactions by likelihood (1 being the most likely).\n",
    "\n",
    "6. **Justification:**\n",
    "- Provide a brief explanation for each selection, citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "### Expected Output (in JSON format):\n",
    "```json\n",
    "[\n",
    "    {\n",
    "        \"reaction_id\": \"xxx\",\n",
    "        \"selected\": \"yes\" or \"no\",\n",
    "        \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "        \"confidence\": <number>,      // Value between 0 and 1\n",
    "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "    },\n",
    "    ...\n",
    "]\n",
    "\"\"\"\n",
    "\n",
    "prompt1 = \"\"\"\n",
    "You are a biochemical expert. Given a protein sequence and a set of potential catalyzed reactions, please analyze and determine the most likely catalyzed reactions. \n",
    "    The information is provided in JSON format.\n",
    "    Important note: If a reaction is represented by a single dash (\"-\"), it means the protein cannot catalyze a reaction. \n",
    "    Please consider this in your analysis.\n",
    "    \n",
    "Guidelines:\n",
    "    1. **Handling Non-Catalytic Cases:**\n",
    "    - A candidate reaction denoted by a single dash (\"-\") indicates that the protein is predicted not to catalyze that particular reaction.\n",
    "    - **Important:** If all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in the output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "\n",
    "    2. **Multiple Reactions:**\n",
    "    - Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "\n",
    "    3. **Ranking:**\n",
    "    - Rank the selected reactions by likelihood (1 being the most likely).\n",
    "\n",
    "    4. **Justification:**\n",
    "    - Provide a brief explanation for each selection, citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "    \n",
    "Expected Output (in JSON format):\n",
    "    ```json\n",
    "    [\n",
    "        {\n",
    "            \"reaction_id\": \"xxx\",\n",
    "            \"selected\": \"yes\" or \"no\",\n",
    "            \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "            \"confidence\": <number>,      // Value between 0 and 1\n",
    "            \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "        },\n",
    "        ...\n",
    "    ]\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "prompt2 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Important Notes:\n",
    "1. If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "2. Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "3. For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,      // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "prompt3 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Structural Chemistry Analysis**\n",
    "   - Parse SMILES to identify reaction centers (e.g., bond cleavage sites)\n",
    "   $$ \\text{Example: } \\text{C-S bond in } \\mathrm{CHEBI:59789} \\rightarrow \\mathrm{CH_3S^- + adenosyl} $$\n",
    "   - Map ChEBI IDs to functional groups (e.g., thioether in CHEBI:59789)\n",
    "\n",
    "2. **Sequence-Structure Mapping**\n",
    "   - Detect catalytic triads using PROSITE patterns:\n",
    "     $$ \\text{Cys-X-X-Cys} \\Rightarrow \\text{Sulfur metabolism} $$\n",
    "     $$ \\text{Ser-Lys-Asp} \\Rightarrow \\text{Hydrolase activity} $$\n",
    "   - Verify cofactor binding with PROSITE PDOC00005\n",
    "\n",
    "3. **Dynamic Compatibility Check**\n",
    "   - Calculate active site volume using:\n",
    "     $$ V_{\\text{site}} = \\frac{4}{3}\\pi r_{\\text{substrate}}^3 \\pm 15\\% $$\n",
    "   - Validate stereochemical constraints from SMILES chirality tags (@, @@)\n",
    "\n",
    "4. **Conflict Resolution Protocol**\n",
    "   - Prioritize reaction chemistry over EC numbers when conflicts exist\n",
    "   - Use UniProtKB/Swiss-Prot annotations (e.g., FT ACT_SITE) as tiebreaker\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,      // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "prompt3 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Structural Chemistry Analysis**\n",
    "   - Parse SMILES to identify reaction centers (e.g., bond cleavage sites)\n",
    "   $$ \\text{Example: } \\text{C-S bond in } \\mathrm{CHEBI:59789} \\rightarrow \\mathrm{CH_3S^- + adenosyl} $$\n",
    "   - Map ChEBI IDs to functional groups (e.g., thioether in CHEBI:59789)\n",
    "\n",
    "2. **Sequence-Structure Mapping**\n",
    "   - Detect catalytic triads using PROSITE patterns:\n",
    "     $$ \\text{Cys-X-X-Cys} \\Rightarrow \\text{Sulfur metabolism} $$\n",
    "     $$ \\text{Ser-Lys-Asp} \\Rightarrow \\text{Hydrolase activity} $$\n",
    "   - Verify cofactor binding with PROSITE PDOC00005\n",
    "\n",
    "3. **Dynamic Compatibility Check**\n",
    "   - Calculate active site volume using:\n",
    "     $$ V_{\\text{site}} = \\frac{4}{3}\\pi r_{\\text{substrate}}^3 \\pm 15\\% $$\n",
    "   - Validate stereochemical constraints from SMILES chirality tags (@, @@)\n",
    "\n",
    "4. **Conflict Resolution Protocol**\n",
    "   - Prioritize reaction chemistry over EC numbers when conflicts exist\n",
    "   - Use UniProtKB/Swiss-Prot annotations (e.g., FT ACT_SITE) as tiebreaker\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,      // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "prompt4 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Omni-enzyme Structural Profiling**\n",
    "   - Perform topological analysis of SMILES using:\n",
    "     $$ \\text{Graph neural networks for reaction center detection} $$\n",
    "     $$ \\text{Electrostatic potential mapping (EPM) for charge complementarity} $$\n",
    "   - Cross-reference ChEBI IDs with MetaCyc reaction pathways:\n",
    "     $$ \\forall R \\in \\text{Reactions}, \\exists \\text{Substrate-Enzyme PCMatrix} \\geq 0.7 $$\n",
    "\n",
    "2. **Quantum Biochemical Evaluation**\n",
    "   - Calculate catalytic proficiency using:\n",
    "     $$ k_{\\text{cat}}/K_m = \\frac{k_B T}{h} e^{-\\Delta G^\\ddagger/RT} $$\n",
    "   - Simulate transition states with:\n",
    "     $$ \\text{QM/MM modeling of active site frontier orbitals} $$\n",
    "     $$ \\text{Reaction coordinate analysis (IRC)} $$\n",
    "\n",
    "3. **Evolutionary Dynamics Integration**\n",
    "   - Calculate phylogenetic catalytic propensity:\n",
    "     $$ P_{\\text{cat}} = \\frac{\\sum \\text{Consurf scores}}{\\text{Active site residues}} \\times \\text{EC_specific_phyletic_pattern} $$\n",
    "   - Detect convergent evolution signatures through:\n",
    "     $$ \\text{Parallel substitution analysis in catalytic residues} $$\n",
    "     $$ \\text{Structural phylogeny using DALI Z-scores} $$\n",
    "\n",
    "4. **Multi-scale Mechanistic Validation**\n",
    "   - Verify allosteric regulation potential:\n",
    "     $$ \\text{Ensemble docking across conformational states} $$\n",
    "     $$ \\text{Principal component analysis of MD trajectories} $$\n",
    "   - Cross-validate with omics data:\n",
    "     $$ \\text{Co-expression with metabolic pathway genes (r ≥ 0.85)} $$\n",
    "     $$ \\text{Proteomic colocalization evidence} $$\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,    // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "prompt5 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Literature Evidence Triangulation** ★最高优先级\n",
    "   a. UniProt/Sequence-based search:\n",
    "     $$ \\text{IF } \\exists \\text{ PMIDs } (P|E) \\rightarrow \\text{ReactID} \\text{ THEN confidence} += 0.4 $$\n",
    "   b. Reaction-centric search:\n",
    "     $$ \\text{Match(Equation|ChEBI|SMILES) → PMIDs → Enzyme } \\cap \\text{Input} $$\n",
    "   c. EC number validation:\n",
    "     $$ \\text{Validate EC}_{obs} = \\frac{\\text{PMID}_{\\text{EC-match}}}{\\text{PMID}_{\\text{EC-total}}} \\geq 0.6 $$\n",
    "\n",
    "2. **Structural Catalytic Competence**\n",
    "   - SMILES拓扑指纹匹配:\n",
    "     $$ \\text{Tanimoto} \\geq 0.85 \\text{ with Catalophore model} $$\n",
    "   - 电荷互补性验证:\n",
    "     $$ \\Delta E_{\\text{elec}} = \\sum \\frac{q_i q_j}{4\\pi\\epsilon r_{ij}} \\leq -15 \\text{ kJ/mol} $$\n",
    "\n",
    "3. **Quantum Transition State Analysis**\n",
    "   - 催化效率计算:\n",
    "     $$ \\ln(k_{\\text{cat}}/K_m) \\propto -\\Delta G^\\ddagger_{\\text{calc}} + \\text{文献修正项} $$\n",
    "   - 轨道相互作用验证:\n",
    "     $$ \\text{HOMO-LUMO gap匹配度} \\geq 82\\% $$\n",
    "\n",
    "4. **Evolutionary & Omics Validation**\n",
    "   - 系统发育可信度:\n",
    "     $$ P_{\\text{cat}} = \\text{Consurf} \\times \\text{EC}_{phylo} \\geq 7.5 $$\n",
    "   - 组学相关性:\n",
    "     $$ r_{\\text{coex}} \\geq 0.9 \\text{ AND } \\text{LocalScore} \\geq 0.8 $$\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,    // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C0HLM8</td>\n",
       "      <td>-;RHEA:15801</td>\n",
       "      <td>{'-': 0.999989, 'RHEA:15801': 0.940825}</td>\n",
       "      <td>GLEETYCSMKIK</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Q7XQ85</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.997593, '-': 0.985379}</td>\n",
       "      <td>MAYQGIDLLSTKAAGDDHGENSSYFDGWKAYDTNPFDLRHNRGGVI...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A0A0P0WIY3</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.990111, '-': 0.7777}</td>\n",
       "      <td>MVGRMLSSPEPTLSTMAMSAAHGEDSPYFAGWRAYDEDPYDPITNP...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Q5W6F9</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.989836, '-': 0.7777}</td>\n",
       "      <td>MGVKLLADGCAGASSSPALSRVATSAAHGEGSPYFAGWKAYDEDPY...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A0P0UZP7</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'RHEA:21744': 0.979187, '-': 0.7777}</td>\n",
       "      <td>MIMSGFHIGIYTSICLYIPLPHLEPWIISSHTPKNLNLLDCLLYCS...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Q9SNN8</td>\n",
       "      <td>RHEA:22104;-;RHEA:21744</td>\n",
       "      <td>{'RHEA:22104': 0.980896, '-': 0.980291, 'RHEA:...</td>\n",
       "      <td>MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRG...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>L0TBY6</td>\n",
       "      <td>-;RHEA:20049</td>\n",
       "      <td>{'-': 0.99999, 'RHEA:20049': 0.989305}</td>\n",
       "      <td>MTHREELLPPMKWDAWGDPAAAKPLSDGVRSLLKQVVGLADSEQPE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M5AWY0</td>\n",
       "      <td>RHEA:34415;-;RHEA:56596</td>\n",
       "      <td>{'RHEA:34415': 0.999452, '-': 0.7777, 'RHEA:56...</td>\n",
       "      <td>MQAGFFHTPYNLPTRTARQMFDWSLKLAQVCDEAGFADFMIGEHST...</td>\n",
       "      <td>RHEA:34415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Q88FI7</td>\n",
       "      <td>RHEA:18049;RHEA:23352;RHEA:10212;RHEA:12268</td>\n",
       "      <td>{'RHEA:18049': 0.99144, 'RHEA:23352': 0.988573...</td>\n",
       "      <td>MNQESISQSIAIVHPITLSHGRNAEVWDTDGKRYIDFVGGIGVLNL...</td>\n",
       "      <td>RHEA:12601;RHEA:10212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A5CAL1</td>\n",
       "      <td>RHEA:10992;RHEA:17905;RHEA:18657;RHEA:57408;RH...</td>\n",
       "      <td>{'RHEA:10992': 0.998241, 'RHEA:17905': 0.99824...</td>\n",
       "      <td>MESIGVLLTCPMNPYLEQELDKRFKLFRFWDFPSANDLFREHSNSI...</td>\n",
       "      <td>RHEA:16369;RHEA:18657;RHEA:18229;RHEA:10992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A7B3K3</td>\n",
       "      <td>RHEA:55380;RHEA:52592;RHEA:52584;RHEA:52588;RH...</td>\n",
       "      <td>{'RHEA:55380': 0.994077, 'RHEA:52592': 0.99407...</td>\n",
       "      <td>MFMMLKNKVAIVTGGTRGIGFAVVKKFIENGAAVSLWGSRQETVDQ...</td>\n",
       "      <td>RHEA:47496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A7AZH2</td>\n",
       "      <td>RHEA:55380;RHEA:52588;RHEA:52584;RHEA:52592;RH...</td>\n",
       "      <td>{'RHEA:55380': 0.987763, 'RHEA:52588': 0.98776...</td>\n",
       "      <td>MNFGGFIMGRFDEKIMLVTGATSGIGRAVAIRAAKEGATVVAVGRN...</td>\n",
       "      <td>RHEA:47520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>P0DX23</td>\n",
       "      <td>-;RHEA:32187;RHEA:14709;RHEA:54796</td>\n",
       "      <td>{'-': 0.998382, 'RHEA:32187': 0.7777, 'RHEA:14...</td>\n",
       "      <td>MKVLVTGATSGLGRNAVEYLRNKGISVRATGRNEAMGKLLSKMGAE...</td>\n",
       "      <td>RHEA:14981;RHEA:14929;RHEA:24616;RHEA:24612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>P0DX24</td>\n",
       "      <td>RHEA:43928;RHEA:43932;RHEA:43924;RHEA:14709;RH...</td>\n",
       "      <td>{'RHEA:43928': 0.995395, 'RHEA:43932': 0.99539...</td>\n",
       "      <td>MGDPTLRTDLGRVLVTGGSGFVGANLVTTLLERGHEVRSFDRVPSP...</td>\n",
       "      <td>RHEA:14929;RHEA:14981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>A0A7H0DNE2</td>\n",
       "      <td>RHEA:24076;RHEA:14709;RHEA:32187</td>\n",
       "      <td>{'RHEA:24076': 0.999931, 'RHEA:14709': 0.99993...</td>\n",
       "      <td>MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...</td>\n",
       "      <td>RHEA:24076;RHEA:14709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A0A7H0DNE2</td>\n",
       "      <td>RHEA:24076;RHEA:14709;RHEA:32187;RHEA:43932</td>\n",
       "      <td>{'RHEA:24076': 0.999931, 'RHEA:14709': 0.99993...</td>\n",
       "      <td>MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...</td>\n",
       "      <td>RHEA:24076;RHEA:14709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>F2XF94</td>\n",
       "      <td>RHEA:32539;-</td>\n",
       "      <td>{'RHEA:32539': 0.999942, '-': 0.7777}</td>\n",
       "      <td>MSVISIVPLVSKPCLYKSFISSTHEPKALRRPISTVGLCRRAKSVT...</td>\n",
       "      <td>RHEA:32539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>R9QMW4</td>\n",
       "      <td>RHEA:25488;RHEA:25496;RHEA:32539;-</td>\n",
       "      <td>{'RHEA:25488': 0.999269, 'RHEA:25496': 0.99926...</td>\n",
       "      <td>MSLISAVPLASSCVSKSLISSVREHKALRRAIATLQMSRPGKSVAA...</td>\n",
       "      <td>RHEA:32539;RHEA:25500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>R9QMW8</td>\n",
       "      <td>RHEA:25488;RHEA:25496;RHEA:32539;-</td>\n",
       "      <td>{'RHEA:25488': 0.999099, 'RHEA:25496': 0.99909...</td>\n",
       "      <td>MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...</td>\n",
       "      <td>RHEA:32539;RHEA:25500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>R9QMR6</td>\n",
       "      <td>RHEA:25488;RHEA:25496;RHEA:32539;-</td>\n",
       "      <td>{'RHEA:25488': 0.999062, 'RHEA:25496': 0.99906...</td>\n",
       "      <td>MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...</td>\n",
       "      <td>RHEA:32539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Q4W8C9</td>\n",
       "      <td>RHEA:14549;RHEA:10432;-;RHEA:15649;RHEA:26510</td>\n",
       "      <td>{'RHEA:14549': 0.977312, 'RHEA:10432': 0.96327...</td>\n",
       "      <td>MSASPRLGFVQCISPAGLHRMAYHEWGDPANPRVLVCAHGLTRTGR...</td>\n",
       "      <td>RHEA:10172;RHEA:11248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>G2X4G0</td>\n",
       "      <td>-;RHEA:68052;RHEA:21580</td>\n",
       "      <td>{'-': 0.999996, 'RHEA:68052': 0.972501, 'RHEA:...</td>\n",
       "      <td>MVSFTSLLAAFSVVSGVLTSPIAVVPEVNTALAKRTPSSTGTSGGF...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>A0A2H5AIX5</td>\n",
       "      <td>RHEA:72423;RHEA:72419;RHEA:19641;RHEA:36299;RH...</td>\n",
       "      <td>{'RHEA:72423': 0.999828, 'RHEA:72419': 0.99976...</td>\n",
       "      <td>MGSIPSEKETIFRSKLPDIYVPDHLPLHSYCFQNLHQFSDRPCLID...</td>\n",
       "      <td>RHEA:19641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>I3PB37</td>\n",
       "      <td>RHEA:72423;RHEA:72419;RHEA:19641;RHEA:36299;RH...</td>\n",
       "      <td>{'RHEA:72423': 0.999619, 'RHEA:72419': 0.99958...</td>\n",
       "      <td>MPMETETNQGDLIFRSKLPDIYIPKHLPLHSYCFENISEFSSRPCL...</td>\n",
       "      <td>RHEA:19641;RHEA:36299;RHEA:36251;RHEA:10132;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>A0A2H5AIY4</td>\n",
       "      <td>RHEA:72423;RHEA:72419;RHEA:19641;RHEA:72431;RH...</td>\n",
       "      <td>{'RHEA:72423': 0.999604, 'RHEA:72419': 0.99946...</td>\n",
       "      <td>MITIAESHPQIHHSPPDTTAPSTPPTATIFRSRLPDIVLSNHLPLH...</td>\n",
       "      <td>RHEA:19641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Q59104</td>\n",
       "      <td>RHEA:33051;RHEA:23872;RHEA:15933;RHEA:23188;-;...</td>\n",
       "      <td>{'RHEA:33051': 0.986769, 'RHEA:23872': 0.9801,...</td>\n",
       "      <td>MAFIYYLTHIHLDFGAVSLLKSECERIGIRRPLLVTDKGVVAAGVA...</td>\n",
       "      <td>RHEA:23948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Q88NF3</td>\n",
       "      <td>RHEA:21152;-</td>\n",
       "      <td>{'RHEA:21152': 0.999719, '-': 0.7777}</td>\n",
       "      <td>MKQIHVIDSHTGGEPTRLVMKGFPQLRGRSMAEQRDELRELHDRWR...</td>\n",
       "      <td>RHEA:21152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Q0CRQ5</td>\n",
       "      <td>RHEA:49628;RHEA:64512;RHEA:12020;-</td>\n",
       "      <td>{'RHEA:49628': 0.992385, 'RHEA:64512': 0.99039...</td>\n",
       "      <td>MTPTLLLCGSQAIQWSEDYLSSLREMLLADSALQPLVHAIRDLPQL...</td>\n",
       "      <td>RHEA:63056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>D3ZMY7</td>\n",
       "      <td>RHEA:19961;RHEA:29379;RHEA:27718;RHEA:28530;RH...</td>\n",
       "      <td>{'RHEA:19961': 0.999983, 'RHEA:29379': 0.99998...</td>\n",
       "      <td>MMTSWSDRLQNAADVPANMDKHALKKYRREAYHRVFVNRSLAMEKI...</td>\n",
       "      <td>RHEA:19961;RHEA:27718;RHEA:29383;RHEA:28530;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>A0A0H3LKL4</td>\n",
       "      <td>RHEA:27333;-</td>\n",
       "      <td>{'RHEA:27333': 0.996948, '-': 0.7777}</td>\n",
       "      <td>MQGKPRIAVIGAGLGGTAGAALMARAGFNVRLYEQAPAFSRLGAGI...</td>\n",
       "      <td>RHEA:27333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Q5C9L7</td>\n",
       "      <td>RHEA:19941;-;RHEA:31907</td>\n",
       "      <td>{'RHEA:19941': 0.998707, '-': 0.7777, 'RHEA:31...</td>\n",
       "      <td>MEMINKENLSSQAKLWNFIYGFADSLVLKSAVQLDLANIIHNHGSP...</td>\n",
       "      <td>RHEA:19941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>A0A224AM54</td>\n",
       "      <td>RHEA:51504;-</td>\n",
       "      <td>{'RHEA:51504': 0.994658, '-': 0.7777}</td>\n",
       "      <td>MSDSGGFDSHPHVALIPSAGMGHLTPFLRLAASLVQHHCRVTLITT...</td>\n",
       "      <td>RHEA:51504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>A0A224AKZ9</td>\n",
       "      <td>RHEA:51504;-</td>\n",
       "      <td>{'RHEA:51504': 0.994714, '-': 0.7777}</td>\n",
       "      <td>MSDSGGFDSHPHVALIPSAGMGHLTPFLRLAASLVQHHCRVTLITT...</td>\n",
       "      <td>RHEA:51504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>A0A517FNC4</td>\n",
       "      <td>RHEA:20585;RHEA:72667;RHEA:35511;RHEA:35499;RH...</td>\n",
       "      <td>{'RHEA:20585': 0.994162, 'RHEA:72667': 0.99416...</td>\n",
       "      <td>MDSRVLGALAALLAAAAAWVMRAAAEWLWWRPRRLERSLRSQGVDG...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>A0A517FNC9</td>\n",
       "      <td>RHEA:35499;-;RHEA:35511</td>\n",
       "      <td>{'RHEA:35499': 0.998593, '-': 0.7777, 'RHEA:35...</td>\n",
       "      <td>MVFLFPTGTIIIWVLTILLAVIPWYLLNKFWLKPKRFEKLLKAQGL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>A0A2R4LMF9</td>\n",
       "      <td>RHEA:20453;RHEA:35627;-</td>\n",
       "      <td>{'RHEA:20453': 0.994928, 'RHEA:35627': 0.99282...</td>\n",
       "      <td>MASITNHKSDPLHFVLFPFMAQGHMIPMVDIARLLAQRGLTITIVT...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>K4GMD6</td>\n",
       "      <td>RHEA:35627;-;RHEA:20453</td>\n",
       "      <td>{'RHEA:35627': 0.994723, '-': 0.7777, 'RHEA:20...</td>\n",
       "      <td>MVSEITHQSYPLHFVLFPYMAQGHMIPMVDIARLLAQRGVKITIVT...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Q59E04</td>\n",
       "      <td>-;RHEA:12532</td>\n",
       "      <td>{'-': 0.999997, 'RHEA:12532': 0.989925}</td>\n",
       "      <td>MLFRSQTHVFPGVYMMVAMLVVALSGYQVQSYSAKDILADVLMTDL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>U5NH37</td>\n",
       "      <td>RHEA:39895;-</td>\n",
       "      <td>{'RHEA:39895': 0.999989, '-': 0.7777}</td>\n",
       "      <td>MGSQETNLPPHVLIFPLPIQGHVNSMLRLAELLCLAELDITFIVSE...</td>\n",
       "      <td>RHEA:39895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>A0A517FND1</td>\n",
       "      <td>RHEA:22644;RHEA:55424;RHEA:13545;RHEA:57844;-;...</td>\n",
       "      <td>{'RHEA:22644': 0.995161, 'RHEA:55424': 0.99187...</td>\n",
       "      <td>MDSFLSQPITIVLAILSVLLYNIWKIRKPSNKFQKGMKLPQLSFAL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>F1P963</td>\n",
       "      <td>RHEA:67608;RHEA:31575;RHEA:67392;RHEA:67604;RH...</td>\n",
       "      <td>{'RHEA:67608': 0.998362, 'RHEA:31575': 0.99836...</td>\n",
       "      <td>MGTSRLYTLVLVLQPERVLLGMKKRGFGAGRWNGFGGKVQEGETIE...</td>\n",
       "      <td>RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Q7ZWC3</td>\n",
       "      <td>RHEA:67608;RHEA:31575;RHEA:67392;RHEA:67604;RH...</td>\n",
       "      <td>{'RHEA:67608': 0.994504, 'RHEA:31575': 0.99450...</td>\n",
       "      <td>MFTSKLLTLVLVVQPGRVLLGMKKRGFGAGKWNGFGGKVQTGETIE...</td>\n",
       "      <td>RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>A0A517FNC7</td>\n",
       "      <td>RHEA:54832;RHEA:54808;RHEA:55164;RHEA:55056;-;...</td>\n",
       "      <td>{'RHEA:54832': 0.990781, 'RHEA:54808': 0.98411...</td>\n",
       "      <td>MDPPSLLSLALLLLAAAAAAAFVLFPRRRSPEKKPSSGNPGSLSEL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>A0A517FND0</td>\n",
       "      <td>RHEA:55056;RHEA:17149;-;RHEA:43728</td>\n",
       "      <td>{'RHEA:55056': 0.988275, 'RHEA:17149': 0.7777,...</td>\n",
       "      <td>MESLSLIFISFITLIVFLVVSASKKRSHPSGYEPDIPKGGCPMLDS...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    uniprot_id                                           RXNRECer  \\\n",
       "5       C0HLM8                                       -;RHEA:15801   \n",
       "12      Q7XQ85                                       RHEA:21744;-   \n",
       "13  A0A0P0WIY3                                       RHEA:21744;-   \n",
       "14      Q5W6F9                                       RHEA:21744;-   \n",
       "15  A0A0P0UZP7                                       RHEA:21744;-   \n",
       "16      Q9SNN8                            RHEA:22104;-;RHEA:21744   \n",
       "17      L0TBY6                                       -;RHEA:20049   \n",
       "18      M5AWY0                            RHEA:34415;-;RHEA:56596   \n",
       "20      Q88FI7        RHEA:18049;RHEA:23352;RHEA:10212;RHEA:12268   \n",
       "21      A5CAL1  RHEA:10992;RHEA:17905;RHEA:18657;RHEA:57408;RH...   \n",
       "29      A7B3K3  RHEA:55380;RHEA:52592;RHEA:52584;RHEA:52588;RH...   \n",
       "31      A7AZH2  RHEA:55380;RHEA:52588;RHEA:52584;RHEA:52592;RH...   \n",
       "32      P0DX23                 -;RHEA:32187;RHEA:14709;RHEA:54796   \n",
       "33      P0DX24  RHEA:43928;RHEA:43932;RHEA:43924;RHEA:14709;RH...   \n",
       "34  A0A7H0DNE2                   RHEA:24076;RHEA:14709;RHEA:32187   \n",
       "35  A0A7H0DNE2        RHEA:24076;RHEA:14709;RHEA:32187;RHEA:43932   \n",
       "37      F2XF94                                       RHEA:32539;-   \n",
       "38      R9QMW4                 RHEA:25488;RHEA:25496;RHEA:32539;-   \n",
       "39      R9QMW8                 RHEA:25488;RHEA:25496;RHEA:32539;-   \n",
       "40      R9QMR6                 RHEA:25488;RHEA:25496;RHEA:32539;-   \n",
       "43      Q4W8C9      RHEA:14549;RHEA:10432;-;RHEA:15649;RHEA:26510   \n",
       "68      G2X4G0                            -;RHEA:68052;RHEA:21580   \n",
       "69  A0A2H5AIX5  RHEA:72423;RHEA:72419;RHEA:19641;RHEA:36299;RH...   \n",
       "70      I3PB37  RHEA:72423;RHEA:72419;RHEA:19641;RHEA:36299;RH...   \n",
       "71  A0A2H5AIY4  RHEA:72423;RHEA:72419;RHEA:19641;RHEA:72431;RH...   \n",
       "74      Q59104  RHEA:33051;RHEA:23872;RHEA:15933;RHEA:23188;-;...   \n",
       "75      Q88NF3                                       RHEA:21152;-   \n",
       "78      Q0CRQ5                 RHEA:49628;RHEA:64512;RHEA:12020;-   \n",
       "79      D3ZMY7  RHEA:19961;RHEA:29379;RHEA:27718;RHEA:28530;RH...   \n",
       "80  A0A0H3LKL4                                       RHEA:27333;-   \n",
       "81      Q5C9L7                            RHEA:19941;-;RHEA:31907   \n",
       "82  A0A224AM54                                       RHEA:51504;-   \n",
       "83  A0A224AKZ9                                       RHEA:51504;-   \n",
       "84  A0A517FNC4  RHEA:20585;RHEA:72667;RHEA:35511;RHEA:35499;RH...   \n",
       "85  A0A517FNC9                            RHEA:35499;-;RHEA:35511   \n",
       "86  A0A2R4LMF9                            RHEA:20453;RHEA:35627;-   \n",
       "87      K4GMD6                            RHEA:35627;-;RHEA:20453   \n",
       "88      Q59E04                                       -;RHEA:12532   \n",
       "89      U5NH37                                       RHEA:39895;-   \n",
       "90  A0A517FND1  RHEA:22644;RHEA:55424;RHEA:13545;RHEA:57844;-;...   \n",
       "91      F1P963  RHEA:67608;RHEA:31575;RHEA:67392;RHEA:67604;RH...   \n",
       "92      Q7ZWC3  RHEA:67608;RHEA:31575;RHEA:67392;RHEA:67604;RH...   \n",
       "93  A0A517FNC7  RHEA:54832;RHEA:54808;RHEA:55164;RHEA:55056;-;...   \n",
       "94  A0A517FND0                 RHEA:55056;RHEA:17149;-;RHEA:43728   \n",
       "\n",
       "                                   RXNRECer_with_prob  \\\n",
       "5             {'-': 0.999989, 'RHEA:15801': 0.940825}   \n",
       "12            {'RHEA:21744': 0.997593, '-': 0.985379}   \n",
       "13              {'RHEA:21744': 0.990111, '-': 0.7777}   \n",
       "14              {'RHEA:21744': 0.989836, '-': 0.7777}   \n",
       "15              {'RHEA:21744': 0.979187, '-': 0.7777}   \n",
       "16  {'RHEA:22104': 0.980896, '-': 0.980291, 'RHEA:...   \n",
       "17             {'-': 0.99999, 'RHEA:20049': 0.989305}   \n",
       "18  {'RHEA:34415': 0.999452, '-': 0.7777, 'RHEA:56...   \n",
       "20  {'RHEA:18049': 0.99144, 'RHEA:23352': 0.988573...   \n",
       "21  {'RHEA:10992': 0.998241, 'RHEA:17905': 0.99824...   \n",
       "29  {'RHEA:55380': 0.994077, 'RHEA:52592': 0.99407...   \n",
       "31  {'RHEA:55380': 0.987763, 'RHEA:52588': 0.98776...   \n",
       "32  {'-': 0.998382, 'RHEA:32187': 0.7777, 'RHEA:14...   \n",
       "33  {'RHEA:43928': 0.995395, 'RHEA:43932': 0.99539...   \n",
       "34  {'RHEA:24076': 0.999931, 'RHEA:14709': 0.99993...   \n",
       "35  {'RHEA:24076': 0.999931, 'RHEA:14709': 0.99993...   \n",
       "37              {'RHEA:32539': 0.999942, '-': 0.7777}   \n",
       "38  {'RHEA:25488': 0.999269, 'RHEA:25496': 0.99926...   \n",
       "39  {'RHEA:25488': 0.999099, 'RHEA:25496': 0.99909...   \n",
       "40  {'RHEA:25488': 0.999062, 'RHEA:25496': 0.99906...   \n",
       "43  {'RHEA:14549': 0.977312, 'RHEA:10432': 0.96327...   \n",
       "68  {'-': 0.999996, 'RHEA:68052': 0.972501, 'RHEA:...   \n",
       "69  {'RHEA:72423': 0.999828, 'RHEA:72419': 0.99976...   \n",
       "70  {'RHEA:72423': 0.999619, 'RHEA:72419': 0.99958...   \n",
       "71  {'RHEA:72423': 0.999604, 'RHEA:72419': 0.99946...   \n",
       "74  {'RHEA:33051': 0.986769, 'RHEA:23872': 0.9801,...   \n",
       "75              {'RHEA:21152': 0.999719, '-': 0.7777}   \n",
       "78  {'RHEA:49628': 0.992385, 'RHEA:64512': 0.99039...   \n",
       "79  {'RHEA:19961': 0.999983, 'RHEA:29379': 0.99998...   \n",
       "80              {'RHEA:27333': 0.996948, '-': 0.7777}   \n",
       "81  {'RHEA:19941': 0.998707, '-': 0.7777, 'RHEA:31...   \n",
       "82              {'RHEA:51504': 0.994658, '-': 0.7777}   \n",
       "83              {'RHEA:51504': 0.994714, '-': 0.7777}   \n",
       "84  {'RHEA:20585': 0.994162, 'RHEA:72667': 0.99416...   \n",
       "85  {'RHEA:35499': 0.998593, '-': 0.7777, 'RHEA:35...   \n",
       "86  {'RHEA:20453': 0.994928, 'RHEA:35627': 0.99282...   \n",
       "87  {'RHEA:35627': 0.994723, '-': 0.7777, 'RHEA:20...   \n",
       "88            {'-': 0.999997, 'RHEA:12532': 0.989925}   \n",
       "89              {'RHEA:39895': 0.999989, '-': 0.7777}   \n",
       "90  {'RHEA:22644': 0.995161, 'RHEA:55424': 0.99187...   \n",
       "91  {'RHEA:67608': 0.998362, 'RHEA:31575': 0.99836...   \n",
       "92  {'RHEA:67608': 0.994504, 'RHEA:31575': 0.99450...   \n",
       "93  {'RHEA:54832': 0.990781, 'RHEA:54808': 0.98411...   \n",
       "94  {'RHEA:55056': 0.988275, 'RHEA:17149': 0.7777,...   \n",
       "\n",
       "                                                  seq  \\\n",
       "5                                        GLEETYCSMKIK   \n",
       "12  MAYQGIDLLSTKAAGDDHGENSSYFDGWKAYDTNPFDLRHNRGGVI...   \n",
       "13  MVGRMLSSPEPTLSTMAMSAAHGEDSPYFAGWRAYDEDPYDPITNP...   \n",
       "14  MGVKLLADGCAGASSSPALSRVATSAAHGEGSPYFAGWKAYDEDPY...   \n",
       "15  MIMSGFHIGIYTSICLYIPLPHLEPWIISSHTPKNLNLLDCLLYCS...   \n",
       "16  MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRG...   \n",
       "17  MTHREELLPPMKWDAWGDPAAAKPLSDGVRSLLKQVVGLADSEQPE...   \n",
       "18  MQAGFFHTPYNLPTRTARQMFDWSLKLAQVCDEAGFADFMIGEHST...   \n",
       "20  MNQESISQSIAIVHPITLSHGRNAEVWDTDGKRYIDFVGGIGVLNL...   \n",
       "21  MESIGVLLTCPMNPYLEQELDKRFKLFRFWDFPSANDLFREHSNSI...   \n",
       "29  MFMMLKNKVAIVTGGTRGIGFAVVKKFIENGAAVSLWGSRQETVDQ...   \n",
       "31  MNFGGFIMGRFDEKIMLVTGATSGIGRAVAIRAAKEGATVVAVGRN...   \n",
       "32  MKVLVTGATSGLGRNAVEYLRNKGISVRATGRNEAMGKLLSKMGAE...   \n",
       "33  MGDPTLRTDLGRVLVTGGSGFVGANLVTTLLERGHEVRSFDRVPSP...   \n",
       "34  MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...   \n",
       "35  MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...   \n",
       "37  MSVISIVPLVSKPCLYKSFISSTHEPKALRRPISTVGLCRRAKSVT...   \n",
       "38  MSLISAVPLASSCVSKSLISSVREHKALRRAIATLQMSRPGKSVAA...   \n",
       "39  MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...   \n",
       "40  MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...   \n",
       "43  MSASPRLGFVQCISPAGLHRMAYHEWGDPANPRVLVCAHGLTRTGR...   \n",
       "68  MVSFTSLLAAFSVVSGVLTSPIAVVPEVNTALAKRTPSSTGTSGGF...   \n",
       "69  MGSIPSEKETIFRSKLPDIYVPDHLPLHSYCFQNLHQFSDRPCLID...   \n",
       "70  MPMETETNQGDLIFRSKLPDIYIPKHLPLHSYCFENISEFSSRPCL...   \n",
       "71  MITIAESHPQIHHSPPDTTAPSTPPTATIFRSRLPDIVLSNHLPLH...   \n",
       "74  MAFIYYLTHIHLDFGAVSLLKSECERIGIRRPLLVTDKGVVAAGVA...   \n",
       "75  MKQIHVIDSHTGGEPTRLVMKGFPQLRGRSMAEQRDELRELHDRWR...   \n",
       "78  MTPTLLLCGSQAIQWSEDYLSSLREMLLADSALQPLVHAIRDLPQL...   \n",
       "79  MMTSWSDRLQNAADVPANMDKHALKKYRREAYHRVFVNRSLAMEKI...   \n",
       "80  MQGKPRIAVIGAGLGGTAGAALMARAGFNVRLYEQAPAFSRLGAGI...   \n",
       "81  MEMINKENLSSQAKLWNFIYGFADSLVLKSAVQLDLANIIHNHGSP...   \n",
       "82  MSDSGGFDSHPHVALIPSAGMGHLTPFLRLAASLVQHHCRVTLITT...   \n",
       "83  MSDSGGFDSHPHVALIPSAGMGHLTPFLRLAASLVQHHCRVTLITT...   \n",
       "84  MDSRVLGALAALLAAAAAWVMRAAAEWLWWRPRRLERSLRSQGVDG...   \n",
       "85  MVFLFPTGTIIIWVLTILLAVIPWYLLNKFWLKPKRFEKLLKAQGL...   \n",
       "86  MASITNHKSDPLHFVLFPFMAQGHMIPMVDIARLLAQRGLTITIVT...   \n",
       "87  MVSEITHQSYPLHFVLFPYMAQGHMIPMVDIARLLAQRGVKITIVT...   \n",
       "88  MLFRSQTHVFPGVYMMVAMLVVALSGYQVQSYSAKDILADVLMTDL...   \n",
       "89  MGSQETNLPPHVLIFPLPIQGHVNSMLRLAELLCLAELDITFIVSE...   \n",
       "90  MDSFLSQPITIVLAILSVLLYNIWKIRKPSNKFQKGMKLPQLSFAL...   \n",
       "91  MGTSRLYTLVLVLQPERVLLGMKKRGFGAGRWNGFGGKVQEGETIE...   \n",
       "92  MFTSKLLTLVLVVQPGRVLLGMKKRGFGAGKWNGFGGKVQTGETIE...   \n",
       "93  MDPPSLLSLALLLLAAAAAAAFVLFPRRRSPEKKPSSGNPGSLSEL...   \n",
       "94  MESLSLIFISFITLIVFLVVSASKKRSHPSGYEPDIPKGGCPMLDS...   \n",
       "\n",
       "                                      rxn_groundtruth  \n",
       "5                                                   -  \n",
       "12                                         RHEA:21744  \n",
       "13                                         RHEA:21744  \n",
       "14                                         RHEA:21744  \n",
       "15                                         RHEA:21744  \n",
       "16                                         RHEA:21744  \n",
       "17                                                  -  \n",
       "18                                         RHEA:34415  \n",
       "20                              RHEA:12601;RHEA:10212  \n",
       "21        RHEA:16369;RHEA:18657;RHEA:18229;RHEA:10992  \n",
       "29                                         RHEA:47496  \n",
       "31                                         RHEA:47520  \n",
       "32        RHEA:14981;RHEA:14929;RHEA:24616;RHEA:24612  \n",
       "33                              RHEA:14929;RHEA:14981  \n",
       "34                              RHEA:24076;RHEA:14709  \n",
       "35                              RHEA:24076;RHEA:14709  \n",
       "37                                         RHEA:32539  \n",
       "38                              RHEA:32539;RHEA:25500  \n",
       "39                              RHEA:32539;RHEA:25500  \n",
       "40                                         RHEA:32539  \n",
       "43                              RHEA:10172;RHEA:11248  \n",
       "68                                                  -  \n",
       "69                                         RHEA:19641  \n",
       "70  RHEA:19641;RHEA:36299;RHEA:36251;RHEA:10132;RH...  \n",
       "71                                         RHEA:19641  \n",
       "74                                         RHEA:23948  \n",
       "75                                         RHEA:21152  \n",
       "78                                         RHEA:63056  \n",
       "79  RHEA:19961;RHEA:27718;RHEA:29383;RHEA:28530;RH...  \n",
       "80                                         RHEA:27333  \n",
       "81                                         RHEA:19941  \n",
       "82                                         RHEA:51504  \n",
       "83                                         RHEA:51504  \n",
       "84                                                  -  \n",
       "85                                                  -  \n",
       "86                                                  -  \n",
       "87                                                  -  \n",
       "88                                                  -  \n",
       "89                                         RHEA:39895  \n",
       "90                                                  -  \n",
       "91  RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...  \n",
       "92  RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...  \n",
       "93                                                  -  \n",
       "94                                                  -  "
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer[res_ecrecer.RXNRECer!=res_ecrecer.rxn_groundtruth]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
      "\n",
      "Guidline:\n",
      "1. **Literature Evidence Triangulation** ★最高优先级\n",
      "   a. UniProt/Sequence-based search:\n",
      "ightarrow \text{ReactID} \text{ THEN confidence} += 0.4 $\n",
      "   b. Reaction-centric search:\n",
      "     $$ \text{Match(Equation|ChEBI|SMILES) → PMIDs → Enzyme } \\cap \text{Input} $$\n",
      "   c. EC number validation:\n",
      "     $$ \text{Validate EC}_{obs} = \frac{\text{PMID}_{\text{EC-match}}}{\text{PMID}_{\text{EC-total}}} \\geq 0.6 $$\n",
      "\n",
      "2. **Structural Catalytic Competence**\n",
      "   - SMILES拓扑指纹匹配:\n",
      "     $$ \text{Tanimoto} \\geq 0.85 \text{ with Catalophore model} $$\n",
      "   - 电荷互补性验证:\n",
      "     $$ \\Delta E_{\text{elec}} = \\sum \frac{q_i q_j}{4\\pi\\epsilon r_{ij}} \\leq -15 \text{ kJ/mol} $$\n",
      "\n",
      "3. **Quantum Transition State Analysis**\n",
      "   - 催化效率计算:\n",
      "     $$ \\ln(k_{\text{cat}}/K_m) \\propto -\\Delta G^\\ddagger_{\text{calc}} + \text{文献修正项} $$\n",
      "   - 轨道相互作用验证:\n",
      "     $$ \text{HOMO-LUMO gap匹配度} \\geq 82\\% $$\n",
      "\n",
      "4. **Evolutionary & Omics Validation**\n",
      "   - 系统发育可信度:\n",
      "     $$ P_{\text{cat}} = \text{Consurf} \times \text{EC}_{phylo} \\geq 7.5 $$\n",
      "   - 组学相关性:\n",
      "     $$ r_{\text{coex}} \\geq 0.9 \text{ AND } \text{LocalScore} \\geq 0.8 $$\n",
      "\n",
      "**Important Notes:**\n",
      "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
      "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
      "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
      "\n",
      "Output your final answer in JSON format using the following template:\n",
      "{\n",
      "    \"reaction_id\": \"xxx\",\n",
      "    \"selected\": \"yes\" or \"no\",\n",
      "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
      "    \"confidence\": <number>,    // A value between 0 and 1\n",
      "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
      "}\n",
      "\n",
      "Input:\n",
      "\n",
      "{\n",
      "    \"protein information\": {\n",
      "        \"uniprot id\": \"Q9SNN8\",\n",
      "        \"protein amino acid sequence\": \"MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\"\n",
      "    },\n",
      "    \"reaction information\": {\n",
      "        \"candidate reaction 1\": {\n",
      "            \"reaction id\": \"RHEA:22104\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + nicotianamine = 3''-deamino-3''-oxonicotianamine + L-glutamate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:58249 = CHEBI:58685 + CHEBI:29985\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+][C@@H](CC[NH2+][C@@H](CC[NH+]1CC[C@H]1C([O-])=O)C([O-])=O)C([O-])=O>>[O-]C(=O)[C@H](CC[NH+]1CC[C@H]1C([O-])=O)[NH2+]CCC(=O)C([O-])=O.[NH3+][C@@H](CCC([O-])=O)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.80\"\n",
      "        },\n",
      "        \"candidate reaction 2\": {\n",
      "            \"reaction id\": \"-\",\n",
      "            \"reaction equation\": \"-\"\n",
      "        },\n",
      "        \"candidate reaction 3\": {\n",
      "            \"reaction id\": \"RHEA:21744\",\n",
      "            \"reaction equation\": \"S-adenosyl-L-methionine = 1-aminocyclopropane-1-carboxylate + H(+) + S-methyl-5'-thioadenosine\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:59789 = CHEBI:58360 + CHEBI:15378 + CHEBI:17509\",\n",
      "            \"reaction equation in SMILES format\": \"C[S+](CC[C@H]([NH3+])C([O-])=O)C[C@H]1O[C@H]([C@H](O)[C@@H]1O)n1cnc2c(N)ncnc12>>[NH3+]C1(CC1)C([O-])=O.[H+].CSC[C@H]1O[C@H]([C@H](O)[C@@H]1O)n1cnc2c(N)ncnc12\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:4.4.1.14\"\n",
      "        }\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "ix = 16\n",
    "str_query = make_make_query(uniprot_id=res_ecrecer.iloc[ix].uniprot_id, seq=res_ecrecer.iloc[ix].seq, rxn_id=res_ecrecer.iloc[ix].RXNRECer, rxn_bank=rxn, prompt=prompt5)\n",
    "\n",
    "print(str_query)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(str_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "cli1 = llmchat.Chat(name='deepseek-ai/DeepSeek-R1-Distill-Qwen-32B', \n",
    "               url='http://172.16.8.75:8123/v1', \n",
    "               api_key = 'bio12345'\n",
    "               )\n",
    "\n",
    "cli2 = llmchat.Chat(name='DeepSeek-R1-Distill-Llama-70B', \n",
    "               url='http://172.16.8.76:8123/v1', \n",
    "               api_key = 'bio12345'\n",
    "                )\n",
    "\n",
    "cli3 = llmchat.Chat(name='deepseek-r1:70b', \n",
    "               url='http://bdcop1.tibhpc.net:11434/v1', \n",
    "               api_key = 'bio12345'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ChatCompletion(id='1959451ab5134958b2e3bf4b3d8677fd', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Okay, so I\\'m trying to figure out which reactions the given protein sequence can catalyze. The protein has the UniProt ID Q9SNN8, and its sequence is provided. There are three candidate reactions, but one of them is just a dash, meaning the protein can\\'t catalyze that one. So I need to focus on the other two reactions.\\n\\nFirst, I\\'ll look at Reaction 1, which has the ID RHEA:22104. The reaction equation is 2-oxoglutarate converting to 3\\'\\'-deamino-3\\'\\'-oxonicotianamine and L-glutamate. The EC number is 2.6.1.80, which I think is a transaminase. Transaminases usually have specific active sites with certain residues like lysine, which is often involved in the catalysis. Looking at the protein sequence, I should check for conserved domains or motifs that match transaminases. Maybe there\\'s a PLP (pyridoxal phosphate) binding site since transaminases often use PLP as a cofactor. I\\'ll scan the sequence for PLP-binding motifs, which might have glycine-rich regions or specific patterns around lysine residues.\\n\\nNext, Reaction 3 has the ID RHEA:21744, and it\\'s about S-adenosyl-L-methionine converting into 1-aminocyclopropane-1-carboxylate, H+, and S-methyl-5\\'-thioadenosine. The EC number here is 4.4.1.14, which I believe is ACC synthase. ACC synthase is involved in ethylene biosynthesis in plants. These enzymes usually have a certain structure, maybe with a conserved active site that includes residues like cysteine or histidine. I\\'ll look for any motifs that resemble those found in ACC synthases, such as specific amino acid patterns around the active site.\\n\\nNow, looking at the protein sequence, I\\'ll check for any known domains. The sequence starts with MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH.\\n\\nI notice that there are several proline-rich regions, like around the beginning with multiple K\\'s (KKKKKK). Also, there are some repeats, like the PPPPPPGSPTA part. These might be involved in protein interactions or specific catalytic functions. Looking for active site residues, I see some aspartic acid (D) and glutamic acid (E) residues, which are common in enzymes. For example, the sequence has VRDDDSLYYAGLRRCAADPYHPVTNPSGI... which includes several D\\'s and E\\'s. These could be part of the active site.\\n\\nFor Reaction 1, if it\\'s a transaminase, I should look for a conserved lysine residue that forms a Schiff base with the substrate. Scanning through, I see a few K\\'s, like in the beginning, but I\\'m not sure if they\\'re in the right context. There\\'s also a region with RHEA:22104\\'s EC number being 2.6.1.80, which is a transaminase. So maybe the protein has a domain that matches this.\\n\\nFor Reaction 3, ACC synthase, I recall that these enzymes have a conserved region around the active site, often with a cysteine residue that acts as a nucleophile. Looking through the sequence, I see some cysteines, like in the CIEPGWFRCCFT part. The CC might indicate a metal-binding site or a disulfide bond, but I\\'m not sure if that\\'s relevant here. Alternatively, the region around VDDLEDFDKGR might have some significance, but I\\'m not certain.\\n\\nI also need to consider the overall structure. Transaminases often have a certain fold, maybe a two-domain structure with the active site in between. ACC synthases might have a different structure, perhaps with a regulatory domain. Without more specific domain predictions, it\\'s a bit hard, but I can make educated guesses based on the sequence features.\\n\\nAnother approach is to look for homologs. If this protein is similar to known transaminases or ACC synthases, that would help. But since I can\\'t access external databases, I\\'ll have to rely on sequence motifs. For example, the presence of a specific PLP-binding motif would support Reaction 1. Alternatively, motifs like GCVK or similar might indicate ACC synthase activity.\\n\\nIn the end, I think Reaction 1 is more likely because the EC number points to a transaminase, and the sequence has features that could support that, like possible PLP-binding regions. Reaction 3\\'s ACC synthase is less clear, so I might not select it unless there are strong motifs indicating otherwise. Since Reaction 2 is a dash, I have to include it if I don\\'t find any other reactions, but since Reaction 1 seems plausible, I\\'ll go with that.\\n</think>\\n\\n```json\\n[\\n    {\\n        \"reaction_id\": \"RHEA:22104\",\\n        \"selected\": \"yes\",\\n        \"rank\": 1,\\n        \"confidence\": 0.8,\\n        \"reason\": \"The protein sequence contains features consistent with a transaminase, such as potential PLP-binding motifs and conserved residues like lysine, which are typical for this enzyme class. The reaction\\'s EC number (2.6.1.80) aligns with transaminase activity.\"\\n    },\\n    {\\n        \"reaction_id\": \"-\",\\n        \"selected\": \"no\",\\n        \"rank\": 0,\\n        \"confidence\": 0,\\n        \"reason\": \"This reaction is represented by a single dash, indicating the protein cannot catalyze this reaction.\"\\n    },\\n    {\\n        \"reaction_id\": \"RHEA:21744\",\\n        \"selected\": \"no\",\\n        \"rank\": 0,\\n        \"confidence\": 0.2,\\n        \"reason\": \"While the EC number suggests ACC synthase activity, the sequence lacks strong motifs typically associated with this enzyme, such as specific cysteine residues in the active site.\"\\n    }\\n]\\n```', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None), matched_stop=128001)], created=1739774754, model='deepseek-ai/DeepSeek-R1-Distill-Qwen-32B', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=1555, prompt_tokens=1268, total_tokens=2823, completion_tokens_details=None, prompt_tokens_details=None))\n",
      "Okay, so I'm trying to figure out which reactions the given protein sequence can catalyze. The protein has the UniProt ID Q9SNN8, and its sequence is provided. There are three candidate reactions, but one of them is just a dash, meaning the protein can't catalyze that one. So I need to focus on the other two reactions.\n",
      "\n",
      "First, I'll look at Reaction 1, which has the ID RHEA:22104. The reaction equation is 2-oxoglutarate converting to 3''-deamino-3''-oxonicotianamine and L-glutamate. The EC number is 2.6.1.80, which I think is a transaminase. Transaminases usually have specific active sites with certain residues like lysine, which is often involved in the catalysis. Looking at the protein sequence, I should check for conserved domains or motifs that match transaminases. Maybe there's a PLP (pyridoxal phosphate) binding site since transaminases often use PLP as a cofactor. I'll scan the sequence for PLP-binding motifs, which might have glycine-rich regions or specific patterns around lysine residues.\n",
      "\n",
      "Next, Reaction 3 has the ID RHEA:21744, and it's about S-adenosyl-L-methionine converting into 1-aminocyclopropane-1-carboxylate, H+, and S-methyl-5'-thioadenosine. The EC number here is 4.4.1.14, which I believe is ACC synthase. ACC synthase is involved in ethylene biosynthesis in plants. These enzymes usually have a certain structure, maybe with a conserved active site that includes residues like cysteine or histidine. I'll look for any motifs that resemble those found in ACC synthases, such as specific amino acid patterns around the active site.\n",
      "\n",
      "Now, looking at the protein sequence, I'll check for any known domains. The sequence starts with MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH.\n",
      "\n",
      "I notice that there are several proline-rich regions, like around the beginning with multiple K's (KKKKKK). Also, there are some repeats, like the PPPPPPGSPTA part. These might be involved in protein interactions or specific catalytic functions. Looking for active site residues, I see some aspartic acid (D) and glutamic acid (E) residues, which are common in enzymes. For example, the sequence has VRDDDSLYYAGLRRCAADPYHPVTNPSGI... which includes several D's and E's. These could be part of the active site.\n",
      "\n",
      "For Reaction 1, if it's a transaminase, I should look for a conserved lysine residue that forms a Schiff base with the substrate. Scanning through, I see a few K's, like in the beginning, but I'm not sure if they're in the right context. There's also a region with RHEA:22104's EC number being 2.6.1.80, which is a transaminase. So maybe the protein has a domain that matches this.\n",
      "\n",
      "For Reaction 3, ACC synthase, I recall that these enzymes have a conserved region around the active site, often with a cysteine residue that acts as a nucleophile. Looking through the sequence, I see some cysteines, like in the CIEPGWFRCCFT part. The CC might indicate a metal-binding site or a disulfide bond, but I'm not sure if that's relevant here. Alternatively, the region around VDDLEDFDKGR might have some significance, but I'm not certain.\n",
      "\n",
      "I also need to consider the overall structure. Transaminases often have a certain fold, maybe a two-domain structure with the active site in between. ACC synthases might have a different structure, perhaps with a regulatory domain. Without more specific domain predictions, it's a bit hard, but I can make educated guesses based on the sequence features.\n",
      "\n",
      "Another approach is to look for homologs. If this protein is similar to known transaminases or ACC synthases, that would help. But since I can't access external databases, I'll have to rely on sequence motifs. For example, the presence of a specific PLP-binding motif would support Reaction 1. Alternatively, motifs like GCVK or similar might indicate ACC synthase activity.\n",
      "\n",
      "In the end, I think Reaction 1 is more likely because the EC number points to a transaminase, and the sequence has features that could support that, like possible PLP-binding regions. Reaction 3's ACC synthase is less clear, so I might not select it unless there are strong motifs indicating otherwise. Since Reaction 2 is a dash, I have to include it if I don't find any other reactions, but since Reaction 1 seems plausible, I'll go with that.\n",
      "</think>\n",
      "\n",
      "```json\n",
      "[\n",
      "    {\n",
      "        \"reaction_id\": \"RHEA:22104\",\n",
      "        \"selected\": \"yes\",\n",
      "        \"rank\": 1,\n",
      "        \"confidence\": 0.8,\n",
      "        \"reason\": \"The protein sequence contains features consistent with a transaminase, such as potential PLP-binding motifs and conserved residues like lysine, which are typical for this enzyme class. The reaction's EC number (2.6.1.80) aligns with transaminase activity.\"\n",
      "    },\n",
      "    {\n",
      "        \"reaction_id\": \"-\",\n",
      "        \"selected\": \"no\",\n",
      "        \"rank\": 0,\n",
      "        \"confidence\": 0,\n",
      "        \"reason\": \"This reaction is represented by a single dash, indicating the protein cannot catalyze this reaction.\"\n",
      "    },\n",
      "    {\n",
      "        \"reaction_id\": \"RHEA:21744\",\n",
      "        \"selected\": \"no\",\n",
      "        \"rank\": 0,\n",
      "        \"confidence\": 0.2,\n",
      "        \"reason\": \"While the EC number suggests ACC synthase activity, the sequence lacks strong motifs typically associated with this enzyme, such as specific cysteine residues in the active site.\"\n",
      "    }\n",
      "]\n",
      "```\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ChatCompletion(id='1959451ab5134958b2e3bf4b3d8677fd', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Okay, so I\\'m trying to figure out which reactions the given protein sequence can catalyze. The protein has the UniProt ID Q9SNN8, and its sequence is provided. There are three candidate reactions, but one of them is just a dash, meaning the protein can\\'t catalyze that one. So I need to focus on the other two reactions.\\n\\nFirst, I\\'ll look at Reaction 1, which has the ID RHEA:22104. The reaction equation is 2-oxoglutarate converting to 3\\'\\'-deamino-3\\'\\'-oxonicotianamine and L-glutamate. The EC number is 2.6.1.80, which I think is a transaminase. Transaminases usually have specific active sites with certain residues like lysine, which is often involved in the catalysis. Looking at the protein sequence, I should check for conserved domains or motifs that match transaminases. Maybe there\\'s a PLP (pyridoxal phosphate) binding site since transaminases often use PLP as a cofactor. I\\'ll scan the sequence for PLP-binding motifs, which might have glycine-rich regions or specific patterns around lysine residues.\\n\\nNext, Reaction 3 has the ID RHEA:21744, and it\\'s about S-adenosyl-L-methionine converting into 1-aminocyclopropane-1-carboxylate, H+, and S-methyl-5\\'-thioadenosine. The EC number here is 4.4.1.14, which I believe is ACC synthase. ACC synthase is involved in ethylene biosynthesis in plants. These enzymes usually have a certain structure, maybe with a conserved active site that includes residues like cysteine or histidine. I\\'ll look for any motifs that resemble those found in ACC synthases, such as specific amino acid patterns around the active site.\\n\\nNow, looking at the protein sequence, I\\'ll check for any known domains. The sequence starts with MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH.\\n\\nI notice that there are several proline-rich regions, like around the beginning with multiple K\\'s (KKKKKK). Also, there are some repeats, like the PPPPPPGSPTA part. These might be involved in protein interactions or specific catalytic functions. Looking for active site residues, I see some aspartic acid (D) and glutamic acid (E) residues, which are common in enzymes. For example, the sequence has VRDDDSLYYAGLRRCAADPYHPVTNPSGI... which includes several D\\'s and E\\'s. These could be part of the active site.\\n\\nFor Reaction 1, if it\\'s a transaminase, I should look for a conserved lysine residue that forms a Schiff base with the substrate. Scanning through, I see a few K\\'s, like in the beginning, but I\\'m not sure if they\\'re in the right context. There\\'s also a region with RHEA:22104\\'s EC number being 2.6.1.80, which is a transaminase. So maybe the protein has a domain that matches this.\\n\\nFor Reaction 3, ACC synthase, I recall that these enzymes have a conserved region around the active site, often with a cysteine residue that acts as a nucleophile. Looking through the sequence, I see some cysteines, like in the CIEPGWFRCCFT part. The CC might indicate a metal-binding site or a disulfide bond, but I\\'m not sure if that\\'s relevant here. Alternatively, the region around VDDLEDFDKGR might have some significance, but I\\'m not certain.\\n\\nI also need to consider the overall structure. Transaminases often have a certain fold, maybe a two-domain structure with the active site in between. ACC synthases might have a different structure, perhaps with a regulatory domain. Without more specific domain predictions, it\\'s a bit hard, but I can make educated guesses based on the sequence features.\\n\\nAnother approach is to look for homologs. If this protein is similar to known transaminases or ACC synthases, that would help. But since I can\\'t access external databases, I\\'ll have to rely on sequence motifs. For example, the presence of a specific PLP-binding motif would support Reaction 1. Alternatively, motifs like GCVK or similar might indicate ACC synthase activity.\\n\\nIn the end, I think Reaction 1 is more likely because the EC number points to a transaminase, and the sequence has features that could support that, like possible PLP-binding regions. Reaction 3\\'s ACC synthase is less clear, so I might not select it unless there are strong motifs indicating otherwise. Since Reaction 2 is a dash, I have to include it if I don\\'t find any other reactions, but since Reaction 1 seems plausible, I\\'ll go with that.\\n</think>\\n\\n```json\\n[\\n    {\\n        \"reaction_id\": \"RHEA:22104\",\\n        \"selected\": \"yes\",\\n        \"rank\": 1,\\n        \"confidence\": 0.8,\\n        \"reason\": \"The protein sequence contains features consistent with a transaminase, such as potential PLP-binding motifs and conserved residues like lysine, which are typical for this enzyme class. The reaction\\'s EC number (2.6.1.80) aligns with transaminase activity.\"\\n    },\\n    {\\n        \"reaction_id\": \"-\",\\n        \"selected\": \"no\",\\n        \"rank\": 0,\\n        \"confidence\": 0,\\n        \"reason\": \"This reaction is represented by a single dash, indicating the protein cannot catalyze this reaction.\"\\n    },\\n    {\\n        \"reaction_id\": \"RHEA:21744\",\\n        \"selected\": \"no\",\\n        \"rank\": 0,\\n        \"confidence\": 0.2,\\n        \"reason\": \"While the EC number suggests ACC synthase activity, the sequence lacks strong motifs typically associated with this enzyme, such as specific cysteine residues in the active site.\"\\n    }\\n]\\n```', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None), matched_stop=128001)], created=1739774754, model='deepseek-ai/DeepSeek-R1-Distill-Qwen-32B', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=1555, prompt_tokens=1268, total_tokens=2823, completion_tokens_details=None, prompt_tokens_details=None))"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cli1.chat(message=str_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ChatCompletion(id='c8e4df94e0194ff4a9cb32d317024a23', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Okay, so I\\'m trying to figure out which reactions the protein Q9SNN8 can catalyze. Let me start by looking at the information provided.\\n\\nFirst, the protein sequence is given. I know that certain motifs or domains in the sequence can indicate what kind of reactions the protein might catalyze. I\\'ll need to analyze the sequence for any conserved domains or active site residues that are characteristic of specific enzymes.\\n\\nLooking at the reaction information, there are three candidate reactions. The second one is a dash, which means the protein can\\'t catalyze that reaction. So I can ignore that one.\\n\\nNow, focusing on the first and third reactions. Reaction 1 is RHEA:22104, which involves 2-oxoglutarate and nicotianamine. The enzyme commission number is EC:2.6.1.80. I remember that EC numbers can give clues about the type of reaction. EC 2.6.1 typically involves enzymes that transfer amino groups, like aminotransferases. But I\\'m not entirely sure, so I should double-check.\\n\\nReaction 3 is RHEA:21744, which involves S-adenosyl-L-methionine breaking down into 1-aminocyclopropane-1-carboxylate and other products. The EC number here is 4.4.1.14. EC 4.4.1 suggests a methyltransferase, which transfers methyl groups. But wait, the reaction equation shows the breakdown of S-adenosyl-L-methionine, which is a common methyl donor. So maybe this is a decarboxylase or something else.\\n\\nLooking at the protein sequence, I should check for any motifs that match known enzymes. For EC 2.6.1.80, I think the enzyme might have a specific active site with residues that bind 2-oxoglutarate. Maybe there are histidine or glutamic acid residues in the sequence that are common in such enzymes.\\n\\nFor EC 4.4.1.14, the enzyme is likely a methyltransferase. These often have a conserved motif like GxGxG or other specific residues that bind the methyl group. I\\'ll scan the sequence for such motifs.\\n\\nWait, the protein sequence is quite long. Let me look for any known domains. I see a part that has \"MEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\". Hmm, I notice some repeats and possible motifs, but I\\'m not sure which ones are significant.\\n\\nI think I should use some bioinformatics tools or databases to check for conserved domains. For example, using InterProScan or looking up the UniProt ID Q9SNN8 might give more information. But since I can\\'t access external databases right now, I\\'ll have to rely on the sequence alone.\\n\\nAnother approach is to look for active site residues. For EC 2.6.1.80, which is a type of aminotransferase, the active site might include a lysine residue that acts as a base. I\\'ll scan the sequence for lysine (K) residues in a context that might form part of an active site.\\n\\nLooking through the sequence, I see several K\\'s. For example, near the beginning: \"MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\".\\n\\nI see a lot of K\\'s, but without knowing the exact structure, it\\'s hard to say. Maybe the presence of multiple K\\'s suggests a possible aminotransferase activity.\\n\\nFor EC 4.4.1.14, which is a methyltransferase, I should look for motifs like GxGxG or other conserved regions. Scanning the sequence, I don\\'t immediately see such motifs, but there are several repeats and possible structural motifs that might indicate a different function.\\n\\nConsidering the reaction equations, Reaction 1 involves the conversion of 2-oxoglutarate and nicotianamine into other compounds, which seems like a transfer of an amino group. Reaction 3 involves the breakdown of S-adenosyl-L-methionine, which is more about releasing a methyl group or amines.\\n\\nGiven that the protein has multiple K\\'s, which are often involved in binding or catalysis, I\\'m leaning towards Reaction 1 being more likely. However, I\\'m not entirely certain because I don\\'t have the structural information.\\n\\nI should also consider the confidence level. Since I\\'m not 100% sure, I\\'ll give Reaction 1 a high confidence but not the maximum. Reaction 3 might be possible, but without clear motifs, I\\'m less confident.\\n\\nSo, I think Reaction 1 is the most likely, followed by Reaction 3 if any, but since Reaction 3\\'s EC number doesn\\'t directly match the sequence features I can see, I might not select it. Alternatively, if the sequence has other features, maybe it\\'s possible.\\n\\nWait, another thought: S-adenosyl-L-methionine is often involved in methylation reactions, and the enzyme that breaks it down might be a decarboxylase. But I\\'m not sure if the sequence supports that.\\n\\nIn conclusion, based on the presence of K residues and the EC number, I\\'ll select Reaction 1 as the most likely, with a high confidence. Reaction 3 might be possible, but without more evidence, I\\'ll rank it lower or not select it.\\n</think>\\n\\n```json\\n[\\n    {\\n        \"reaction_id\": \"RHEA:22104\",\\n        \"selected\": \"yes\",\\n        \"rank\": 1,\\n        \"confidence\": 0.85,\\n        \"reason\": \"The protein sequence contains multiple lysine residues, which are often involved in catalytic activity, particularly in aminotransferases. The reaction involves the transfer of an amino group, aligning with the enzyme\\'s potential function.\"\\n    },\\n    {\\n        \"reaction_id\": \"RHEA:21744\",\\n        \"selected\": \"no\",\\n        \"rank\": 2,\\n        \"confidence\": 0.30,\\n        \"reason\": \"While S-adenosyl-L-methionine is a common substrate, the sequence lacks clear motifs typically associated with methyltransferases or decarboxylases, making this reaction less likely.\"\\n    },\\n    {\\n        \"reaction_id\": \"-\",\\n        \"selected\": \"no\",\\n        \"rank\": 3,\\n        \"confidence\": 0.00,\\n        \"reason\": \"This reaction is marked as non-catalytic.\"\\n    }\\n]\\n```', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None), matched_stop=151643)], created=1739774834, model='DeepSeek-R1-Distill-Llama-70B', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=1855, prompt_tokens=1306, total_tokens=3161, completion_tokens_details=None, prompt_tokens_details=None))\n",
      "Okay, so I'm trying to figure out which reactions the protein Q9SNN8 can catalyze. Let me start by looking at the information provided.\n",
      "\n",
      "First, the protein sequence is given. I know that certain motifs or domains in the sequence can indicate what kind of reactions the protein might catalyze. I'll need to analyze the sequence for any conserved domains or active site residues that are characteristic of specific enzymes.\n",
      "\n",
      "Looking at the reaction information, there are three candidate reactions. The second one is a dash, which means the protein can't catalyze that reaction. So I can ignore that one.\n",
      "\n",
      "Now, focusing on the first and third reactions. Reaction 1 is RHEA:22104, which involves 2-oxoglutarate and nicotianamine. The enzyme commission number is EC:2.6.1.80. I remember that EC numbers can give clues about the type of reaction. EC 2.6.1 typically involves enzymes that transfer amino groups, like aminotransferases. But I'm not entirely sure, so I should double-check.\n",
      "\n",
      "Reaction 3 is RHEA:21744, which involves S-adenosyl-L-methionine breaking down into 1-aminocyclopropane-1-carboxylate and other products. The EC number here is 4.4.1.14. EC 4.4.1 suggests a methyltransferase, which transfers methyl groups. But wait, the reaction equation shows the breakdown of S-adenosyl-L-methionine, which is a common methyl donor. So maybe this is a decarboxylase or something else.\n",
      "\n",
      "Looking at the protein sequence, I should check for any motifs that match known enzymes. For EC 2.6.1.80, I think the enzyme might have a specific active site with residues that bind 2-oxoglutarate. Maybe there are histidine or glutamic acid residues in the sequence that are common in such enzymes.\n",
      "\n",
      "For EC 4.4.1.14, the enzyme is likely a methyltransferase. These often have a conserved motif like GxGxG or other specific residues that bind the methyl group. I'll scan the sequence for such motifs.\n",
      "\n",
      "Wait, the protein sequence is quite long. Let me look for any known domains. I see a part that has \"MEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\". Hmm, I notice some repeats and possible motifs, but I'm not sure which ones are significant.\n",
      "\n",
      "I think I should use some bioinformatics tools or databases to check for conserved domains. For example, using InterProScan or looking up the UniProt ID Q9SNN8 might give more information. But since I can't access external databases right now, I'll have to rely on the sequence alone.\n",
      "\n",
      "Another approach is to look for active site residues. For EC 2.6.1.80, which is a type of aminotransferase, the active site might include a lysine residue that acts as a base. I'll scan the sequence for lysine (K) residues in a context that might form part of an active site.\n",
      "\n",
      "Looking through the sequence, I see several K's. For example, near the beginning: \"MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\".\n",
      "\n",
      "I see a lot of K's, but without knowing the exact structure, it's hard to say. Maybe the presence of multiple K's suggests a possible aminotransferase activity.\n",
      "\n",
      "For EC 4.4.1.14, which is a methyltransferase, I should look for motifs like GxGxG or other conserved regions. Scanning the sequence, I don't immediately see such motifs, but there are several repeats and possible structural motifs that might indicate a different function.\n",
      "\n",
      "Considering the reaction equations, Reaction 1 involves the conversion of 2-oxoglutarate and nicotianamine into other compounds, which seems like a transfer of an amino group. Reaction 3 involves the breakdown of S-adenosyl-L-methionine, which is more about releasing a methyl group or amines.\n",
      "\n",
      "Given that the protein has multiple K's, which are often involved in binding or catalysis, I'm leaning towards Reaction 1 being more likely. However, I'm not entirely certain because I don't have the structural information.\n",
      "\n",
      "I should also consider the confidence level. Since I'm not 100% sure, I'll give Reaction 1 a high confidence but not the maximum. Reaction 3 might be possible, but without clear motifs, I'm less confident.\n",
      "\n",
      "So, I think Reaction 1 is the most likely, followed by Reaction 3 if any, but since Reaction 3's EC number doesn't directly match the sequence features I can see, I might not select it. Alternatively, if the sequence has other features, maybe it's possible.\n",
      "\n",
      "Wait, another thought: S-adenosyl-L-methionine is often involved in methylation reactions, and the enzyme that breaks it down might be a decarboxylase. But I'm not sure if the sequence supports that.\n",
      "\n",
      "In conclusion, based on the presence of K residues and the EC number, I'll select Reaction 1 as the most likely, with a high confidence. Reaction 3 might be possible, but without more evidence, I'll rank it lower or not select it.\n",
      "</think>\n",
      "\n",
      "```json\n",
      "[\n",
      "    {\n",
      "        \"reaction_id\": \"RHEA:22104\",\n",
      "        \"selected\": \"yes\",\n",
      "        \"rank\": 1,\n",
      "        \"confidence\": 0.85,\n",
      "        \"reason\": \"The protein sequence contains multiple lysine residues, which are often involved in catalytic activity, particularly in aminotransferases. The reaction involves the transfer of an amino group, aligning with the enzyme's potential function.\"\n",
      "    },\n",
      "    {\n",
      "        \"reaction_id\": \"RHEA:21744\",\n",
      "        \"selected\": \"no\",\n",
      "        \"rank\": 2,\n",
      "        \"confidence\": 0.30,\n",
      "        \"reason\": \"While S-adenosyl-L-methionine is a common substrate, the sequence lacks clear motifs typically associated with methyltransferases or decarboxylases, making this reaction less likely.\"\n",
      "    },\n",
      "    {\n",
      "        \"reaction_id\": \"-\",\n",
      "        \"selected\": \"no\",\n",
      "        \"rank\": 3,\n",
      "        \"confidence\": 0.00,\n",
      "        \"reason\": \"This reaction is marked as non-catalytic.\"\n",
      "    }\n",
      "]\n",
      "```\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ChatCompletion(id='c8e4df94e0194ff4a9cb32d317024a23', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='Okay, so I\\'m trying to figure out which reactions the protein Q9SNN8 can catalyze. Let me start by looking at the information provided.\\n\\nFirst, the protein sequence is given. I know that certain motifs or domains in the sequence can indicate what kind of reactions the protein might catalyze. I\\'ll need to analyze the sequence for any conserved domains or active site residues that are characteristic of specific enzymes.\\n\\nLooking at the reaction information, there are three candidate reactions. The second one is a dash, which means the protein can\\'t catalyze that reaction. So I can ignore that one.\\n\\nNow, focusing on the first and third reactions. Reaction 1 is RHEA:22104, which involves 2-oxoglutarate and nicotianamine. The enzyme commission number is EC:2.6.1.80. I remember that EC numbers can give clues about the type of reaction. EC 2.6.1 typically involves enzymes that transfer amino groups, like aminotransferases. But I\\'m not entirely sure, so I should double-check.\\n\\nReaction 3 is RHEA:21744, which involves S-adenosyl-L-methionine breaking down into 1-aminocyclopropane-1-carboxylate and other products. The EC number here is 4.4.1.14. EC 4.4.1 suggests a methyltransferase, which transfers methyl groups. But wait, the reaction equation shows the breakdown of S-adenosyl-L-methionine, which is a common methyl donor. So maybe this is a decarboxylase or something else.\\n\\nLooking at the protein sequence, I should check for any motifs that match known enzymes. For EC 2.6.1.80, I think the enzyme might have a specific active site with residues that bind 2-oxoglutarate. Maybe there are histidine or glutamic acid residues in the sequence that are common in such enzymes.\\n\\nFor EC 4.4.1.14, the enzyme is likely a methyltransferase. These often have a conserved motif like GxGxG or other specific residues that bind the methyl group. I\\'ll scan the sequence for such motifs.\\n\\nWait, the protein sequence is quite long. Let me look for any known domains. I see a part that has \"MEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\". Hmm, I notice some repeats and possible motifs, but I\\'m not sure which ones are significant.\\n\\nI think I should use some bioinformatics tools or databases to check for conserved domains. For example, using InterProScan or looking up the UniProt ID Q9SNN8 might give more information. But since I can\\'t access external databases right now, I\\'ll have to rely on the sequence alone.\\n\\nAnother approach is to look for active site residues. For EC 2.6.1.80, which is a type of aminotransferase, the active site might include a lysine residue that acts as a base. I\\'ll scan the sequence for lysine (K) residues in a context that might form part of an active site.\\n\\nLooking through the sequence, I see several K\\'s. For example, near the beginning: \"MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\".\\n\\nI see a lot of K\\'s, but without knowing the exact structure, it\\'s hard to say. Maybe the presence of multiple K\\'s suggests a possible aminotransferase activity.\\n\\nFor EC 4.4.1.14, which is a methyltransferase, I should look for motifs like GxGxG or other conserved regions. Scanning the sequence, I don\\'t immediately see such motifs, but there are several repeats and possible structural motifs that might indicate a different function.\\n\\nConsidering the reaction equations, Reaction 1 involves the conversion of 2-oxoglutarate and nicotianamine into other compounds, which seems like a transfer of an amino group. Reaction 3 involves the breakdown of S-adenosyl-L-methionine, which is more about releasing a methyl group or amines.\\n\\nGiven that the protein has multiple K\\'s, which are often involved in binding or catalysis, I\\'m leaning towards Reaction 1 being more likely. However, I\\'m not entirely certain because I don\\'t have the structural information.\\n\\nI should also consider the confidence level. Since I\\'m not 100% sure, I\\'ll give Reaction 1 a high confidence but not the maximum. Reaction 3 might be possible, but without clear motifs, I\\'m less confident.\\n\\nSo, I think Reaction 1 is the most likely, followed by Reaction 3 if any, but since Reaction 3\\'s EC number doesn\\'t directly match the sequence features I can see, I might not select it. Alternatively, if the sequence has other features, maybe it\\'s possible.\\n\\nWait, another thought: S-adenosyl-L-methionine is often involved in methylation reactions, and the enzyme that breaks it down might be a decarboxylase. But I\\'m not sure if the sequence supports that.\\n\\nIn conclusion, based on the presence of K residues and the EC number, I\\'ll select Reaction 1 as the most likely, with a high confidence. Reaction 3 might be possible, but without more evidence, I\\'ll rank it lower or not select it.\\n</think>\\n\\n```json\\n[\\n    {\\n        \"reaction_id\": \"RHEA:22104\",\\n        \"selected\": \"yes\",\\n        \"rank\": 1,\\n        \"confidence\": 0.85,\\n        \"reason\": \"The protein sequence contains multiple lysine residues, which are often involved in catalytic activity, particularly in aminotransferases. The reaction involves the transfer of an amino group, aligning with the enzyme\\'s potential function.\"\\n    },\\n    {\\n        \"reaction_id\": \"RHEA:21744\",\\n        \"selected\": \"no\",\\n        \"rank\": 2,\\n        \"confidence\": 0.30,\\n        \"reason\": \"While S-adenosyl-L-methionine is a common substrate, the sequence lacks clear motifs typically associated with methyltransferases or decarboxylases, making this reaction less likely.\"\\n    },\\n    {\\n        \"reaction_id\": \"-\",\\n        \"selected\": \"no\",\\n        \"rank\": 3,\\n        \"confidence\": 0.00,\\n        \"reason\": \"This reaction is marked as non-catalytic.\"\\n    }\\n]\\n```', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None), matched_stop=151643)], created=1739774834, model='DeepSeek-R1-Distill-Llama-70B', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=1855, prompt_tokens=1306, total_tokens=3161, completion_tokens_details=None, prompt_tokens_details=None))"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cli2.chat(message= str_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_cli3 = cli3.chat(message=str_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChatCompletion(id='chatcmpl-327', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='<think>\\nOkay, so I\\'m trying to figure out which reactions the given protein might catalyze. Let\\'s start by looking at the information provided.\\n\\nThe protein has a UniProt ID of Q9SNN8 and a pretty long amino acid sequence. The task is to analyze this sequence against three candidate reactions. Two of them have valid IDs, RHEA:22104 and RHEA:21744, while the third one is just a dash, meaning it\\'s not catalyzed.\\n\\nFirst, I\\'ll look at Reaction 1 (RHEA:22104). The reaction equation involves 2-oxoglutarate and nicotianamine producing 3\\'\\'-deamino-3\\'\\'-oxonicotianamine and L-glutamate. The EC number is 2.6.1.80, which I think corresponds to an aminotransferase enzyme. These enzymes usually have specific active site residues like lysine, which is often involved in the catalysis.\\n\\nLooking at the protein sequence, I notice several conserved domains and motifs. There\\'s a stretch with \"GVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH.\" This part has a lot of glycine and valine residues, which might be part of an active site. Also, the presence of lysine (K) and arginine (R) could indicate potential catalytic sites.\\n\\nReaction 3 (RHEA:21744) involves S-adenosyl-L-methionine converting into several products, associated with EC 4.4.1.14, which is an ACC synthase. These enzymes are usually involved in ethylene biosynthesis and have specific motifs like the PLP-binding domain.\\n\\nLooking at the sequence again, I see a region with \"SPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH.\" This area has multiple arginines and lysines, which are often found in enzyme active sites. The presence of a possible PLP-binding site (like a lysine residue) could support this reaction.\\n\\nReaction 2 is just a dash, so it\\'s not catalyzed by the protein. Now, considering that proteins can have multiple active sites, I think both Reaction 1 and Reaction 3 might be catalyzed. However, based on the sequence features, Reaction 1 seems more likely because of the clear presence of residues matching an aminotransferase. Reaction 3 is possible but less confident.\\n\\nSo, I\\'ll rank Reaction 1 as high confidence and Reaction 3 as moderate, with Reaction 2 not applicable.\\n</think>\\n\\nBased on the analysis of the protein sequence Q9SNN8 against the provided reactions:\\n\\n1. **Reaction 1 (RHEA:22104)**:\\n   - **EC Number**: 2.6.1.80 (aminotransferase)\\n   - **Confidence**: High\\n   - **Reasoning**: The presence of lysine and arginine residues suggests a potential catalytic site, aligning with the characteristics of an aminotransferase enzyme.\\n\\n2. **Reaction 3 (RHEA:21744)**:\\n   - **EC Number**: 4.4.1.14 (ACC synthase)\\n   - **Confidence**: Moderate\\n   - **Reasoning**: The sequence contains arginines and lysines, which are common in enzyme active sites, possibly indicating a PLP-binding domain.\\n\\n3. **Reaction 2**:\\n   - **Status**: Not catalyzed\\n   - **Reasoning**: Indicated by the dash, meaning no catalytic activity for this reaction.\\n\\n**Conclusion**: The protein is likely to catalyze Reaction 1 with high confidence and may also catalyze Reaction 3 with moderate confidence.', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1739779270, model='deepseek-r1:70b', object='chat.completion', service_tier=None, system_fingerprint='fp_ollama', usage=CompletionUsage(completion_tokens=1137, prompt_tokens=1173, total_tokens=2310, completion_tokens_details=None, prompt_tokens_details=None))"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_cli3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "Okay, so I'm trying to figure out which reactions the given protein might catalyze. Let's start by looking at the information provided.\n",
      "\n",
      "The protein has a UniProt ID of Q9SNN8 and a pretty long amino acid sequence. The task is to analyze this sequence against three candidate reactions. Two of them have valid IDs, RHEA:22104 and RHEA:21744, while the third one is just a dash, meaning it's not catalyzed.\n",
      "\n",
      "First, I'll look at Reaction 1 (RHEA:22104). The reaction equation involves 2-oxoglutarate and nicotianamine producing 3''-deamino-3''-oxonicotianamine and L-glutamate. The EC number is 2.6.1.80, which I think corresponds to an aminotransferase enzyme. These enzymes usually have specific active site residues like lysine, which is often involved in the catalysis.\n",
      "\n",
      "Looking at the protein sequence, I notice several conserved domains and motifs. There's a stretch with \"GVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH.\" This part has a lot of glycine and valine residues, which might be part of an active site. Also, the presence of lysine (K) and arginine (R) could indicate potential catalytic sites.\n",
      "\n",
      "Reaction 3 (RHEA:21744) involves S-adenosyl-L-methionine converting into several products, associated with EC 4.4.1.14, which is an ACC synthase. These enzymes are usually involved in ethylene biosynthesis and have specific motifs like the PLP-binding domain.\n",
      "\n",
      "Looking at the sequence again, I see a region with \"SPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH.\" This area has multiple arginines and lysines, which are often found in enzyme active sites. The presence of a possible PLP-binding site (like a lysine residue) could support this reaction.\n",
      "\n",
      "Reaction 2 is just a dash, so it's not catalyzed by the protein. Now, considering that proteins can have multiple active sites, I think both Reaction 1 and Reaction 3 might be catalyzed. However, based on the sequence features, Reaction 1 seems more likely because of the clear presence of residues matching an aminotransferase. Reaction 3 is possible but less confident.\n",
      "\n",
      "So, I'll rank Reaction 1 as high confidence and Reaction 3 as moderate, with Reaction 2 not applicable.\n",
      "</think>\n",
      "\n",
      "Based on the analysis of the protein sequence Q9SNN8 against the provided reactions:\n",
      "\n",
      "1. **Reaction 1 (RHEA:22104)**:\n",
      "   - **EC Number**: 2.6.1.80 (aminotransferase)\n",
      "   - **Confidence**: High\n",
      "   - **Reasoning**: The presence of lysine and arginine residues suggests a potential catalytic site, aligning with the characteristics of an aminotransferase enzyme.\n",
      "\n",
      "2. **Reaction 3 (RHEA:21744)**:\n",
      "   - **EC Number**: 4.4.1.14 (ACC synthase)\n",
      "   - **Confidence**: Moderate\n",
      "   - **Reasoning**: The sequence contains arginines and lysines, which are common in enzyme active sites, possibly indicating a PLP-binding domain.\n",
      "\n",
      "3. **Reaction 2**:\n",
      "   - **Status**: Not catalyzed\n",
      "   - **Reasoning**: Indicated by the dash, meaning no catalytic activity for this reaction.\n",
      "\n",
      "**Conclusion**: The protein is likely to catalyze Reaction 1 with high confidence and may also catalyze Reaction 3 with moderate confidence.\n"
     ]
    }
   ],
   "source": [
    "print(res_cli3.choices[0].message.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "You are a biochemical expert. Given a protein sequence and a set of potential catalyzed reactions, please analyze and determine the most likely catalyzed reactions. \n",
      "    The information is provided in JSON format.\n",
      "    Important note: \n",
      "    1. If a reaction is represented by a single dash (\"-\"), it means the protein cannot catalyze a reaction. \n",
      "    Please consider this in your analysis. If all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in the output to indicate that the protein is not catalytically active for any provided reaction.\n",
      "    2. Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
      "    3. Provide a brief explanation for each selection, citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
      "\n",
      "Please give out the finale output in json format Output template:\n",
      "{\n",
      "    \"reaction_id\": \"xxx\",\n",
      "    \"selected\": \"yes\" or \"no\",\n",
      "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
      "    \"confidence\": <number>,      // Value between 0 and 1\n",
      "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
      "}\n",
      "    \n",
      "Input:\n",
      "{\n",
      "    \"protein information\": {\n",
      "        \"uniprot id\": \"Q9SNN8\",\n",
      "        \"protein amino acid sequence\": \"MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRGGLVLGSLIPCALFYFFQLYIKRNRASPPPPPGSPTAASAAAVSPIHRSLSRGLLAPRAALPAISARGASVRDDDSLYYAGLRRCAADPYHPVTNPSGIIQLGLAENYLSLDLVGRWMEEHAAEAASMAGGEDEDERELSIRGLAAYQPYDGILALKMALAGFMRQIMQGSVSFEPSQVVITSGATPAMEILSFCLADPGNAFLVPSPYYPGWDRDIKWRTGIELIPVPCRSTDNFNISITALEIAYNQAKKRGIKVRGVLISNPNNPTGSFVPKQTLHDLLEFAAEKNIHLISDEVFAGSTYGSGKFVSVAEVVDDLEDFDKGRVHIIYGLSKDLSLAGFRVGVIYSYNESIVTAAAKIARFSSVSTPTQRLLVAMLSDQKFISDYLKVNRERLRKMYHLFVDALDQVGIECYKSSGGFYCWADMSKFIRSYSEKGERKLWDRLLEEAKVNVTPGSSCHCIEPGWFRCCFTTLSEHDIPVLVQRLRTITDSHKPNH\"\n",
      "    },\n",
      "    \"reaction information\": {\n",
      "        \"candidate reaction 1\": {\n",
      "            \"reaction id\": \"RHEA:22104\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + nicotianamine = 3''-deamino-3''-oxonicotianamine + L-glutamate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:58249 = CHEBI:58685 + CHEBI:29985\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+][C@@H](CC[NH2+][C@@H](CC[NH+]1CC[C@H]1C([O-])=O)C([O-])=O)C([O-])=O>>[O-]C(=O)[C@H](CC[NH+]1CC[C@H]1C([O-])=O)[NH2+]CCC(=O)C([O-])=O.[NH3+][C@@H](CCC([O-])=O)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.80\"\n",
      "        },\n",
      "        \"candidate reaction 2\": {\n",
      "            \"reaction id\": \"-\",\n",
      "            \"reaction equation\": \"-\"\n",
      "        },\n",
      "        \"candidate reaction 3\": {\n",
      "            \"reaction id\": \"RHEA:21744\",\n",
      "            \"reaction equation\": \"S-adenosyl-L-methionine = 1-aminocyclopropane-1-carboxylate + H(+) + S-methyl-5'-thioadenosine\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:59789 = CHEBI:58360 + CHEBI:15378 + CHEBI:17509\",\n",
      "            \"reaction equation in SMILES format\": \"C[S+](CC[C@H]([NH3+])C([O-])=O)C[C@H]1O[C@H]([C@H](O)[C@@H]1O)n1cnc2c(N)ncnc12>>[NH3+]C1(CC1)C([O-])=O.[H+].CSC[C@H]1O[C@H]([C@H](O)[C@@H]1O)n1cnc2c(N)ncnc12\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:4.4.1.14\"\n",
      "        }\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(str_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rxnrecer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
