{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 基于LLM bagging 的质控+ 可解释\n",
    "> zhenkun.shi@tib.cas.cn   \n",
    "> 2025-05-15 \n",
    "\n",
    "## 1. 导入必要的包"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 192 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "import sys,os\n",
    "sys.path.insert(0, os.path.dirname(os.path.realpath('__file__')))\n",
    "sys.path.insert(1,'../../')\n",
    "from config import conf as cfg\n",
    "import pandas as pd\n",
    "import json\n",
    "import plotly.graph_objects as go\n",
    "from tools import btools\n",
    "from tools import uniprottool as uptool\n",
    "from tqdm import tqdm\n",
    "import rxnrecer as production\n",
    "\n",
    "from modules.llm import chat as llmchat\n",
    "import tools.bioFunctionLib as bfl\n",
    "from IPython.display import HTML\n",
    "from pandarallel import pandarallel # 导入pandaralle\n",
    "FIRST_TIME_RUN = False\n",
    "pandarallel.initialize(progress_bar=False)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. 加载数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>functionCounts</th>\n",
       "      <th>ec_specific_level</th>\n",
       "      <th>isenzyme</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A9JLI2</td>\n",
       "      <td>MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A9JLI3</td>\n",
       "      <td>MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A9JLI5</td>\n",
       "      <td>MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id                                                seq reaction_id  \\\n",
       "0     A9JLI2  MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...           -   \n",
       "1     A9JLI3  MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...           -   \n",
       "2     A9JLI5  MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...           -   \n",
       "\n",
       "  ec_number  functionCounts  ec_specific_level  isenzyme  \\\n",
       "0         -               0                  0     False   \n",
       "1         -               0                  0     False   \n",
       "2         -               0                  0     False   \n",
       "\n",
       "                                               label  \n",
       "0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_test = pd.read_feather(cfg.FILE_DS_TEST)\n",
    "ds_test.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. RXNRECer EXP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A9JLI2</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997, 'RHEA:10012': None, 'RHEA:1002...</td>\n",
       "      <td>MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A9JLI3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998, 'RHEA:10012': None, 'RHEA:1002...</td>\n",
       "      <td>MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A9JLI5</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0, 'RHEA:10012': None, 'RHEA:10024': N...</td>\n",
       "      <td>MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id RXNRECer                                 RXNRECer_with_prob  \\\n",
       "0     A9JLI2        -  {'-': 0.999997, 'RHEA:10012': None, 'RHEA:1002...   \n",
       "1     A9JLI3        -  {'-': 0.999998, 'RHEA:10012': None, 'RHEA:1002...   \n",
       "2     A9JLI5        -  {'-': 1.0, 'RHEA:10012': None, 'RHEA:10024': N...   \n",
       "\n",
       "                                                 seq rxn_groundtruth  \n",
       "0  MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...               -  \n",
       "1  MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...               -  \n",
       "2  MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...               -  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if FIRST_TIME_RUN:\n",
    "    res_ecrecer = production.step_by_step_prediction(input_data=ds_test[['uniprot_id', 'seq']], Ensemble=True, batch_size=100)\n",
    "    res_ecrecer.to_feather('res/res_ecrecer.feather')\n",
    "    res_ecrecer = res_ecrecer.rename(columns={'input_id':'uniprot_id'})\n",
    "    res_ecrecer = res_ecrecer.merge(ds_test[['uniprot_id', 'seq', 'reaction_id']], on='uniprot_id', how='left').rename(columns={'reaction_id':'rxn_groundtruth'})\n",
    "else:\n",
    "    res_ecrecer = pd.read_feather('res/res_ecrecer.feather')\n",
    "    \n",
    "res_ecrecer.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A9JLI2</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A9JLI3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A9JLI5</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "      <td>MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id RXNRECer RXNRECer_with_prob  \\\n",
       "0     A9JLI2        -    {'-': 0.999997}   \n",
       "1     A9JLI3        -    {'-': 0.999998}   \n",
       "2     A9JLI5        -         {'-': 1.0}   \n",
       "\n",
       "                                                 seq rxn_groundtruth  \n",
       "0  MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...               -  \n",
       "1  MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...               -  \n",
       "2  MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...               -  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer['RXNRECer_with_prob'] = res_ecrecer.RXNRECer_with_prob.apply(lambda x: {k: v for k, v in x.items() if v is not None})\n",
    "res_ecrecer.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# res_ecrecer.to_pickle('res/res_ecrecer.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'-': 0.980291, 'RHEA:21744': 0.7777, 'RHEA:22104': 0.98089599609375}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_data = {k: v for k, v in res_ecrecer.RXNRECer_with_prob.values[16].items() if v is not None}\n",
    "filtered_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# res_ecrecer.to_pickle(cfg.FILE_CASE_RESULTS_RXNRECER_REACTION)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Quality Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>equation</th>\n",
       "      <th>chebi_id</th>\n",
       "      <th>ec_number</th>\n",
       "      <th>equation_chebi</th>\n",
       "      <th>equation_smiles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHEA:22636</td>\n",
       "      <td>dCTP + H2O = dCMP + diphosphate + H(+)</td>\n",
       "      <td>CHEBI:61481;CHEBI:15377;CHEBI:57566;CHEBI:3301...</td>\n",
       "      <td>EC:3.6.1.9;EC:3.6.1.12;EC:3.6.1.65</td>\n",
       "      <td>CHEBI:61481 + CHEBI:15377 = CHEBI:57566 + CHEB...</td>\n",
       "      <td>Nc1ccn([C@H]2C[C@H](O)[C@@H](COP([O-])(=O)OP([...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RHEA:22640</td>\n",
       "      <td>NADP(+) + sphinganine = 3-oxosphinganine + H(+...</td>\n",
       "      <td>CHEBI:58349;CHEBI:57817;CHEBI:58299;CHEBI:1537...</td>\n",
       "      <td>EC:1.1.1.102</td>\n",
       "      <td>CHEBI:58349 + CHEBI:57817 = CHEBI:58299 + CHEB...</td>\n",
       "      <td>NC(=O)c1ccc[n+](c1)[C@@H]1O[C@H](COP([O-])(=O)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RHEA:22644</td>\n",
       "      <td>O2 + protopine + reduced [NADPH--hemoprotein r...</td>\n",
       "      <td>CHEBI:15379;CHEBI:16415;CHEBI:57618;CHEBI:1710...</td>\n",
       "      <td>EC:1.14.14.98</td>\n",
       "      <td>CHEBI:15379 + CHEBI:16415 + CHEBI:57618 = CHEB...</td>\n",
       "      <td>O=O.CN1CCc2cc3OCOc3cc2C(=O)Cc2ccc3OCOc3c2C1.Cc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id                                           equation  \\\n",
       "0  RHEA:22636             dCTP + H2O = dCMP + diphosphate + H(+)   \n",
       "1  RHEA:22640  NADP(+) + sphinganine = 3-oxosphinganine + H(+...   \n",
       "2  RHEA:22644  O2 + protopine + reduced [NADPH--hemoprotein r...   \n",
       "\n",
       "                                            chebi_id  \\\n",
       "0  CHEBI:61481;CHEBI:15377;CHEBI:57566;CHEBI:3301...   \n",
       "1  CHEBI:58349;CHEBI:57817;CHEBI:58299;CHEBI:1537...   \n",
       "2  CHEBI:15379;CHEBI:16415;CHEBI:57618;CHEBI:1710...   \n",
       "\n",
       "                            ec_number  \\\n",
       "0  EC:3.6.1.9;EC:3.6.1.12;EC:3.6.1.65   \n",
       "1                        EC:1.1.1.102   \n",
       "2                       EC:1.14.14.98   \n",
       "\n",
       "                                      equation_chebi  \\\n",
       "0  CHEBI:61481 + CHEBI:15377 = CHEBI:57566 + CHEB...   \n",
       "1  CHEBI:58349 + CHEBI:57817 = CHEBI:58299 + CHEB...   \n",
       "2  CHEBI:15379 + CHEBI:16415 + CHEBI:57618 = CHEB...   \n",
       "\n",
       "                                     equation_smiles  \n",
       "0  Nc1ccn([C@H]2C[C@H](O)[C@@H](COP([O-])(=O)OP([...  \n",
       "1  NC(=O)c1ccc[n+](c1)[C@@H]1O[C@H](COP([O-])(=O)...  \n",
       "2  O=O.CN1CCc2cc3OCOc3cc2C(=O)Cc2ccc3OCOc3c2C1.Cc...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 4. Load RXN\n",
    "rxn = pd.read_feather(cfg.FILE_RHEA_REACTION)\n",
    "rxn.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P0DTH6</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MKKIFLTMTPLPYMNHLPENHYFLSLTVVISIIHLFTTCVPQHNHS...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Q7XQ85</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'-': 0.9853789806365967, 'RHEA:21744': 0.997593}</td>\n",
       "      <td>MAYQGIDLLSTKAAGDDHGENSSYFDGWKAYDTNPFDLRHNRGGVI...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A0A0P0WIY3</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:21744': 0.990111}</td>\n",
       "      <td>MVGRMLSSPEPTLSTMAMSAAHGEDSPYFAGWRAYDEDPYDPITNP...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Q5W6F9</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:21744': 0.989836}</td>\n",
       "      <td>MGVKLLADGCAGASSSPALSRVATSAAHGEGSPYFAGWKAYDEDPY...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A0P0UZP7</td>\n",
       "      <td>RHEA:21744;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:21744': 0.97918701171875}</td>\n",
       "      <td>MIMSGFHIGIYTSICLYIPLPHLEPWIISSHTPKNLNLLDCLLYCS...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Q9SNN8</td>\n",
       "      <td>RHEA:22104;-;RHEA:21744</td>\n",
       "      <td>{'-': 0.980291, 'RHEA:21744': 0.7777, 'RHEA:22...</td>\n",
       "      <td>MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRG...</td>\n",
       "      <td>RHEA:21744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>L0TBY6</td>\n",
       "      <td>-;RHEA:20049</td>\n",
       "      <td>{'-': 0.99999, 'RHEA:20049': 0.9893050193786621}</td>\n",
       "      <td>MTHREELLPPMKWDAWGDPAAAKPLSDGVRSLLKQVVGLADSEQPE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M5AWY0</td>\n",
       "      <td>RHEA:34415;-;RHEA:56596</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:34415': 0.9994519948959351...</td>\n",
       "      <td>MQAGFFHTPYNLPTRTARQMFDWSLKLAQVCDEAGFADFMIGEHST...</td>\n",
       "      <td>RHEA:34415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Q6I621</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MWKGFLSKLPRKTSASGRGADLDSGQCSNGAGNGNPIQRTSSCGSI...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Q88FI7</td>\n",
       "      <td>RHEA:18049;RHEA:23352;RHEA:10212;RHEA:12268</td>\n",
       "      <td>{'RHEA:10212': 0.7777, 'RHEA:12268': 0.7777, '...</td>\n",
       "      <td>MNQESISQSIAIVHPITLSHGRNAEVWDTDGKRYIDFVGGIGVLNL...</td>\n",
       "      <td>RHEA:12601;RHEA:10212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A5CAL1</td>\n",
       "      <td>RHEA:10992;RHEA:17905;RHEA:18657;RHEA:57704;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:10776': 0.997845, 'RHEA:10...</td>\n",
       "      <td>MESIGVLLTCPMNPYLEQELDKRFKLFRFWDFPSANDLFREHSNSI...</td>\n",
       "      <td>RHEA:16369;RHEA:18657;RHEA:18229;RHEA:10992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Q2PS07</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MKLFIILATATLLIAATQATYPRDEGFDLGETQMSSKCMRQVKMNE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>E9NX73</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>MKLFIILATATLLIAATQAKYLRDEGFDLGETQMSSKCTRQVKMME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P93198</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999994}</td>\n",
       "      <td>AALLVALLFVANAAAFRTTITTMEIDEDIDNPRRRGEGCREQIQRQ...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    uniprot_id                                           RXNRECer  \\\n",
       "11      P0DTH6                                                  -   \n",
       "12      Q7XQ85                                       RHEA:21744;-   \n",
       "13  A0A0P0WIY3                                       RHEA:21744;-   \n",
       "14      Q5W6F9                                       RHEA:21744;-   \n",
       "15  A0A0P0UZP7                                       RHEA:21744;-   \n",
       "16      Q9SNN8                            RHEA:22104;-;RHEA:21744   \n",
       "17      L0TBY6                                       -;RHEA:20049   \n",
       "18      M5AWY0                            RHEA:34415;-;RHEA:56596   \n",
       "19      Q6I621                                                  -   \n",
       "20      Q88FI7        RHEA:18049;RHEA:23352;RHEA:10212;RHEA:12268   \n",
       "21      A5CAL1  RHEA:10992;RHEA:17905;RHEA:18657;RHEA:57704;RH...   \n",
       "22      Q2PS07                                                  -   \n",
       "23      E9NX73                                                  -   \n",
       "24      P93198                                                  -   \n",
       "\n",
       "                                   RXNRECer_with_prob  \\\n",
       "11                                    {'-': 0.999998}   \n",
       "12  {'-': 0.9853789806365967, 'RHEA:21744': 0.997593}   \n",
       "13              {'-': 0.7777, 'RHEA:21744': 0.990111}   \n",
       "14              {'-': 0.7777, 'RHEA:21744': 0.989836}   \n",
       "15      {'-': 0.7777, 'RHEA:21744': 0.97918701171875}   \n",
       "16  {'-': 0.980291, 'RHEA:21744': 0.7777, 'RHEA:22...   \n",
       "17   {'-': 0.99999, 'RHEA:20049': 0.9893050193786621}   \n",
       "18  {'-': 0.7777, 'RHEA:34415': 0.9994519948959351...   \n",
       "19                                    {'-': 0.999998}   \n",
       "20  {'RHEA:10212': 0.7777, 'RHEA:12268': 0.7777, '...   \n",
       "21  {'-': 0.7777, 'RHEA:10776': 0.997845, 'RHEA:10...   \n",
       "22                                    {'-': 0.999997}   \n",
       "23                                    {'-': 0.999996}   \n",
       "24                                    {'-': 0.999994}   \n",
       "\n",
       "                                                  seq  \\\n",
       "11  MKKIFLTMTPLPYMNHLPENHYFLSLTVVISIIHLFTTCVPQHNHS...   \n",
       "12  MAYQGIDLLSTKAAGDDHGENSSYFDGWKAYDTNPFDLRHNRGGVI...   \n",
       "13  MVGRMLSSPEPTLSTMAMSAAHGEDSPYFAGWRAYDEDPYDPITNP...   \n",
       "14  MGVKLLADGCAGASSSPALSRVATSAAHGEGSPYFAGWKAYDEDPY...   \n",
       "15  MIMSGFHIGIYTSICLYIPLPHLEPWIISSHTPKNLNLLDCLLYCS...   \n",
       "16  MRRSGNGGAAKKKKKRSASAASERRPRADGGMRIVVPLQGVVQGRG...   \n",
       "17  MTHREELLPPMKWDAWGDPAAAKPLSDGVRSLLKQVVGLADSEQPE...   \n",
       "18  MQAGFFHTPYNLPTRTARQMFDWSLKLAQVCDEAGFADFMIGEHST...   \n",
       "19  MWKGFLSKLPRKTSASGRGADLDSGQCSNGAGNGNPIQRTSSCGSI...   \n",
       "20  MNQESISQSIAIVHPITLSHGRNAEVWDTDGKRYIDFVGGIGVLNL...   \n",
       "21  MESIGVLLTCPMNPYLEQELDKRFKLFRFWDFPSANDLFREHSNSI...   \n",
       "22  MKLFIILATATLLIAATQATYPRDEGFDLGETQMSSKCMRQVKMNE...   \n",
       "23  MKLFIILATATLLIAATQAKYLRDEGFDLGETQMSSKCTRQVKMME...   \n",
       "24  AALLVALLFVANAAAFRTTITTMEIDEDIDNPRRRGEGCREQIQRQ...   \n",
       "\n",
       "                                rxn_groundtruth  \n",
       "11                                            -  \n",
       "12                                   RHEA:21744  \n",
       "13                                   RHEA:21744  \n",
       "14                                   RHEA:21744  \n",
       "15                                   RHEA:21744  \n",
       "16                                   RHEA:21744  \n",
       "17                                            -  \n",
       "18                                   RHEA:34415  \n",
       "19                                            -  \n",
       "20                        RHEA:12601;RHEA:10212  \n",
       "21  RHEA:16369;RHEA:18657;RHEA:18229;RHEA:10992  \n",
       "22                                            -  \n",
       "23                                            -  \n",
       "24                                            -  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer.iloc[11:25]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rxn_detail(rxn_id, rxn_bank):\n",
    "    if rxn_id == '-':\n",
    "        return {\n",
    "            'reaction id': '-',\n",
    "            'reaction equation': '-'\n",
    "        }\n",
    "    \n",
    "    rxn_record = \n",
    "    [rxn_bank.reaction_id == rxn_id]\n",
    "    \n",
    "    if rxn_record.empty:\n",
    "        return {}  # Avoid errors if the record is missing\n",
    "\n",
    "    rxn_record = rxn_record.iloc[0]  # Take the first match\n",
    "\n",
    "    return {\n",
    "        'reaction id': rxn_record.reaction_id,\n",
    "        'reaction equation': rxn_record.equation,\n",
    "        'reaction equation in ChEBI format': rxn_record.equation_chebi,\n",
    "        'reaction equation in SMILES format': rxn_record.equation_smiles,\n",
    "        'reaction associated Enzyme Commission Number': rxn_record.ec_number\n",
    "    }\n",
    "\n",
    "def make_make_query(uniprot_id, seq, rxn_id, rxn_bank, prompt):\n",
    "    if not rxn_id or cfg.SPLITER not in rxn_id:\n",
    "        return {}\n",
    "\n",
    "    rxns = rxn_id.split(cfg.SPLITER)\n",
    "\n",
    "\n",
    "\n",
    "    query_info = {\n",
    "        'protein information': {\n",
    "            'uniprot id': uniprot_id,\n",
    "            'protein amino acid sequence': seq\n",
    "        },\n",
    "        'reaction information': {\n",
    "            f'candidate reaction {i+1}': get_rxn_detail(rxn, rxn_bank)\n",
    "            for i, rxn in enumerate(rxns)\n",
    "        }\n",
    "    }\n",
    "\n",
    "    llm_query = prompt  + json.dumps(query_info, indent=4, ensure_ascii=False)\n",
    "\n",
    "    return llm_query\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt0 = \"\"\"You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data.\n",
    "\n",
    "### Guidelines:\n",
    "1. **Input Data:**\n",
    "- **Protein Information:** Includes the Uniprot ID and the protein's amino acid sequence.\n",
    "- **Reaction Information:** A list of candidate reactions. Each reaction is provided in:\n",
    "    - Standard reaction equation\n",
    "    - Reaction equation in ChEBI format\n",
    "    - Reaction equation in SMILES format\n",
    "\n",
    "2. **Sequence Analysis & Functional Inference:**\n",
    "- **Basic Properties:** Consider the protein's length, molecular weight, and amino acid composition to see if they match known enzyme classes.\n",
    "- **Domain & Active Site Identification:** Look for conserved domains and key catalytic residues (or cofactor-binding motifs) that are critical for catalysis.\n",
    "- **Structural Considerations:** Evaluate whether the predicted secondary/tertiary structure supports the formation of an active site suitable for the reaction’s substrate.\n",
    "- **Database & Homology Evidence:** Use similarities to annotated proteins and known enzyme families to infer likely catalytic activity.\n",
    "\n",
    "3. **Handling Non-Catalytic Cases:**\n",
    "- A candidate reaction denoted by a single dash (\"-\") indicates that the protein is predicted not to catalyze that particular reaction.\n",
    "- **Important:** If all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in the output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "\n",
    "4. **Multiple Reactions:**\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "\n",
    "5. **Ranking:**\n",
    "- Rank the selected reactions by likelihood (1 being the most likely).\n",
    "\n",
    "6. **Justification:**\n",
    "- Provide a brief explanation for each selection, citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "### Expected Output (in JSON format):\n",
    "```json\n",
    "[\n",
    "    {\n",
    "        \"reaction_id\": \"xxx\",\n",
    "        \"selected\": \"yes\" or \"no\",\n",
    "        \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "        \"confidence\": <number>,      // Value between 0 and 1\n",
    "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "    },\n",
    "    ...\n",
    "]\n",
    "\"\"\"\n",
    "\n",
    "prompt1 = \"\"\"\n",
    "You are a biochemical expert. Given a protein sequence and a set of potential catalyzed reactions, please analyze and determine the most likely catalyzed reactions. \n",
    "    The information is provided in JSON format.\n",
    "    Important note: If a reaction is represented by a single dash (\"-\"), it means the protein cannot catalyze a reaction. \n",
    "    Please consider this in your analysis.\n",
    "    \n",
    "Guidelines:\n",
    "    1. **Handling Non-Catalytic Cases:**\n",
    "    - A candidate reaction denoted by a single dash (\"-\") indicates that the protein is predicted not to catalyze that particular reaction.\n",
    "    - **Important:** If all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in the output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "\n",
    "    2. **Multiple Reactions:**\n",
    "    - Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "\n",
    "    3. **Ranking:**\n",
    "    - Rank the selected reactions by likelihood (1 being the most likely).\n",
    "\n",
    "    4. **Justification:**\n",
    "    - Provide a brief explanation for each selection, citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "    \n",
    "    5. **Confidence Scoring:**  \n",
    "    - For each reaction, assign a confidence score between 0 and 1 based on how well the protein sequence matches known features (e.g., conserved domains, motifs, known homologs) for the reaction. Use this score to determine the ranking.  \n",
    "    - If none of the reactions are plausible, assign the highest confidence to the dash (\"-\") reaction to indicate a strong negative prediction.\n",
    "\n",
    "    \n",
    "Expected Output (in JSON format):\n",
    "    ```json\n",
    "    [\n",
    "        {\n",
    "            \"reaction_id\": \"xxx\",\n",
    "            \"selected\": \"yes\" or \"no\",\n",
    "            \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "            \"confidence\": <number>,      // Value between 0 and 1\n",
    "            \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "        },\n",
    "        ...\n",
    "    ]\n",
    "⚠️ Please do not use any other field name such as `\"analysis\"`. Only `\"results\"` is allowed.\n",
    "\n",
    "Input:\n",
    "\"\"\"\n",
    "\n",
    "prompt2 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Important Notes:\n",
    "1. If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "2. Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "3. For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,      // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "prompt3 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Structural Chemistry Analysis**\n",
    "   - Parse SMILES to identify reaction centers (e.g., bond cleavage sites)\n",
    "   $$ \\text{Example: } \\text{C-S bond in } \\mathrm{CHEBI:59789} \\rightarrow \\mathrm{CH_3S^- + adenosyl} $$\n",
    "   - Map ChEBI IDs to functional groups (e.g., thioether in CHEBI:59789)\n",
    "\n",
    "2. **Sequence-Structure Mapping**\n",
    "   - Detect catalytic triads using PROSITE patterns:\n",
    "     $$ \\text{Cys-X-X-Cys} \\Rightarrow \\text{Sulfur metabolism} $$\n",
    "     $$ \\text{Ser-Lys-Asp} \\Rightarrow \\text{Hydrolase activity} $$\n",
    "   - Verify cofactor binding with PROSITE PDOC00005\n",
    "\n",
    "3. **Dynamic Compatibility Check**\n",
    "   - Calculate active site volume using:\n",
    "     $$ V_{\\text{site}} = \\frac{4}{3}\\pi r_{\\text{substrate}}^3 \\pm 15\\% $$\n",
    "   - Validate stereochemical constraints from SMILES chirality tags (@, @@)\n",
    "\n",
    "4. **Conflict Resolution Protocol**\n",
    "   - Prioritize reaction chemistry over EC numbers when conflicts exist\n",
    "   - Use UniProtKB/Swiss-Prot annotations (e.g., FT ACT_SITE) as tiebreaker\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,      // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "prompt3 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Structural Chemistry Analysis**\n",
    "   - Parse SMILES to identify reaction centers (e.g., bond cleavage sites)\n",
    "   $$ \\text{Example: } \\text{C-S bond in } \\mathrm{CHEBI:59789} \\rightarrow \\mathrm{CH_3S^- + adenosyl} $$\n",
    "   - Map ChEBI IDs to functional groups (e.g., thioether in CHEBI:59789)\n",
    "\n",
    "2. **Sequence-Structure Mapping**\n",
    "   - Detect catalytic triads using PROSITE patterns:\n",
    "     $$ \\text{Cys-X-X-Cys} \\Rightarrow \\text{Sulfur metabolism} $$\n",
    "     $$ \\text{Ser-Lys-Asp} \\Rightarrow \\text{Hydrolase activity} $$\n",
    "   - Verify cofactor binding with PROSITE PDOC00005\n",
    "\n",
    "3. **Dynamic Compatibility Check**\n",
    "   - Calculate active site volume using:\n",
    "     $$ V_{\\text{site}} = \\frac{4}{3}\\pi r_{\\text{substrate}}^3 \\pm 15\\% $$\n",
    "   - Validate stereochemical constraints from SMILES chirality tags (@, @@)\n",
    "\n",
    "4. **Conflict Resolution Protocol**\n",
    "   - Prioritize reaction chemistry over EC numbers when conflicts exist\n",
    "   - Use UniProtKB/Swiss-Prot annotations (e.g., FT ACT_SITE) as tiebreaker\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,      // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "prompt4 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Omni-enzyme Structural Profiling**\n",
    "   - Perform topological analysis of SMILES using:\n",
    "     $$ \\text{Graph neural networks for reaction center detection} $$\n",
    "     $$ \\text{Electrostatic potential mapping (EPM) for charge complementarity} $$\n",
    "   - Cross-reference ChEBI IDs with MetaCyc reaction pathways:\n",
    "     $$ \\forall R \\in \\text{Reactions}, \\exists \\text{Substrate-Enzyme PCMatrix} \\geq 0.7 $$\n",
    "\n",
    "2. **Quantum Biochemical Evaluation**\n",
    "   - Calculate catalytic proficiency using:\n",
    "     $$ k_{\\text{cat}}/K_m = \\frac{k_B T}{h} e^{-\\Delta G^\\ddagger/RT} $$\n",
    "   - Simulate transition states with:\n",
    "     $$ \\text{QM/MM modeling of active site frontier orbitals} $$\n",
    "     $$ \\text{Reaction coordinate analysis (IRC)} $$\n",
    "\n",
    "3. **Evolutionary Dynamics Integration**\n",
    "   - Calculate phylogenetic catalytic propensity:\n",
    "     $$ P_{\\text{cat}} = \\frac{\\sum \\text{Consurf scores}}{\\text{Active site residues}} \\times \\text{EC_specific_phyletic_pattern} $$\n",
    "   - Detect convergent evolution signatures through:\n",
    "     $$ \\text{Parallel substitution analysis in catalytic residues} $$\n",
    "     $$ \\text{Structural phylogeny using DALI Z-scores} $$\n",
    "\n",
    "4. **Multi-scale Mechanistic Validation**\n",
    "   - Verify allosteric regulation potential:\n",
    "     $$ \\text{Ensemble docking across conformational states} $$\n",
    "     $$ \\text{Principal component analysis of MD trajectories} $$\n",
    "   - Cross-validate with omics data:\n",
    "     $$ \\text{Co-expression with metabolic pathway genes (r ≥ 0.85)} $$\n",
    "     $$ \\text{Proteomic colocalization evidence} $$\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,    // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "prompt5 = \"\"\"\n",
    "You are a biochemical expert. Your task is to analyze a protein and determine which reaction(s) it is most likely to catalyze based on the provided data in JSON format.\n",
    "\n",
    "Guidline:\n",
    "1. **Literature Evidence Triangulation** ★最高优先级\n",
    "   a. UniProt/Sequence-based search:\n",
    "     $$ \\text{IF } \\exists \\text{ PMIDs } (P|E) \\rightarrow \\text{ReactID} \\text{ THEN confidence} += 0.4 $$\n",
    "   b. Reaction-centric search:\n",
    "     $$ \\text{Match(Equation|ChEBI|SMILES) → PMIDs → Enzyme } \\cap \\text{Input} $$\n",
    "   c. EC number validation:\n",
    "     $$ \\text{Validate EC}_{obs} = \\frac{\\text{PMID}_{\\text{EC-match}}}{\\text{PMID}_{\\text{EC-total}}} \\geq 0.6 $$\n",
    "\n",
    "2. **Structural Catalytic Competence**\n",
    "   - SMILES拓扑指纹匹配:\n",
    "     $$ \\text{Tanimoto} \\geq 0.85 \\text{ with Catalophore model} $$\n",
    "   - 电荷互补性验证:\n",
    "     $$ \\Delta E_{\\text{elec}} = \\sum \\frac{q_i q_j}{4\\pi\\epsilon r_{ij}} \\leq -15 \\text{ kJ/mol} $$\n",
    "\n",
    "3. **Quantum Transition State Analysis**\n",
    "   - 催化效率计算:\n",
    "     $$ \\ln(k_{\\text{cat}}/K_m) \\propto -\\Delta G^\\ddagger_{\\text{calc}} + \\text{文献修正项} $$\n",
    "   - 轨道相互作用验证:\n",
    "     $$ \\text{HOMO-LUMO gap匹配度} \\geq 82\\% $$\n",
    "\n",
    "4. **Evolutionary & Omics Validation**\n",
    "   - 系统发育可信度:\n",
    "     $$ P_{\\text{cat}} = \\text{Consurf} \\times \\text{EC}_{phylo} \\geq 7.5 $$\n",
    "   - 组学相关性:\n",
    "     $$ r_{\\text{coex}} \\geq 0.9 \\text{ AND } \\text{LocalScore} \\geq 0.8 $$\n",
    "\n",
    "**Important Notes:**\n",
    "- If a reaction is represented by a single dash (\"-\"), it means the protein is predicted not to catalyze that reaction. However, if all candidate reactions are marked with a dash, or if your analysis suggests that none of the reactions with valid IDs are likely catalyzed by the protein, then include the dash candidate reaction in your output to indicate that the protein is not catalytically active for any provided reaction.\n",
    "- Proteins may have multiple active sites. If applicable, select more than one reaction.\n",
    "- For each reaction, provide a brief explanation citing relevant sequence features (e.g., conserved domains, active site residues, structural motifs) and reaction details.\n",
    "\n",
    "Output your final answer in JSON format using the following template:\n",
    "{\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,         // 1 for the most likely, 2 for the next, etc.\n",
    "    \"confidence\": <number>,    // A value between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "}\n",
    "\n",
    "Input:\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "schema = {\n",
    "    \"type\": \"function\",   \n",
    "    \"name\": \"return_predictions\",\n",
    "    \"parameters\": {\n",
    "        \"type\": \"object\",\n",
    "        \"properties\": {\n",
    "            \"results\": {\n",
    "                \"type\": \"array\",\n",
    "                \"items\": {\n",
    "                    \"type\": \"object\",\n",
    "                    \"properties\": {\n",
    "                        \"reaction_id\":  {\"type\": \"string\"},\n",
    "                        \"selected\":     {\"type\": \"string\"},\n",
    "                        \"rank\":         {\"type\": [\"integer\", \"null\"]},\n",
    "                        \"confidence\":   {\"type\": \"number\"},\n",
    "                        \"reason\":       {\"type\": \"string\"}\n",
    "                    },\n",
    "                    \"required\": [\"reaction_id\", \"selected\", \"confidence\", \"reason\"]\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "        \"required\": [\"results\"]\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "system_prompt = \"\"\"\n",
    "You are a biochemical expert. Given a protein sequence and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein.\n",
    "You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
    "Key considerations:\n",
    "1. **Non-Catalytic Cases:**\n",
    "   - A reaction represented by a single dash (\"-\") indicates a predicted lack of catalytic activity.\n",
    "   - If all reactions are marked as \"-\", or none of the valid reactions are plausible, select the dash (\"-\") reaction to indicate non-catalytic function.\n",
    "\n",
    "2. **Multiple Reactions:**\n",
    "   - Proteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
    "\n",
    "3. **Ranking:**\n",
    "   - Assign a likelihood-based ranking to selected reactions (1 = most likely).\n",
    "\n",
    "4. **Justification:**\n",
    "   - Provide a brief explanation for each decision, referencing conserved domains, active site residues, structural motifs, or known catalytic mechanisms.\n",
    "\n",
    "5. **Confidence Score:**\n",
    "   - Assign a confidence score between 0 and 1 based on how well the protein sequence aligns with known catalytic features for each reaction.\n",
    "   - If no valid reaction is likely, assign the highest confidence to the dash (\"-\") reaction.\n",
    "\n",
    "Output format (strictly JSON):\n",
    "[\n",
    "    {\n",
    "        \"reaction_id\": \"xxx\",\n",
    "        \"selected\": \"yes\" or \"no\",\n",
    "        \"rank\": <integer>,     // Only for selected = \"yes\"\n",
    "        \"confidence\": <float>, // Between 0 and 1\n",
    "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "    },\n",
    "    ...\n",
    "]\n",
    "\n",
    "⚠️ Only use the key 'results' in the output. Do not include any extra explanation or metadata.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "markdown"
    }
   },
   "outputs": [],
   "source": [
    "system_prompt = \"\"\"\n",
    "You are a biochemical expert. Given a protein sequence and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein.\n",
    "You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
    "Key considerations:\n",
    "1. **Non-Catalytic Cases:**\n",
    "   - A reaction represented by a single dash (\"-\") indicates a predicted lack of catalytic activity.\n",
    "   - If all reactions are marked as \"-\", or none of the valid reactions are plausible, select the dash (\"-\") reaction to indicate non-catalytic function.\n",
    "\n",
    "2. **Multiple Reactions:**\n",
    "   - Proteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
    "\n",
    "3. **Ranking:**\n",
    "   - Assign a likelihood-based ranking to selected reactions (1 = most likely).\n",
    "\n",
    "4. **Justification:**\n",
    "   - Provide a brief explanation for each decision, referencing conserved domains, active site residues, structural motifs, or known catalytic mechanisms.\n",
    "\n",
    "5. **Confidence Score:**\n",
    "   - Assign a confidence score between 0 and 1 based on how well the protein sequence aligns with known catalytic features for each reaction.\n",
    "   - If no valid reaction is likely, assign the highest confidence to the dash (\"-\") reaction.\n",
    "\n",
    "Output format (strictly JSON):\n",
    "[\n",
    "    {\n",
    "        \"reaction_id\": \"xxx\",\n",
    "        \"selected\": \"yes\" or \"no\",\n",
    "        \"rank\": <integer>,     // Only for selected = \"yes\"\n",
    "        \"confidence\": <float>, // Between 0 and 1\n",
    "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "    },\n",
    "    ...\n",
    "]\n",
    "\n",
    "⚠️ Only use the key 'results' in the output. Do not include any extra explanation or metadata.\n",
    "\"\"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "markdown"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'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'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "markdown"
    }
   },
   "outputs": [],
   "source": [
    "! source ~/.zshrc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "markdown"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "APIKEY = os.environ.get(\"OPENROUTER_API_KEY\")\n",
    "cli5 = llmchat.Chat(name='openai/gpt-4.1', \n",
    "               url='https://openrouter.ai/api/v1', \n",
    "               api_key = APIKEY,\n",
    "            #    proxy='http://172.16.10.2:7897'\n",
    ")\n",
    "\n",
    "\n",
    "def make_make_user_query(uniprot_id, seq, rxn_id, rxn_bank):\n",
    "    if not rxn_id or cfg.SPLITER not in rxn_id:\n",
    "        return {}\n",
    "    rxns = rxn_id.split(cfg.SPLITER)\n",
    "\n",
    "    query_info = {\n",
    "        'protein information': {\n",
    "            'uniprot id': uniprot_id,\n",
    "            'protein amino acid sequence': seq\n",
    "        },\n",
    "        'reaction information': {\n",
    "            f'candidate reaction {i+1}': get_rxn_detail(rxn, rxn_bank)\n",
    "            for i, rxn in enumerate(rxns)\n",
    "        }\n",
    "    }\n",
    "\n",
    "    llm_query =  json.dumps(query_info, indent=4, ensure_ascii=False)\n",
    "\n",
    "    return llm_query\n",
    "\n",
    "def session_chat(chat_bot, ix, system_prompt, debug=True):\n",
    "    \n",
    "    query_user = make_make_user_query(uniprot_id=res_ecrecer.iloc[ix].uniprot_id, seq=res_ecrecer.iloc[ix].seq, rxn_id=res_ecrecer.iloc[ix].RXNRECer, rxn_bank=rxn)\n",
    "    if debug:\n",
    "        print(system_prompt)\n",
    "        print(query_user)\n",
    "   \n",
    "    response = chat_bot.chat(message=query_user, system_prompt=system_prompt, response_format={\"type\": \"json_object\"})   # 关键行，指定返回格式为json_object)\n",
    "    print(response)\n",
    "    records = json.loads(response.choices[0].message.content)\n",
    "    \n",
    "    json_results = records\n",
    "    df_results = pd.DataFrame(records['results']).fillna(float('inf'))\n",
    "    \n",
    "    \n",
    "    if debug:\n",
    "\n",
    "         print(response)\n",
    "         print(json_results)\n",
    "    \n",
    "    return df_results, json_results\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "You are a biochemical expert. Given a protein sequence and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein.\n",
      "You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
      "Key considerations:\n",
      "1. **Non-Catalytic Cases:**\n",
      "   - A reaction represented by a single dash (\"-\") indicates a predicted lack of catalytic activity.\n",
      "   - If all reactions are marked as \"-\", or none of the valid reactions are plausible, select the dash (\"-\") reaction to indicate non-catalytic function.\n",
      "\n",
      "2. **Multiple Reactions:**\n",
      "   - Proteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
      "\n",
      "3. **Ranking:**\n",
      "   - Assign a likelihood-based ranking to selected reactions (1 = most likely).\n",
      "\n",
      "4. **Justification:**\n",
      "   - Provide a brief explanation for each decision, referencing conserved domains, active site residues, structural motifs, or known catalytic mechanisms.\n",
      "\n",
      "5. **Confidence Score:**\n",
      "   - Assign a confidence score between 0 and 1 based on how well the protein sequence aligns with known catalytic features for each reaction.\n",
      "   - If no valid reaction is likely, assign the highest confidence to the dash (\"-\") reaction.\n",
      "\n",
      "Output format (strictly JSON):\n",
      "[\n",
      "    {\n",
      "        \"reaction_id\": \"xxx\",\n",
      "        \"selected\": \"yes\" or \"no\",\n",
      "        \"rank\": <integer>,     // Only for selected = \"yes\"\n",
      "        \"confidence\": <float>, // Between 0 and 1\n",
      "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
      "    },\n",
      "    ...\n",
      "]\n",
      "\n",
      "⚠️ Only use the key 'results' in the output. Do not include any extra explanation or metadata.\n",
      "\n",
      "{\n",
      "    \"protein information\": {\n",
      "        \"uniprot id\": \"Q7XQ85\",\n",
      "        \"protein amino acid sequence\": \"MAYQGIDLLSTKAAGDDHGENSSYFDGWKAYDTNPFDLRHNRGGVIQMGLAENQLSLDLIEEWSKNHPEASICTPEGVSQFKRIANFQDYHGLPEFRKAMAQFMGQVRGGKATFDPDRVVMSGGATGAQETLAFCLANPGEAFLVPTPYYPAFDRDCCWRSGIKLLPIECHSFNDFRLTKEALVSAYDGARRQGISVKGILITNPSNPLGTITDRDTLAMLATFATEHRVHLVCDEIYAGSVFATPEYVSIAEVIERDVPWCNRDLIHVVYSLSKDFGLPGFRVGIIYSYNDAVVAAARRMSSFGLVSSQTQYFLARMLSDEEFIGRFLQESKCRLVARHERFTSGLREVGIGCLRGNAGLFSWMDLRRMLREKTAEAELELWRVIVHQVKLNVSPGTSFHCREPGWFRVCHANMDDETMEVALGRIHDFVRQHQQRRVKAERWAANRQLRLSLPHHHHLSPAHLSSPLALLSPQSPMVRATS\"\n",
      "    },\n",
      "    \"reaction information\": {\n",
      "        \"candidate reaction 1\": {\n",
      "            \"reaction id\": \"RHEA:21744\",\n",
      "            \"reaction equation\": \"S-adenosyl-L-methionine = 1-aminocyclopropane-1-carboxylate + H(+) + S-methyl-5'-thioadenosine\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:59789 = CHEBI:58360 + CHEBI:15378 + CHEBI:17509\",\n",
      "            \"reaction equation in SMILES format\": \"C[S+](CC[C@H]([NH3+])C([O-])=O)C[C@H]1O[C@H]([C@H](O)[C@@H]1O)n1cnc2c(N)ncnc12>>[NH3+]C1(CC1)C([O-])=O.[H+].CSC[C@H]1O[C@H]([C@H](O)[C@@H]1O)n1cnc2c(N)ncnc12\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:4.4.1.14\"\n",
      "        },\n",
      "        \"candidate reaction 2\": {\n",
      "            \"reaction id\": \"-\",\n",
      "            \"reaction equation\": \"-\"\n",
      "        }\n",
      "    }\n",
      "}\n",
      "ChatCompletion(id='gen-1748579674-zhalainNX4yOmp8gClOj', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\\n  \"results\": [\\n    {\\n      \"reaction_id\": \"RHEA:21744\",\\n      \"selected\": \"yes\",\\n      \"rank\": 1,\\n      \"confidence\": 0.98,\\n      \"reason\": \"The UniProt ID Q7XQ85 corresponds to the protein 1-aminocyclopropane-1-carboxylate synthase (ACC synthase), which catalyzes the conversion of S-adenosyl-L-methionine to 1-aminocyclopropane-1-carboxylate, H(+), and S-methyl-5\\'-thioadenosine (EC 4.4.1.14). The sequence contains conserved motifs and active site residues characteristic of the ACC synthase family, including the pyridoxal phosphate (PLP)-binding domain, which is essential for this lyase activity. This matches the reaction RHEA:21744 with high confidence.\"\\n    },\\n    {\\n      \"reaction_id\": \"-\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.01,\\n      \"reason\": \"The protein sequence and UniProt annotation strongly support catalytic activity as ACC synthase. There is no evidence for a non-catalytic function.\"\\n    }\\n  ]\\n}', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning=None), native_finish_reason='stop')], created=1748579675, model='openai/gpt-4.1', object='chat.completion', service_tier=None, system_fingerprint='fp_b3f1157249', usage=CompletionUsage(completion_tokens=263, prompt_tokens=966, total_tokens=1229, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=0, rejected_prediction_tokens=None), prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0)), provider='OpenAI')\n",
      "ChatCompletion(id='gen-1748579674-zhalainNX4yOmp8gClOj', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\\n  \"results\": [\\n    {\\n      \"reaction_id\": \"RHEA:21744\",\\n      \"selected\": \"yes\",\\n      \"rank\": 1,\\n      \"confidence\": 0.98,\\n      \"reason\": \"The UniProt ID Q7XQ85 corresponds to the protein 1-aminocyclopropane-1-carboxylate synthase (ACC synthase), which catalyzes the conversion of S-adenosyl-L-methionine to 1-aminocyclopropane-1-carboxylate, H(+), and S-methyl-5\\'-thioadenosine (EC 4.4.1.14). The sequence contains conserved motifs and active site residues characteristic of the ACC synthase family, including the pyridoxal phosphate (PLP)-binding domain, which is essential for this lyase activity. This matches the reaction RHEA:21744 with high confidence.\"\\n    },\\n    {\\n      \"reaction_id\": \"-\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.01,\\n      \"reason\": \"The protein sequence and UniProt annotation strongly support catalytic activity as ACC synthase. There is no evidence for a non-catalytic function.\"\\n    }\\n  ]\\n}', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning=None), native_finish_reason='stop')], created=1748579675, model='openai/gpt-4.1', object='chat.completion', service_tier=None, system_fingerprint='fp_b3f1157249', usage=CompletionUsage(completion_tokens=263, prompt_tokens=966, total_tokens=1229, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=0, rejected_prediction_tokens=None), prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0)), provider='OpenAI')\n",
      "{'results': [{'reaction_id': 'RHEA:21744', 'selected': 'yes', 'rank': 1, 'confidence': 0.98, 'reason': \"The UniProt ID Q7XQ85 corresponds to the protein 1-aminocyclopropane-1-carboxylate synthase (ACC synthase), which catalyzes the conversion of S-adenosyl-L-methionine to 1-aminocyclopropane-1-carboxylate, H(+), and S-methyl-5'-thioadenosine (EC 4.4.1.14). The sequence contains conserved motifs and active site residues characteristic of the ACC synthase family, including the pyridoxal phosphate (PLP)-binding domain, which is essential for this lyase activity. This matches the reaction RHEA:21744 with high confidence.\"}, {'reaction_id': '-', 'selected': 'no', 'confidence': 0.01, 'reason': 'The protein sequence and UniProt annotation strongly support catalytic activity as ACC synthase. There is no evidence for a non-catalytic function.'}]}\n"
     ]
    }
   ],
   "source": [
    "res_df, res_json = session_chat(chat_bot=cli5, ix=12, system_prompt=system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'results': [{'reaction_id': 'RHEA:21744',\n",
       "   'selected': 'yes',\n",
       "   'rank': 1,\n",
       "   'confidence': 0.98,\n",
       "   'reason': \"The UniProt ID Q7XQ85 corresponds to the protein 1-aminocyclopropane-1-carboxylate synthase (ACC synthase), which catalyzes the conversion of S-adenosyl-L-methionine to 1-aminocyclopropane-1-carboxylate, H(+), and S-methyl-5'-thioadenosine (EC 4.4.1.14). The sequence contains conserved motifs and active site residues characteristic of the ACC synthase family, including the pyridoxal phosphate (PLP)-binding domain, which is essential for this lyase activity. This matches the reaction RHEA:21744 with high confidence.\"},\n",
       "  {'reaction_id': '-',\n",
       "   'selected': 'no',\n",
       "   'confidence': 0.01,\n",
       "   'reason': 'The protein sequence and UniProt annotation strongly support catalytic activity as ACC synthase. There is no evidence for a non-catalytic function.'}]}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "You are a biochemical expert. Given a protein sequence and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein.\n",
      "You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
      "Key considerations:\n",
      "1. **Non-Catalytic Cases:**\n",
      "   - A reaction represented by a single dash (\"-\") indicates a predicted lack of catalytic activity.\n",
      "   - If all reactions are marked as \"-\", or none of the valid reactions are plausible, select the dash (\"-\") reaction to indicate non-catalytic function.\n",
      "\n",
      "2. **Multiple Reactions:**\n",
      "   - Proteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
      "\n",
      "3. **Ranking:**\n",
      "   - Assign a likelihood-based ranking to selected reactions (1 = most likely).\n",
      "\n",
      "4. **Justification:**\n",
      "   - Provide a brief explanation for each decision, referencing conserved domains, active site residues, structural motifs, or known catalytic mechanisms.\n",
      "\n",
      "5. **Confidence Score:**\n",
      "   - Assign a confidence score between 0 and 1 based on how well the protein sequence aligns with known catalytic features for each reaction.\n",
      "   - If no valid reaction is likely, assign the highest confidence to the dash (\"-\") reaction.\n",
      "\n",
      "Output format (strictly JSON):\n",
      "[\n",
      "    {\n",
      "        \"reaction_id\": \"xxx\",\n",
      "        \"selected\": \"yes\" or \"no\",\n",
      "        \"rank\": <integer>,     // Only for selected = \"yes\"\n",
      "        \"confidence\": <float>, // Between 0 and 1\n",
      "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
      "    },\n",
      "    ...\n",
      "]\n",
      "\n",
      "⚠️ Only use the key 'results' in the output. Do not include any extra explanation or metadata.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'results': [{'reaction_id': 'RHEA:21744',\n",
       "   'selected': 'yes',\n",
       "   'rank': 1,\n",
       "   'confidence': 0.98,\n",
       "   'reason': \"The UniProt ID Q7XQ85 corresponds to the protein 1-aminocyclopropane-1-carboxylate synthase (ACC synthase), which catalyzes the conversion of S-adenosyl-L-methionine to 1-aminocyclopropane-1-carboxylate, H(+), and S-methyl-5'-thioadenosine (EC 4.4.1.14). The sequence contains conserved motifs and active site residues characteristic of the ACC synthase family, including the pyridoxal phosphate (PLP)-binding domain, which is essential for this lyase activity. This matches the reaction RHEA:21744 with high confidence.\"},\n",
       "  {'reaction_id': '-',\n",
       "   'selected': 'no',\n",
       "   'confidence': 0.01,\n",
       "   'reason': 'The protein sequence and UniProt annotation strongly support catalytic activity as ACC synthase. There is no evidence for a non-catalytic function.'}]}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_df, res_json = session_chat(chat_bot=cli5, ix=12, system_prompt=system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>selected</th>\n",
       "      <th>confidence</th>\n",
       "      <th>reason</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHEA:20049</td>\n",
       "      <td>no</td>\n",
       "      <td>0.15</td>\n",
       "      <td>The protein L0TBY6 (UniProt) is annotated as a...</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.95</td>\n",
       "      <td>The sequence lacks identifiable catalytic doma...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id selected  confidence  \\\n",
       "0  RHEA:20049       no        0.15   \n",
       "1           -      yes        0.95   \n",
       "\n",
       "                                              reason  rank  \n",
       "0  The protein L0TBY6 (UniProt) is annotated as a...   inf  \n",
       "1  The sequence lacks identifiable catalytic doma...   1.0  "
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session_chat(chat_bot=cli5, ix=17)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>selected</th>\n",
       "      <th>rank</th>\n",
       "      <th>confidence</th>\n",
       "      <th>reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHEA:52584</td>\n",
       "      <td>yes</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.98</td>\n",
       "      <td>The UniProt entry A7B3K3 corresponds to BaiA1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RHEA:52596</td>\n",
       "      <td>yes</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.97</td>\n",
       "      <td>This is another substrate-specific reaction fo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RHEA:52588</td>\n",
       "      <td>yes</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.96</td>\n",
       "      <td>This reaction is also catalyzed by 3α-hydroxys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RHEA:52592</td>\n",
       "      <td>yes</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.96</td>\n",
       "      <td>Deoxycholoyl-CoA is a known substrate for BaiA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RHEA:55380</td>\n",
       "      <td>yes</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.99</td>\n",
       "      <td>This is the generic reaction for 3α-hydroxyste...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RHEA:11728</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.10</td>\n",
       "      <td>Cyclopentanol dehydrogenase (EC 1.1.1.163) is ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RHEA:10044</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.10</td>\n",
       "      <td>Cyclohexanol dehydrogenase (EC 1.1.1.245) is n...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RHEA:17397</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.10</td>\n",
       "      <td>3-oxoacyl-[ACP] reductase (EC 1.1.1.100) is a ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RHEA:47496</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.20</td>\n",
       "      <td>This reaction is catalyzed by 3α-hydroxysteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.01</td>\n",
       "      <td>The protein is clearly catalytic, with strong ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id selected  rank  confidence  \\\n",
       "0  RHEA:52584      yes   1.0        0.98   \n",
       "1  RHEA:52596      yes   2.0        0.97   \n",
       "2  RHEA:52588      yes   3.0        0.96   \n",
       "3  RHEA:52592      yes   4.0        0.96   \n",
       "4  RHEA:55380      yes   5.0        0.99   \n",
       "5  RHEA:11728       no   inf        0.10   \n",
       "6  RHEA:10044       no   inf        0.10   \n",
       "7  RHEA:17397       no   inf        0.10   \n",
       "8  RHEA:47496       no   inf        0.20   \n",
       "9           -       no   inf        0.01   \n",
       "\n",
       "                                              reason  \n",
       "0  The UniProt entry A7B3K3 corresponds to BaiA1,...  \n",
       "1  This is another substrate-specific reaction fo...  \n",
       "2  This reaction is also catalyzed by 3α-hydroxys...  \n",
       "3  Deoxycholoyl-CoA is a known substrate for BaiA...  \n",
       "4  This is the generic reaction for 3α-hydroxyste...  \n",
       "5  Cyclopentanol dehydrogenase (EC 1.1.1.163) is ...  \n",
       "6  Cyclohexanol dehydrogenase (EC 1.1.1.245) is n...  \n",
       "7  3-oxoacyl-[ACP] reductase (EC 1.1.1.100) is a ...  \n",
       "8  This reaction is catalyzed by 3α-hydroxysteroi...  \n",
       "9  The protein is clearly catalytic, with strong ...  "
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session_chat(chat_bot=cli5, ix=29)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>selected</th>\n",
       "      <th>confidence</th>\n",
       "      <th>reason</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHEA:55056</td>\n",
       "      <td>no</td>\n",
       "      <td>0.15</td>\n",
       "      <td>This reaction is catalyzed by EC 1.14.14.49, a...</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.95</td>\n",
       "      <td>The protein sequence lacks conserved domains, ...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RHEA:17149</td>\n",
       "      <td>no</td>\n",
       "      <td>0.10</td>\n",
       "      <td>This is a generic cytochrome P450 monooxygenas...</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RHEA:43728</td>\n",
       "      <td>no</td>\n",
       "      <td>0.10</td>\n",
       "      <td>This reaction is catalyzed by EC 1.14.14.127, ...</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id selected  confidence  \\\n",
       "0  RHEA:55056       no        0.15   \n",
       "1           -      yes        0.95   \n",
       "2  RHEA:17149       no        0.10   \n",
       "3  RHEA:43728       no        0.10   \n",
       "\n",
       "                                              reason  rank  \n",
       "0  This reaction is catalyzed by EC 1.14.14.49, a...   inf  \n",
       "1  The protein sequence lacks conserved domains, ...   1.0  \n",
       "2  This is a generic cytochrome P450 monooxygenas...   inf  \n",
       "3  This reaction is catalyzed by EC 1.14.14.127, ...   inf  "
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session_chat(chat_bot=cli5, ix=94)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"protein information\": {\n",
      "        \"uniprot id\": \"L0TBY6\",\n",
      "        \"protein amino acid sequence\": \"MTHREELLPPMKWDAWGDPAAAKPLSDGVRSLLKQVVGLADSEQPELDPAQVQLRPSALSGADHDALARIVGTEYFRTADRDRLLHAGGKSTPDLLRRKDTGVQDAPDAVLLPGGPNGGGRRRRHLALLLRPRHCRGPVWWRHQRRWWA\"\n",
      "    },\n",
      "    \"reaction information\": {\n",
      "        \"candidate reaction 1\": {\n",
      "            \"reaction id\": \"-\",\n",
      "            \"reaction equation\": \"-\"\n",
      "        },\n",
      "        \"candidate reaction 2\": {\n",
      "            \"reaction id\": \"RHEA:20049\",\n",
      "            \"reaction equation\": \"1-(5-phospho-beta-D-ribosyl)-5'-AMP + H2O = 1-(5-phospho-beta-D-ribosyl)-5-[(5-phospho-beta-D-ribosylamino)methylideneamino]imidazole-4-carboxamide\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:59457 + CHEBI:15377 = CHEBI:58435\",\n",
      "            \"reaction equation in SMILES format\": \"O[C@H]1[C@@H](O)[C@@H](O[C@@H]1COP([O-])([O-])=O)n1cnc2c1ncn([C@@H]1O[C@H](COP([O-])([O-])=O)[C@@H](O)[C@H]1O)c2=N.[H]O[H]>>NC(=O)c1ncn([C@@H]2O[C@H](COP([O-])([O-])=O)[C@@H](O)[C@H]2O)c1\\\\\\\\N=C\\\\\\\\N[C@@H]1O[C@H](COP([O-])([O-])=O)[C@@H](O)[C@H]1O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:3.5.4.19\"\n",
      "        }\n",
      "    }\n",
      "}\n",
      "ChatCompletion(id='gen-1747293837-pWS5nHiTtgjMLgoJt3Sn', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\\n  \"results\": [\\n    {\\n      \"reaction_id\": \"-\",\\n      \"selected\": \"yes\",\\n      \"rank\": 1,\\n      \"confidence\": 0.95,\\n      \"reason\": \"The protein sequence (UniProt L0TBY6) does not show significant homology to known enzymes catalyzing the candidate reaction (EC 3.5.4.19, phosphoribosylformylglycinamidine cyclo-ligase). No conserved domains, active site motifs, or sequence features characteristic of this enzyme class are present. Therefore, the most likely scenario is a lack of catalytic activity.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:20049\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.10,\\n      \"reason\": \"No evidence of conserved domains or sequence motifs associated with EC 3.5.4.19 (phosphoribosylformylglycinamidine cyclo-ligase) is found in the provided sequence. The protein does not align with known catalytic features for this reaction.\"\\n    }\\n  ]\\n}', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning=None), native_finish_reason='stop')], created=1747293837, model='openai/gpt-4.1', object='chat.completion', service_tier=None, system_fingerprint='fp_b38e740b47', usage=CompletionUsage(completion_tokens=232, prompt_tokens=852, total_tokens=1084, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=0, rejected_prediction_tokens=None), prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0)), provider='OpenAI')\n",
      "{'results': [{'reaction_id': '-', 'selected': 'yes', 'rank': 1, 'confidence': 0.95, 'reason': 'The protein sequence (UniProt L0TBY6) does not show significant homology to known enzymes catalyzing the candidate reaction (EC 3.5.4.19, phosphoribosylformylglycinamidine cyclo-ligase). No conserved domains, active site motifs, or sequence features characteristic of this enzyme class are present. Therefore, the most likely scenario is a lack of catalytic activity.'}, {'reaction_id': 'RHEA:20049', 'selected': 'no', 'confidence': 0.1, 'reason': 'No evidence of conserved domains or sequence motifs associated with EC 3.5.4.19 (phosphoribosylformylglycinamidine cyclo-ligase) is found in the provided sequence. The protein does not align with known catalytic features for this reaction.'}]}\n"
     ]
    }
   ],
   "source": [
    "res_df_17, res_json_17 = session_chat(chat_bot=cli5, ix=17, system_prompt=system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'results': [{'reaction_id': '-',\n",
       "   'selected': 'yes',\n",
       "   'rank': 1,\n",
       "   'confidence': 0.95,\n",
       "   'reason': 'The protein sequence (UniProt L0TBY6) does not show significant homology to known enzymes catalyzing the candidate reaction (EC 3.5.4.19, phosphoribosylformylglycinamidine cyclo-ligase). No conserved domains, active site motifs, or sequence features characteristic of this enzyme class are present. Therefore, the most likely scenario is a lack of catalytic activity.'},\n",
       "  {'reaction_id': 'RHEA:20049',\n",
       "   'selected': 'no',\n",
       "   'confidence': 0.1,\n",
       "   'reason': 'No evidence of conserved domains or sequence motifs associated with EC 3.5.4.19 (phosphoribosylformylglycinamidine cyclo-ligase) is found in the provided sequence. The protein does not align with known catalytic features for this reaction.'}]}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_json_17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "You are a biochemical expert. Given a protein sequence and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein.\n",
      "You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
      "Key considerations:\n",
      "1. **Non-Catalytic Cases:**\n",
      "   - A reaction represented by a single dash (\"-\") indicates a predicted lack of catalytic activity.\n",
      "   - If all reactions are marked as \"-\", or none of the valid reactions are plausible, select the dash (\"-\") reaction to indicate non-catalytic function.\n",
      "\n",
      "2. **Multiple Reactions:**\n",
      "   - Proteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
      "\n",
      "3. **Ranking:**\n",
      "   - Assign a likelihood-based ranking to selected reactions (1 = most likely).\n",
      "\n",
      "4. **Justification:**\n",
      "   - Provide a brief explanation for each decision, referencing conserved domains, active site residues, structural motifs, or known catalytic mechanisms.\n",
      "\n",
      "5. **Confidence Score:**\n",
      "   - Assign a confidence score between 0 and 1 based on how well the protein sequence aligns with known catalytic features for each reaction.\n",
      "   - If no valid reaction is likely, assign the highest confidence to the dash (\"-\") reaction.\n",
      "\n",
      "Output format (strictly JSON):\n",
      "[\n",
      "    {\n",
      "        \"reaction_id\": \"xxx\",\n",
      "        \"selected\": \"yes\" or \"no\",\n",
      "        \"rank\": <integer>,     // Only for selected = \"yes\"\n",
      "        \"confidence\": <float>, // Between 0 and 1\n",
      "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
      "    },\n",
      "    ...\n",
      "]\n",
      "\n",
      "⚠️ Only use the key 'results' in the output. Do not include any extra explanation or metadata.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"protein information\": {\n",
      "        \"uniprot id\": \"Q88FI7\",\n",
      "        \"protein amino acid sequence\": \"MNQESISQSIAIVHPITLSHGRNAEVWDTDGKRYIDFVGGIGVLNLGHCNPAVVEAIQAQATRLTHYAFNAAPHGPYLALMEQLSQFVPVSYPLAGMLTNSGAEAAENALKVARGATGKRAIIAFDGGFHGRTLATLNLNGKVAPYKQRVGELPGPVYHLPYPSADTGVTCEQALKAMDRLFSVELAVEDVAAFIFEPVQGEGGFLALDPAFAQALRRFCDERGILIIIDEIQSGFGRTGQRFAFPRLGIEPDLLLLAKSIAGGMPLGAVVGRKELMAALPKGGLGGTYSGNPISCAAALASLAQMTDENLATWGERQEQAIVSRYERWKASGLSPYIGRLTGVGAMRGIEFANADGSPAPAQLAKVMEAARARGLLLMPSGKARHIIRLLAPLTIEAEVLEEGLDILEQCLAELN\"\n",
      "    },\n",
      "    \"reaction information\": {\n",
      "        \"candidate reaction 1\": {\n",
      "            \"reaction id\": \"RHEA:18049\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + N(2)-acetyl-L-ornithine = L-glutamate + N-acetyl-L-glutamate 5-semialdehyde\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:57805 = CHEBI:29985 + CHEBI:29123\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.CC(=O)N[C@@H](CCC[NH3+])C([O-])=O>>[NH3+][C@@H](CCC([O-])=O)C([O-])=O.CC(=O)N[C@@H](CCC=O)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.11\"\n",
      "        },\n",
      "        \"candidate reaction 2\": {\n",
      "            \"reaction id\": \"RHEA:23352\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + 4-aminobutanoate = L-glutamate + succinate semialdehyde\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:59888 = CHEBI:29985 + CHEBI:57706\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+]CCCC([O-])=O>>[NH3+][C@@H](CCC([O-])=O)C([O-])=O.[O-]C(=O)CCC=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.19\"\n",
      "        },\n",
      "        \"candidate reaction 3\": {\n",
      "            \"reaction id\": \"RHEA:10212\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + 5-aminopentanoate = 5-oxopentanoate + L-glutamate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:356010 = CHEBI:16120 + CHEBI:29985\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+]CCCCC([O-])=O>>[H]C(=O)CCCC([O-])=O.[NH3+][C@@H](CCC([O-])=O)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.48\"\n",
      "        },\n",
      "        \"candidate reaction 4\": {\n",
      "            \"reaction id\": \"RHEA:12268\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + putrescine = 1-pyrroline + H2O + L-glutamate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:326268 = CHEBI:36781 + CHEBI:15377 + CHEBI:29985\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+]CCCC[NH3+]>>C1CC=[NH+]C1.[H]O[H].[NH3+][C@@H](CCC([O-])=O)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.82\"\n",
      "        }\n",
      "    }\n",
      "}\n",
      "ChatCompletion(id='gen-1747297671-lnhkdntZM00TE0J7Sqsn', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\\n  \"results\": [\\n    {\\n      \"reaction_id\": \"RHEA:18049\",\\n      \"selected\": \"yes\",\\n      \"rank\": 1,\\n      \"confidence\": 0.98,\\n      \"reason\": \"The UniProt ID Q88FI7 corresponds to N-acetylornithine aminotransferase (EC 2.6.1.11), which catalyzes the transamination of N(2)-acetyl-L-ornithine and 2-oxoglutarate to form N-acetyl-L-glutamate 5-semialdehyde and L-glutamate. The protein sequence contains conserved motifs and active site residues characteristic of aminotransferases, including the PLP-binding lysine. This matches the reaction RHEA:18049 with high confidence.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:23352\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.15,\\n      \"reason\": \"This reaction is catalyzed by 4-aminobutyrate aminotransferase (EC 2.6.1.19). While both are aminotransferases, sequence and domain analysis show Q88FI7 is specific for N-acetylornithine, not 4-aminobutanoate. No evidence for this substrate specificity in Q88FI7.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:10212\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.10,\\n      \"reason\": \"This reaction is catalyzed by 5-aminovalerate transaminase (EC 2.6.1.48). Q88FI7 lacks the sequence features and substrate specificity for 5-aminopentanoate, as supported by UniProt and BRENDA annotations.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:12268\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.10,\\n      \"reason\": \"This reaction is catalyzed by putrescine transaminase (EC 2.6.1.82). Q88FI7 does not contain the necessary active site adaptations for putrescine binding and catalysis, and is not annotated for this activity in major databases.\"\\n    }\\n  ]\\n}', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning=None), native_finish_reason='stop')], created=1747297671, model='openai/gpt-4.1', object='chat.completion', service_tier=None, system_fingerprint='fp_b38e740b47', usage=CompletionUsage(completion_tokens=484, prompt_tokens=1448, total_tokens=1932, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=0, rejected_prediction_tokens=None), prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0)), provider='OpenAI')\n",
      "{'results': [{'reaction_id': 'RHEA:18049', 'selected': 'yes', 'rank': 1, 'confidence': 0.98, 'reason': 'The UniProt ID Q88FI7 corresponds to N-acetylornithine aminotransferase (EC 2.6.1.11), which catalyzes the transamination of N(2)-acetyl-L-ornithine and 2-oxoglutarate to form N-acetyl-L-glutamate 5-semialdehyde and L-glutamate. The protein sequence contains conserved motifs and active site residues characteristic of aminotransferases, including the PLP-binding lysine. This matches the reaction RHEA:18049 with high confidence.'}, {'reaction_id': 'RHEA:23352', 'selected': 'no', 'confidence': 0.15, 'reason': 'This reaction is catalyzed by 4-aminobutyrate aminotransferase (EC 2.6.1.19). While both are aminotransferases, sequence and domain analysis show Q88FI7 is specific for N-acetylornithine, not 4-aminobutanoate. No evidence for this substrate specificity in Q88FI7.'}, {'reaction_id': 'RHEA:10212', 'selected': 'no', 'confidence': 0.1, 'reason': 'This reaction is catalyzed by 5-aminovalerate transaminase (EC 2.6.1.48). Q88FI7 lacks the sequence features and substrate specificity for 5-aminopentanoate, as supported by UniProt and BRENDA annotations.'}, {'reaction_id': 'RHEA:12268', 'selected': 'no', 'confidence': 0.1, 'reason': 'This reaction is catalyzed by putrescine transaminase (EC 2.6.1.82). Q88FI7 does not contain the necessary active site adaptations for putrescine binding and catalysis, and is not annotated for this activity in major databases.'}]}\n"
     ]
    }
   ],
   "source": [
    "res_df_20, res_json_20 = session_chat(chat_bot=cli5, ix=20, system_prompt=system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>selected</th>\n",
       "      <th>rank</th>\n",
       "      <th>confidence</th>\n",
       "      <th>reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHEA:18049</td>\n",
       "      <td>yes</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.98</td>\n",
       "      <td>The UniProt ID Q88FI7 corresponds to N-acetylo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RHEA:23352</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.15</td>\n",
       "      <td>This reaction is catalyzed by 4-aminobutyrate ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RHEA:10212</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.10</td>\n",
       "      <td>This reaction is catalyzed by 5-aminovalerate ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RHEA:12268</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.10</td>\n",
       "      <td>This reaction is catalyzed by putrescine trans...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id selected  rank  confidence  \\\n",
       "0  RHEA:18049      yes   1.0        0.98   \n",
       "1  RHEA:23352       no   inf        0.15   \n",
       "2  RHEA:10212       no   inf        0.10   \n",
       "3  RHEA:12268       no   inf        0.10   \n",
       "\n",
       "                                              reason  \n",
       "0  The UniProt ID Q88FI7 corresponds to N-acetylo...  \n",
       "1  This reaction is catalyzed by 4-aminobutyrate ...  \n",
       "2  This reaction is catalyzed by 5-aminovalerate ...  \n",
       "3  This reaction is catalyzed by putrescine trans...  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df_20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RHEA:12601;RHEA:10212"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>C0HK72</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "      <td>RGICLEPKVVGPCKARIRRFYYDSETGKCTPFIYGGCGGNGNNFET...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A7AZH2</td>\n",
       "      <td>RHEA:55380;RHEA:52588;RHEA:52584;RHEA:52592;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:17397': 0.7777, 'RHEA:1883...</td>\n",
       "      <td>MNFGGFIMGRFDEKIMLVTGATSGIGRAVAIRAAKEGATVVAVGRN...</td>\n",
       "      <td>RHEA:47520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>P0DX23</td>\n",
       "      <td>-;RHEA:14709;RHEA:32187;RHEA:54796</td>\n",
       "      <td>{'-': 0.998382, 'RHEA:14709': 0.7777, 'RHEA:32...</td>\n",
       "      <td>MKVLVTGATSGLGRNAVEYLRNKGISVRATGRNEAMGKLLSKMGAE...</td>\n",
       "      <td>RHEA:14981;RHEA:14929;RHEA:24616;RHEA:24612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>P0DX24</td>\n",
       "      <td>RHEA:35459;RHEA:43936;RHEA:43932;RHEA:43928;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:14709': 0.9953950047492981...</td>\n",
       "      <td>MGDPTLRTDLGRVLVTGGSGFVGANLVTTLLERGHEVRSFDRVPSP...</td>\n",
       "      <td>RHEA:14929;RHEA:14981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>A0A7H0DNE2</td>\n",
       "      <td>RHEA:14709;RHEA:24076;RHEA:32187</td>\n",
       "      <td>{'RHEA:14709': 0.9999309778213501, 'RHEA:24076...</td>\n",
       "      <td>MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...</td>\n",
       "      <td>RHEA:24076;RHEA:14709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A0A7H0DNE2</td>\n",
       "      <td>RHEA:14709;RHEA:24076;RHEA:43932;RHEA:32187</td>\n",
       "      <td>{'RHEA:14709': 0.9999309778213501, 'RHEA:24076...</td>\n",
       "      <td>MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...</td>\n",
       "      <td>RHEA:24076;RHEA:14709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>C0HK73</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>GSICLEPKVVGPCKAGIRRFYFDSETGKCTLFLYGGCKGNGNNFET...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>F2XF94</td>\n",
       "      <td>RHEA:32539;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:32539': 0.9999420046806335}</td>\n",
       "      <td>MSVISIVPLVSKPCLYKSFISSTHEPKALRRPISTVGLCRRAKSVT...</td>\n",
       "      <td>RHEA:32539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>R9QMW4</td>\n",
       "      <td>RHEA:25488;RHEA:25496;RHEA:32539;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:25488': 0.9992690086364746...</td>\n",
       "      <td>MSLISAVPLASSCVSKSLISSVREHKALRRAIATLQMSRPGKSVAA...</td>\n",
       "      <td>RHEA:32539;RHEA:25500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>R9QMW8</td>\n",
       "      <td>RHEA:25488;RHEA:25496;RHEA:32539;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:25488': 0.9990990161895752...</td>\n",
       "      <td>MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...</td>\n",
       "      <td>RHEA:32539;RHEA:25500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>R9QMR6</td>\n",
       "      <td>RHEA:25488;RHEA:25496;RHEA:32539;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:25488': 0.9990620017051697...</td>\n",
       "      <td>MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...</td>\n",
       "      <td>RHEA:32539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>C0HK74</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>GSICLEPKVVGPCTAYFPRFYFDSETGKCTPFIYGGCEGNGNNFET...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>A0A3G2FQK2</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>GFYFRSIQGFYFKRIQGNICSEPKKVGRCRGSFPRFYFDSETGKCT...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Q4W8C9</td>\n",
       "      <td>RHEA:14549;RHEA:10432;-;RHEA:15649;RHEA:26510</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:10432': 0.96327, 'RHEA:145...</td>\n",
       "      <td>MSASPRLGFVQCISPAGLHRMAYHEWGDPANPRVLVCAHGLTRTGR...</td>\n",
       "      <td>RHEA:10172;RHEA:11248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>P0DQQ1</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>KRCYRTPDLKSQTCPPGEDLCYTKKWCADWCTSRGKVIELGCVATC...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>P0DQQ2</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>IRCFITPDVTSQICADGHVCYTKTWCDAWCTSRGKRVDLGCAATCP...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>C0HM09</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>IRCFITPDITSKDCPNGHVCYTKTWCDGFCSIRGERVDLDGCAATC...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>P0DSM9</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>LTCLICPEKYCNKVHTCLNGEKIYDQRKYIRGCADTCPVRKPREIV...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>E2ITZ3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MKTLLLTLLVVTIVCLDLGYTLECHNQQSSQTPTTTGCSGGETNCY...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>C0HLK3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>MICYNQQSSEPPTTKTCSEGQCYKKTWSDHRGTIIERGCACPNVKP...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>C0HM08</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>LECHNQQSSQTPTTTDCSGGETNCYKKWWSDHRGTIIERGCGCPTV...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>A0A7T7DMY7</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "      <td>MKTLLLTLVVVTIVCLDLGYTLKCNKLVPLFYKTCPAGKNLCYKMY...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>P0DSN1</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999989}</td>\n",
       "      <td>LKCNRLIPPFWKTCPEGKNLCYKMTMRLAPKVPVKRGCIDVCPKSS...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>P0DQH3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>LKCKNKLVPFLSKTCPEGKNLCYKMTLKKVTPKIKRG</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>P0DQH4</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999994}</td>\n",
       "      <td>LKCKNKVVPFLKCKNKVVPFLCYKMTLKKVTPKIKRG</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    uniprot_id                                           RXNRECer  \\\n",
       "30      C0HK72                                                  -   \n",
       "31      A7AZH2  RHEA:55380;RHEA:52588;RHEA:52584;RHEA:52592;RH...   \n",
       "32      P0DX23                 -;RHEA:14709;RHEA:32187;RHEA:54796   \n",
       "33      P0DX24  RHEA:35459;RHEA:43936;RHEA:43932;RHEA:43928;RH...   \n",
       "34  A0A7H0DNE2                   RHEA:14709;RHEA:24076;RHEA:32187   \n",
       "35  A0A7H0DNE2        RHEA:14709;RHEA:24076;RHEA:43932;RHEA:32187   \n",
       "36      C0HK73                                                  -   \n",
       "37      F2XF94                                       RHEA:32539;-   \n",
       "38      R9QMW4                 RHEA:25488;RHEA:25496;RHEA:32539;-   \n",
       "39      R9QMW8                 RHEA:25488;RHEA:25496;RHEA:32539;-   \n",
       "40      R9QMR6                 RHEA:25488;RHEA:25496;RHEA:32539;-   \n",
       "41      C0HK74                                                  -   \n",
       "42  A0A3G2FQK2                                                  -   \n",
       "43      Q4W8C9      RHEA:14549;RHEA:10432;-;RHEA:15649;RHEA:26510   \n",
       "44      P0DQQ1                                                  -   \n",
       "45      P0DQQ2                                                  -   \n",
       "46      C0HM09                                                  -   \n",
       "47      P0DSM9                                                  -   \n",
       "48      E2ITZ3                                                  -   \n",
       "49      C0HLK3                                                  -   \n",
       "50      C0HM08                                                  -   \n",
       "51  A0A7T7DMY7                                                  -   \n",
       "52      P0DSN1                                                  -   \n",
       "53      P0DQH3                                                  -   \n",
       "54      P0DQH4                                                  -   \n",
       "\n",
       "                                   RXNRECer_with_prob  \\\n",
       "30                                         {'-': 1.0}   \n",
       "31  {'-': 0.7777, 'RHEA:17397': 0.7777, 'RHEA:1883...   \n",
       "32  {'-': 0.998382, 'RHEA:14709': 0.7777, 'RHEA:32...   \n",
       "33  {'-': 0.7777, 'RHEA:14709': 0.9953950047492981...   \n",
       "34  {'RHEA:14709': 0.9999309778213501, 'RHEA:24076...   \n",
       "35  {'RHEA:14709': 0.9999309778213501, 'RHEA:24076...   \n",
       "36                                    {'-': 0.999998}   \n",
       "37    {'-': 0.7777, 'RHEA:32539': 0.9999420046806335}   \n",
       "38  {'-': 0.7777, 'RHEA:25488': 0.9992690086364746...   \n",
       "39  {'-': 0.7777, 'RHEA:25488': 0.9990990161895752...   \n",
       "40  {'-': 0.7777, 'RHEA:25488': 0.9990620017051697...   \n",
       "41                                    {'-': 0.999998}   \n",
       "42                                    {'-': 0.999998}   \n",
       "43  {'-': 0.7777, 'RHEA:10432': 0.96327, 'RHEA:145...   \n",
       "44                                    {'-': 0.999998}   \n",
       "45                                    {'-': 0.999997}   \n",
       "46                                    {'-': 0.999996}   \n",
       "47                                    {'-': 0.999997}   \n",
       "48                                    {'-': 0.999998}   \n",
       "49                                    {'-': 0.999996}   \n",
       "50                                    {'-': 0.999996}   \n",
       "51                                         {'-': 1.0}   \n",
       "52                                    {'-': 0.999989}   \n",
       "53                                    {'-': 0.999997}   \n",
       "54                                    {'-': 0.999994}   \n",
       "\n",
       "                                                  seq  \\\n",
       "30  RGICLEPKVVGPCKARIRRFYYDSETGKCTPFIYGGCGGNGNNFET...   \n",
       "31  MNFGGFIMGRFDEKIMLVTGATSGIGRAVAIRAAKEGATVVAVGRN...   \n",
       "32  MKVLVTGATSGLGRNAVEYLRNKGISVRATGRNEAMGKLLSKMGAE...   \n",
       "33  MGDPTLRTDLGRVLVTGGSGFVGANLVTTLLERGHEVRSFDRVPSP...   \n",
       "34  MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...   \n",
       "35  MAVYAVTGGAGFLGRYIVKLLISADDVQEIRVIDIVEDPQPITSKV...   \n",
       "36  GSICLEPKVVGPCKAGIRRFYFDSETGKCTLFLYGGCKGNGNNFET...   \n",
       "37  MSVISIVPLVSKPCLYKSFISSTHEPKALRRPISTVGLCRRAKSVT...   \n",
       "38  MSLISAVPLASSCVSKSLISSVREHKALRRAIATLQMSRPGKSVAA...   \n",
       "39  MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...   \n",
       "40  MSLISAVPLASSCVSKSLISSVREHTALRRAIATLQMSRRGKSVAA...   \n",
       "41  GSICLEPKVVGPCTAYFPRFYFDSETGKCTPFIYGGCEGNGNNFET...   \n",
       "42  GFYFRSIQGFYFKRIQGNICSEPKKVGRCRGSFPRFYFDSETGKCT...   \n",
       "43  MSASPRLGFVQCISPAGLHRMAYHEWGDPANPRVLVCAHGLTRTGR...   \n",
       "44  KRCYRTPDLKSQTCPPGEDLCYTKKWCADWCTSRGKVIELGCVATC...   \n",
       "45  IRCFITPDVTSQICADGHVCYTKTWCDAWCTSRGKRVDLGCAATCP...   \n",
       "46  IRCFITPDITSKDCPNGHVCYTKTWCDGFCSIRGERVDLDGCAATC...   \n",
       "47  LTCLICPEKYCNKVHTCLNGEKIYDQRKYIRGCADTCPVRKPREIV...   \n",
       "48  MKTLLLTLLVVTIVCLDLGYTLECHNQQSSQTPTTTGCSGGETNCY...   \n",
       "49  MICYNQQSSEPPTTKTCSEGQCYKKTWSDHRGTIIERGCACPNVKP...   \n",
       "50  LECHNQQSSQTPTTTDCSGGETNCYKKWWSDHRGTIIERGCGCPTV...   \n",
       "51  MKTLLLTLVVVTIVCLDLGYTLKCNKLVPLFYKTCPAGKNLCYKMY...   \n",
       "52  LKCNRLIPPFWKTCPEGKNLCYKMTMRLAPKVPVKRGCIDVCPKSS...   \n",
       "53              LKCKNKLVPFLSKTCPEGKNLCYKMTLKKVTPKIKRG   \n",
       "54              LKCKNKVVPFLKCKNKVVPFLCYKMTLKKVTPKIKRG   \n",
       "\n",
       "                                rxn_groundtruth  \n",
       "30                                            -  \n",
       "31                                   RHEA:47520  \n",
       "32  RHEA:14981;RHEA:14929;RHEA:24616;RHEA:24612  \n",
       "33                        RHEA:14929;RHEA:14981  \n",
       "34                        RHEA:24076;RHEA:14709  \n",
       "35                        RHEA:24076;RHEA:14709  \n",
       "36                                            -  \n",
       "37                                   RHEA:32539  \n",
       "38                        RHEA:32539;RHEA:25500  \n",
       "39                        RHEA:32539;RHEA:25500  \n",
       "40                                   RHEA:32539  \n",
       "41                                            -  \n",
       "42                                            -  \n",
       "43                        RHEA:10172;RHEA:11248  \n",
       "44                                            -  \n",
       "45                                            -  \n",
       "46                                            -  \n",
       "47                                            -  \n",
       "48                                            -  \n",
       "49                                            -  \n",
       "50                                            -  \n",
       "51                                            -  \n",
       "52                                            -  \n",
       "53                                            -  \n",
       "54                                            -  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer.iloc[30:55]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>U5NH37</td>\n",
       "      <td>RHEA:39895;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:39895': 0.9999889731407166}</td>\n",
       "      <td>MGSQETNLPPHVLIFPLPIQGHVNSMLRLAELLCLAELDITFIVSE...</td>\n",
       "      <td>RHEA:39895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>F1P963</td>\n",
       "      <td>RHEA:31583;RHEA:67392;RHEA:67604;RHEA:65396;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:31575': 0.9983620047569275...</td>\n",
       "      <td>MGTSRLYTLVLVLQPERVLLGMKKRGFGAGRWNGFGGKVQEGETIE...</td>\n",
       "      <td>RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Q7ZWC3</td>\n",
       "      <td>RHEA:31583;RHEA:67392;RHEA:67604;RHEA:65396;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:31575': 0.9945039749145508...</td>\n",
       "      <td>MFTSKLLTLVLVVQPGRVLLGMKKRGFGAGKWNGFGGKVQTGETIE...</td>\n",
       "      <td>RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>D3ZJ25</td>\n",
       "      <td>RHEA:75347;RHEA:70855;RHEA:66324;RHEA:70891;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:34923': 0.9976549744606018...</td>\n",
       "      <td>MAVDPPKADPKGVVAVDPTANCGSGLKSREDQGAKAGGCCSSRDQV...</td>\n",
       "      <td>RHEA:70891;RHEA:70875;RHEA:70863;RHEA:70887;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>P0DUR6</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MPHPMVAPALRRRRRSSFNPHVAAVMTGSILAGAMLSLSFIAVPVL...</td>\n",
       "      <td>RHEA:64268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>P0DUR5</td>\n",
       "      <td>RHEA:64264;-;RHEA:62680</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:62680': 0.526458, 'RHEA:64...</td>\n",
       "      <td>MSLPIPPRPADNSTSHSLCLTICGYRRPGMSEFDYRQHMTQVSAPM...</td>\n",
       "      <td>RHEA:64264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>A0A1L9WLD9</td>\n",
       "      <td>RHEA:64232;RHEA:64512;-;RHEA:12020</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:12020': 0.199639, 'RHEA:64...</td>\n",
       "      <td>MVTYTPPAEAQDLPTKLKVLYFSNEFPTDDLSTLLRRLHSHSKHSS...</td>\n",
       "      <td>RHEA:64232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>A0A1L9WLF1</td>\n",
       "      <td>RHEA:64236;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:64236': 0.9974650144577026}</td>\n",
       "      <td>MGEGGYRQINKTLNVCAFDDYLVTQQSRLPKLLDIEQLSPRVLRVL...</td>\n",
       "      <td>RHEA:64236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>A0A1L9WLH9</td>\n",
       "      <td>RHEA:64292;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:64292': 0.99652099609375}</td>\n",
       "      <td>MTVSSAPYCLAGQVALVTGSGRGIGAAIAVELGRLGASVVVNYANS...</td>\n",
       "      <td>RHEA:64292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>I0DFJ0</td>\n",
       "      <td>RHEA:19497;RHEA:15689;RHEA:12272;-;RHEA:18533</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:12272': 0.7777, 'RHEA:1568...</td>\n",
       "      <td>MSENLQLSAEEMRQLGYQAVDLIIDHMNHLKSKPVSETIDSDILRN...</td>\n",
       "      <td>RHEA:30339;RHEA:12272;RHEA:18533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>A0A0E3T552</td>\n",
       "      <td>RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:14381': 0.998886, 'RHEA:15...</td>\n",
       "      <td>MAIQIPSRLLFIDGEWREPVLKKRIPIINPATEEIIGHIPAATAED...</td>\n",
       "      <td>RHEA:19105;RHEA:30695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Q8VWZ1</td>\n",
       "      <td>RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:14381': 0.99864, 'RHEA:153...</td>\n",
       "      <td>MAITVSSRQLFIDGEWRVPILNKRIPNINPSTENIIGDIPAATKED...</td>\n",
       "      <td>RHEA:19105;RHEA:30695;RHEA:14381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Q56R04</td>\n",
       "      <td>RHEA:15305;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:15305': 0.99942}</td>\n",
       "      <td>MANRNVPIPRRQLYIGGEWREPVKKNRIPIINPATEEIIGDIPAAT...</td>\n",
       "      <td>RHEA:30695;RHEA:17985;RHEA:19105;RHEA:14381;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>C6KEM4</td>\n",
       "      <td>RHEA:15305;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:15305': 0.999946}</td>\n",
       "      <td>MAPPQTIPRRGLFIGGAWREPCLGRRLPVVNPATEATIGDIPAGTA...</td>\n",
       "      <td>RHEA:19105;RHEA:30695;RHEA:14381;RHEA:17985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>A0A0E3T3B5</td>\n",
       "      <td>RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:14381': 0.999133, 'RHEA:15...</td>\n",
       "      <td>MAIQIPSRQLFIDGEWREPVLKKRIPIINPATEQIIGDIPAATAED...</td>\n",
       "      <td>RHEA:19105;RHEA:30695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Q93YB2</td>\n",
       "      <td>RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:14381': 0.999041, 'RHEA:15...</td>\n",
       "      <td>MDIPIPTRQLFINGDWKAPVLNKRIPVINPATQNIIGDIPAATKED...</td>\n",
       "      <td>RHEA:19105;RHEA:30695;RHEA:14381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>B6ECN9</td>\n",
       "      <td>RHEA:15305;RHEA:19105;RHEA:30695;RHEA:17985;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:14381': 0.9957919716835022...</td>\n",
       "      <td>MAIPNIRIPCRQLFIDGEWREPLKKNRLPIINPANEEIIGYIPAAT...</td>\n",
       "      <td>RHEA:19105;RHEA:30695;RHEA:14381;RHEA:17985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>P0DTV9</td>\n",
       "      <td>RHEA:63236;RHEA:69544;RHEA:69540;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:63236': 0.9965959787368774...</td>\n",
       "      <td>MRSEKEVYDIVLNFAKTDKRIRMVTLEGSRTNTNIPPDDFQDFDIT...</td>\n",
       "      <td>RHEA:63236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>H9A1V3</td>\n",
       "      <td>RHEA:16117;RHEA:15805;RHEA:16861;-;RHEA:23176;...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:15805': 0.96253502368927, ...</td>\n",
       "      <td>MGKNYKSLDSVVASDFIALGITSEVAETLHGRLAEIVCNYGAATPQ...</td>\n",
       "      <td>RHEA:43740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Q93K96</td>\n",
       "      <td>RHEA:75903;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:75903': 0.9966149926185608}</td>\n",
       "      <td>MTFLSPPEIPTIKADNGTYYDFNNGARILFPKGEWHVNIIDEESGN...</td>\n",
       "      <td>RHEA:75903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>A3PMF8</td>\n",
       "      <td>RHEA:21824;RHEA:22880;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:21824': 0.999054, 'RHEA:22...</td>\n",
       "      <td>MAFLSDTLARVKPSQTIAVTNKARELAAAGRDVIGLGAGEPDFDTP...</td>\n",
       "      <td>RHEA:21824;RHEA:22880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>Q8KDS8</td>\n",
       "      <td>RHEA:21824;RHEA:22880;RHEA:12601;RHEA:20445;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:12601': 0.9918789863586426...</td>\n",
       "      <td>MSVESFERFLSRRVLSMQESQTMKITGLAKKMQAEGKDVVSLSAGE...</td>\n",
       "      <td>RHEA:21824;RHEA:20445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>Q82WA8</td>\n",
       "      <td>RHEA:21824;RHEA:22880;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:21824': 0.999695, 'RHEA:22...</td>\n",
       "      <td>MKLSQRVQAIKPSPTLAVTAKAARLKAEGKNIIGLGAGEPDFDTPL...</td>\n",
       "      <td>RHEA:21824;RHEA:22880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>B6HLT9</td>\n",
       "      <td>RHEA:20720</td>\n",
       "      <td>{'RHEA:20720': 1.0}</td>\n",
       "      <td>MLHILCQGTPFEIGYEHGSAAKAVIARSIDFAVDLIRGKTKKTDEE...</td>\n",
       "      <td>RHEA:20720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>A0A1H7VGH3</td>\n",
       "      <td>RHEA:16301;-</td>\n",
       "      <td>{'-': 0.971298, 'RHEA:16301': 0.9843170046806335}</td>\n",
       "      <td>MKNRSYEYDVALSFAGENRAYVERVANSLKTKGVKVFYDLFEEANL...</td>\n",
       "      <td>RHEA:19849</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     uniprot_id                                           RXNRECer  \\\n",
       "89       U5NH37                                       RHEA:39895;-   \n",
       "91       F1P963  RHEA:31583;RHEA:67392;RHEA:67604;RHEA:65396;RH...   \n",
       "92       Q7ZWC3  RHEA:31583;RHEA:67392;RHEA:67604;RHEA:65396;RH...   \n",
       "111      D3ZJ25  RHEA:75347;RHEA:70855;RHEA:66324;RHEA:70891;RH...   \n",
       "118      P0DUR6                                                  -   \n",
       "120      P0DUR5                            RHEA:64264;-;RHEA:62680   \n",
       "123  A0A1L9WLD9                 RHEA:64232;RHEA:64512;-;RHEA:12020   \n",
       "124  A0A1L9WLF1                                       RHEA:64236;-   \n",
       "125  A0A1L9WLH9                                       RHEA:64292;-   \n",
       "131      I0DFJ0      RHEA:19497;RHEA:15689;RHEA:12272;-;RHEA:18533   \n",
       "132  A0A0E3T552  RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...   \n",
       "133      Q8VWZ1  RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...   \n",
       "134      Q56R04                                       RHEA:15305;-   \n",
       "135      C6KEM4                                       RHEA:15305;-   \n",
       "136  A0A0E3T3B5  RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...   \n",
       "137      Q93YB2  RHEA:15305;RHEA:30695;RHEA:14381;RHEA:17985;RH...   \n",
       "138      B6ECN9  RHEA:15305;RHEA:19105;RHEA:30695;RHEA:17985;RH...   \n",
       "139      P0DTV9                 RHEA:63236;RHEA:69544;RHEA:69540;-   \n",
       "140      H9A1V3  RHEA:16117;RHEA:15805;RHEA:16861;-;RHEA:23176;...   \n",
       "141      Q93K96                                       RHEA:75903;-   \n",
       "155      A3PMF8                            RHEA:21824;RHEA:22880;-   \n",
       "156      Q8KDS8      RHEA:21824;RHEA:22880;RHEA:12601;RHEA:20445;-   \n",
       "157      Q82WA8                            RHEA:21824;RHEA:22880;-   \n",
       "162      B6HLT9                                         RHEA:20720   \n",
       "163  A0A1H7VGH3                                       RHEA:16301;-   \n",
       "\n",
       "                                    RXNRECer_with_prob  \\\n",
       "89     {'-': 0.7777, 'RHEA:39895': 0.9999889731407166}   \n",
       "91   {'-': 0.7777, 'RHEA:31575': 0.9983620047569275...   \n",
       "92   {'-': 0.7777, 'RHEA:31575': 0.9945039749145508...   \n",
       "111  {'-': 0.7777, 'RHEA:34923': 0.9976549744606018...   \n",
       "118                                    {'-': 0.999997}   \n",
       "120  {'-': 0.7777, 'RHEA:62680': 0.526458, 'RHEA:64...   \n",
       "123  {'-': 0.7777, 'RHEA:12020': 0.199639, 'RHEA:64...   \n",
       "124    {'-': 0.7777, 'RHEA:64236': 0.9974650144577026}   \n",
       "125      {'-': 0.7777, 'RHEA:64292': 0.99652099609375}   \n",
       "131  {'-': 0.7777, 'RHEA:12272': 0.7777, 'RHEA:1568...   \n",
       "132  {'-': 0.7777, 'RHEA:14381': 0.998886, 'RHEA:15...   \n",
       "133  {'-': 0.7777, 'RHEA:14381': 0.99864, 'RHEA:153...   \n",
       "134               {'-': 0.7777, 'RHEA:15305': 0.99942}   \n",
       "135              {'-': 0.7777, 'RHEA:15305': 0.999946}   \n",
       "136  {'-': 0.7777, 'RHEA:14381': 0.999133, 'RHEA:15...   \n",
       "137  {'-': 0.7777, 'RHEA:14381': 0.999041, 'RHEA:15...   \n",
       "138  {'-': 0.7777, 'RHEA:14381': 0.9957919716835022...   \n",
       "139  {'-': 0.7777, 'RHEA:63236': 0.9965959787368774...   \n",
       "140  {'-': 0.7777, 'RHEA:15805': 0.96253502368927, ...   \n",
       "141    {'-': 0.7777, 'RHEA:75903': 0.9966149926185608}   \n",
       "155  {'-': 0.7777, 'RHEA:21824': 0.999054, 'RHEA:22...   \n",
       "156  {'-': 0.7777, 'RHEA:12601': 0.9918789863586426...   \n",
       "157  {'-': 0.7777, 'RHEA:21824': 0.999695, 'RHEA:22...   \n",
       "162                                {'RHEA:20720': 1.0}   \n",
       "163  {'-': 0.971298, 'RHEA:16301': 0.9843170046806335}   \n",
       "\n",
       "                                                   seq  \\\n",
       "89   MGSQETNLPPHVLIFPLPIQGHVNSMLRLAELLCLAELDITFIVSE...   \n",
       "91   MGTSRLYTLVLVLQPERVLLGMKKRGFGAGRWNGFGGKVQEGETIE...   \n",
       "92   MFTSKLLTLVLVVQPGRVLLGMKKRGFGAGKWNGFGGKVQTGETIE...   \n",
       "111  MAVDPPKADPKGVVAVDPTANCGSGLKSREDQGAKAGGCCSSRDQV...   \n",
       "118  MPHPMVAPALRRRRRSSFNPHVAAVMTGSILAGAMLSLSFIAVPVL...   \n",
       "120  MSLPIPPRPADNSTSHSLCLTICGYRRPGMSEFDYRQHMTQVSAPM...   \n",
       "123  MVTYTPPAEAQDLPTKLKVLYFSNEFPTDDLSTLLRRLHSHSKHSS...   \n",
       "124  MGEGGYRQINKTLNVCAFDDYLVTQQSRLPKLLDIEQLSPRVLRVL...   \n",
       "125  MTVSSAPYCLAGQVALVTGSGRGIGAAIAVELGRLGASVVVNYANS...   \n",
       "131  MSENLQLSAEEMRQLGYQAVDLIIDHMNHLKSKPVSETIDSDILRN...   \n",
       "132  MAIQIPSRLLFIDGEWREPVLKKRIPIINPATEEIIGHIPAATAED...   \n",
       "133  MAITVSSRQLFIDGEWRVPILNKRIPNINPSTENIIGDIPAATKED...   \n",
       "134  MANRNVPIPRRQLYIGGEWREPVKKNRIPIINPATEEIIGDIPAAT...   \n",
       "135  MAPPQTIPRRGLFIGGAWREPCLGRRLPVVNPATEATIGDIPAGTA...   \n",
       "136  MAIQIPSRQLFIDGEWREPVLKKRIPIINPATEQIIGDIPAATAED...   \n",
       "137  MDIPIPTRQLFINGDWKAPVLNKRIPVINPATQNIIGDIPAATKED...   \n",
       "138  MAIPNIRIPCRQLFIDGEWREPLKKNRLPIINPANEEIIGYIPAAT...   \n",
       "139  MRSEKEVYDIVLNFAKTDKRIRMVTLEGSRTNTNIPPDDFQDFDIT...   \n",
       "140  MGKNYKSLDSVVASDFIALGITSEVAETLHGRLAEIVCNYGAATPQ...   \n",
       "141  MTFLSPPEIPTIKADNGTYYDFNNGARILFPKGEWHVNIIDEESGN...   \n",
       "155  MAFLSDTLARVKPSQTIAVTNKARELAAAGRDVIGLGAGEPDFDTP...   \n",
       "156  MSVESFERFLSRRVLSMQESQTMKITGLAKKMQAEGKDVVSLSAGE...   \n",
       "157  MKLSQRVQAIKPSPTLAVTAKAARLKAEGKNIIGLGAGEPDFDTPL...   \n",
       "162  MLHILCQGTPFEIGYEHGSAAKAVIARSIDFAVDLIRGKTKKTDEE...   \n",
       "163  MKNRSYEYDVALSFAGENRAYVERVANSLKTKGVKVFYDLFEEANL...   \n",
       "\n",
       "                                       rxn_groundtruth  \n",
       "89                                          RHEA:39895  \n",
       "91   RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...  \n",
       "92   RHEA:31583;RHEA:31575;RHEA:67604;RHEA:67608;RH...  \n",
       "111  RHEA:70891;RHEA:70875;RHEA:70863;RHEA:70887;RH...  \n",
       "118                                         RHEA:64268  \n",
       "120                                         RHEA:64264  \n",
       "123                                         RHEA:64232  \n",
       "124                                         RHEA:64236  \n",
       "125                                         RHEA:64292  \n",
       "131                   RHEA:30339;RHEA:12272;RHEA:18533  \n",
       "132                              RHEA:19105;RHEA:30695  \n",
       "133                   RHEA:19105;RHEA:30695;RHEA:14381  \n",
       "134  RHEA:30695;RHEA:17985;RHEA:19105;RHEA:14381;RH...  \n",
       "135        RHEA:19105;RHEA:30695;RHEA:14381;RHEA:17985  \n",
       "136                              RHEA:19105;RHEA:30695  \n",
       "137                   RHEA:19105;RHEA:30695;RHEA:14381  \n",
       "138        RHEA:19105;RHEA:30695;RHEA:14381;RHEA:17985  \n",
       "139                                         RHEA:63236  \n",
       "140                                         RHEA:43740  \n",
       "141                                         RHEA:75903  \n",
       "155                              RHEA:21824;RHEA:22880  \n",
       "156                              RHEA:21824;RHEA:20445  \n",
       "157                              RHEA:21824;RHEA:22880  \n",
       "162                                         RHEA:20720  \n",
       "163                                         RHEA:19849  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer[res_ecrecer.rxn_groundtruth!='-'].iloc[30:55]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"protein information\": {\n",
      "        \"uniprot id\": \"Q8KDS8\",\n",
      "        \"protein amino acid sequence\": \"MSVESFERFLSRRVLSMQESQTMKITGLAKKMQAEGKDVVSLSAGEPDFPTPENVCEAGIEAIRKGFTRYTANSGIPELKKAIIRKLQRDNGLEYAEDEIIVSNGGKQALANTFLALCDEGDEVIVPAPYWVSFPEMARLAEATPVIVETSIETGYKMTPEQLAAAITPKTRILVLNSPSNPSGAVYNEAEVRALMQVIEGKEIFVLSDEMYDMICYGGVRPFSPARIPEMKPWVIVSNGTSKSYSMTGWRIGYLAAPKWIINACDKIQSQTTSNANSIAQKAAVAALDGDQSIVEQRRAEFEKRRDFMFRELNTISGIECTLPEGAFYIFPSIKGLLGKTFGGKVMKDSTDVAEYLLTEHYVATVPGDAFGAPENLRLSYAASIEELAEAVNRIRKAFS\"\n",
      "    },\n",
      "    \"reaction information\": {\n",
      "        \"candidate reaction 1\": {\n",
      "            \"reaction id\": \"RHEA:21824\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + L-aspartate = L-glutamate + oxaloacetate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:29991 = CHEBI:29985 + CHEBI:16452\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+][C@@H](CC([O-])=O)C([O-])=O>>[NH3+][C@@H](CCC([O-])=O)C([O-])=O.[O-]C(=O)CC(=O)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.1\"\n",
      "        },\n",
      "        \"candidate reaction 2\": {\n",
      "            \"reaction id\": \"RHEA:22880\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + L-arogenate = L-glutamate + prephenate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:58180 = CHEBI:29985 + CHEBI:29934\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+][C@@H](CC1(C=CC(O)C=C1)C([O-])=O)C([O-])=O>>[NH3+][C@@H](CCC([O-])=O)C([O-])=O.O[C@H]1C=C[C@](CC(=O)C([O-])=O)(C=C1)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.79\"\n",
      "        },\n",
      "        \"candidate reaction 3\": {\n",
      "            \"reaction id\": \"RHEA:12601\",\n",
      "            \"reaction equation\": \"2-oxoglutarate + L-2-aminoadipate = 2-oxoadipate + L-glutamate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:16810 + CHEBI:58672 = CHEBI:57499 + CHEBI:29985\",\n",
      "            \"reaction equation in SMILES format\": \"[O-]C(=O)CCC(=O)C([O-])=O.[NH3+][C@@H](CCCC([O-])=O)C([O-])=O>>[O-]C(=O)CCCC(=O)C([O-])=O.[NH3+][C@@H](CCC([O-])=O)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.39\"\n",
      "        },\n",
      "        \"candidate reaction 4\": {\n",
      "            \"reaction id\": \"RHEA:20445\",\n",
      "            \"reaction equation\": \"L-arogenate + oxaloacetate = L-aspartate + prephenate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:58180 + CHEBI:16452 = CHEBI:29991 + CHEBI:29934\",\n",
      "            \"reaction equation in SMILES format\": \"[NH3+][C@@H](CC1(C=CC(O)C=C1)C([O-])=O)C([O-])=O.[O-]C(=O)CC(=O)C([O-])=O>>[NH3+][C@@H](CC([O-])=O)C([O-])=O.O[C@H]1C=C[C@](CC(=O)C([O-])=O)(C=C1)C([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:2.6.1.78\"\n",
      "        },\n",
      "        \"candidate reaction 5\": {\n",
      "            \"reaction id\": \"-\",\n",
      "            \"reaction equation\": \"-\"\n",
      "        }\n",
      "    }\n",
      "}\n",
      "ChatCompletion(id='gen-1747300015-B1zG9r8edgwGxcMbCQrG', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\\n  \"results\": [\\n    {\\n      \"reaction_id\": \"RHEA:21824\",\\n      \"selected\": \"yes\",\\n      \"rank\": 1,\\n      \"confidence\": 0.98,\\n      \"reason\": \"The UniProt ID Q8KDS8 corresponds to Aspartate aminotransferase (EC 2.6.1.1), which catalyzes the transamination between 2-oxoglutarate and L-aspartate to form L-glutamate and oxaloacetate. The sequence contains conserved motifs (e.g., the lysine residue for PLP binding) characteristic of aminotransferases, supporting this function.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:22880\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.35,\\n      \"reason\": \"While this is a related aminotransferase reaction, the substrate specificity for L-arogenate is not supported by the known function or sequence features of Q8KDS8, which is specific for L-aspartate.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:12601\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.30,\\n      \"reason\": \"This reaction is catalyzed by 2-aminoadipate transaminase (EC 2.6.1.39). Q8KDS8 lacks the sequence features and substrate specificity for L-2-aminoadipate, as supported by UniProt and BRENDA annotations.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:20445\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.20,\\n      \"reason\": \"This reaction involves L-arogenate and is catalyzed by arogenate transaminases (EC 2.6.1.78). Q8KDS8 is not annotated for this activity and lacks the necessary substrate recognition motifs.\"\\n    },\\n    {\\n      \"reaction_id\": \"-\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.01,\\n      \"reason\": \"Q8KDS8 is a well-characterized aminotransferase with clear catalytic activity; the dash is not appropriate.\"\\n    }\\n  ]\\n}', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning=None), native_finish_reason='stop')], created=1747300015, model='openai/gpt-4.1', object='chat.completion', service_tier=None, system_fingerprint='fp_b38e740b47', usage=CompletionUsage(completion_tokens=471, prompt_tokens=1536, total_tokens=2007, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=0, rejected_prediction_tokens=None), prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0)), provider='OpenAI')\n",
      "{'results': [{'reaction_id': 'RHEA:21824', 'selected': 'yes', 'rank': 1, 'confidence': 0.98, 'reason': 'The UniProt ID Q8KDS8 corresponds to Aspartate aminotransferase (EC 2.6.1.1), which catalyzes the transamination between 2-oxoglutarate and L-aspartate to form L-glutamate and oxaloacetate. The sequence contains conserved motifs (e.g., the lysine residue for PLP binding) characteristic of aminotransferases, supporting this function.'}, {'reaction_id': 'RHEA:22880', 'selected': 'no', 'confidence': 0.35, 'reason': 'While this is a related aminotransferase reaction, the substrate specificity for L-arogenate is not supported by the known function or sequence features of Q8KDS8, which is specific for L-aspartate.'}, {'reaction_id': 'RHEA:12601', 'selected': 'no', 'confidence': 0.3, 'reason': 'This reaction is catalyzed by 2-aminoadipate transaminase (EC 2.6.1.39). Q8KDS8 lacks the sequence features and substrate specificity for L-2-aminoadipate, as supported by UniProt and BRENDA annotations.'}, {'reaction_id': 'RHEA:20445', 'selected': 'no', 'confidence': 0.2, 'reason': 'This reaction involves L-arogenate and is catalyzed by arogenate transaminases (EC 2.6.1.78). Q8KDS8 is not annotated for this activity and lacks the necessary substrate recognition motifs.'}, {'reaction_id': '-', 'selected': 'no', 'confidence': 0.01, 'reason': 'Q8KDS8 is a well-characterized aminotransferase with clear catalytic activity; the dash is not appropriate.'}]}\n"
     ]
    }
   ],
   "source": [
    "res_df_156, res_json_156 = session_chat(chat_bot=cli5, ix=156, system_prompt=system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>selected</th>\n",
       "      <th>rank</th>\n",
       "      <th>confidence</th>\n",
       "      <th>reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHEA:21824</td>\n",
       "      <td>yes</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.98</td>\n",
       "      <td>The UniProt ID Q8KDS8 corresponds to Aspartate...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RHEA:22880</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.35</td>\n",
       "      <td>While this is a related aminotransferase react...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RHEA:12601</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.30</td>\n",
       "      <td>This reaction is catalyzed by 2-aminoadipate t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RHEA:20445</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.20</td>\n",
       "      <td>This reaction involves L-arogenate and is cata...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-</td>\n",
       "      <td>no</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Q8KDS8 is a well-characterized aminotransferas...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id selected  rank  confidence  \\\n",
       "0  RHEA:21824      yes   1.0        0.98   \n",
       "1  RHEA:22880       no   inf        0.35   \n",
       "2  RHEA:12601       no   inf        0.30   \n",
       "3  RHEA:20445       no   inf        0.20   \n",
       "4           -       no   inf        0.01   \n",
       "\n",
       "                                              reason  \n",
       "0  The UniProt ID Q8KDS8 corresponds to Aspartate...  \n",
       "1  While this is a related aminotransferase react...  \n",
       "2  This reaction is catalyzed by 2-aminoadipate t...  \n",
       "3  This reaction involves L-arogenate and is cata...  \n",
       "4  Q8KDS8 is a well-characterized aminotransferas...  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df_156"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>Q9VK50</td>\n",
       "      <td>RHEA:15389;RHEA:13665</td>\n",
       "      <td>{'RHEA:13665': 0.668856, 'RHEA:15389': 0.97908...</td>\n",
       "      <td>MPRSLGNCQLNYSKERMWEPGYLKAKCAELRLESEFRLYRIRLWKS...</td>\n",
       "      <td>RHEA:15389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>Q8LPL7</td>\n",
       "      <td>RHEA:16001;RHEA:24408;RHEA:28190;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:16001': 0.9868509769439697...</td>\n",
       "      <td>MEWIQSLPKIELHAHLNGSIRDSTLLELARVLGEKGVIVFADVEHV...</td>\n",
       "      <td>RHEA:16001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>A0A509ALD0</td>\n",
       "      <td>RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;-;...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:12745': 0.987500011920929,...</td>\n",
       "      <td>MEIPNEEIKFLKKEDIKNINLNGMNKKERYEIWKKIPKVELHCHLD...</td>\n",
       "      <td>RHEA:25025;RHEA:24408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>K6UCV4</td>\n",
       "      <td>RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:12745': 0.9861199855804443...</td>\n",
       "      <td>MNILQEPIDFLKKDEIKNIDLSQMSKKERYKIWKRIPKCELHCHLD...</td>\n",
       "      <td>RHEA:25025;RHEA:24408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>Q8IJA9</td>\n",
       "      <td>RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:12745': 0.9844340085983276...</td>\n",
       "      <td>MNCKNMDTSYEIINYLTKDELDIDLSCMDKKERYKIWKRLPKCELH...</td>\n",
       "      <td>RHEA:25025;RHEA:24408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>A0A0K1SC59</td>\n",
       "      <td>RHEA:12745;RHEA:52644;RHEA:52648;RHEA:24408;RH...</td>\n",
       "      <td>{'RHEA:12745': 0.9848179817199707, 'RHEA:23688...</td>\n",
       "      <td>MTILHEEINFLKKDELNINLKCLDKKERYKIWRRIPKCELHCHLDL...</td>\n",
       "      <td>RHEA:24408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>A0A1Y3DYH2</td>\n",
       "      <td>RHEA:24408;RHEA:28190;RHEA:23688;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:23688': 0.976769, 'RHEA:24...</td>\n",
       "      <td>MNILQEPIDFLKKDELKNIDLSQMDKKERYKIWKRIPKCELHCHLD...</td>\n",
       "      <td>RHEA:25025;RHEA:24408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>A5KE01</td>\n",
       "      <td>RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:12745': 0.985480010509491,...</td>\n",
       "      <td>MNILQEPIDFLKKEELKNIDLSQMSKKERYKIWKRIPKCELHCHLD...</td>\n",
       "      <td>RHEA:25025;RHEA:24408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>A0A1S3YCW2</td>\n",
       "      <td>RHEA:17641</td>\n",
       "      <td>{'RHEA:17641': 0.999725}</td>\n",
       "      <td>MPALGCCVDAAVSPPPGYSFLWDSSLPAPEIFPSGVPPSTNTAVAT...</td>\n",
       "      <td>RHEA:17641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>A0A1S3WYR7</td>\n",
       "      <td>RHEA:17641</td>\n",
       "      <td>{'RHEA:17641': 0.999783}</td>\n",
       "      <td>MPALGCCVDAAVSPPPGYSFLWDSSLPAPEIFPSGVPLSTNTAATT...</td>\n",
       "      <td>RHEA:17641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>P72539</td>\n",
       "      <td>RHEA:21732;RHEA:11672</td>\n",
       "      <td>{'RHEA:11672': 0.994222, 'RHEA:21732': 0.996486}</td>\n",
       "      <td>MRTVRTLLIDNYDSFTYNLFQMLAEVNGAAPLVVRNDDTRTWQALA...</td>\n",
       "      <td>RHEA:11672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>F2RB79</td>\n",
       "      <td>RHEA:21732;RHEA:11672</td>\n",
       "      <td>{'RHEA:11672': 0.9908679723739624, 'RHEA:21732...</td>\n",
       "      <td>MRTLLIDNYDSFTHNLFQYIGEATGQPPVVVPNDADWSRLPLEDFD...</td>\n",
       "      <td>RHEA:11672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>C0P9J6</td>\n",
       "      <td>RHEA:15305;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:15305': 0.999959}</td>\n",
       "      <td>MASPAMVPLRQLFVDGEWRPPAQGRRLPVVNPTTEAHIGEIPAGTA...</td>\n",
       "      <td>RHEA:30695;RHEA:17985;RHEA:19105;RHEA:14381;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>G5DDC2</td>\n",
       "      <td>RHEA:15305;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:15305': 0.999959}</td>\n",
       "      <td>MMASQAMVPLRQLFVDGEWRPPAQGRRLPVVNPTTEAHIGEIPAGT...</td>\n",
       "      <td>RHEA:30695;RHEA:17985;RHEA:19105;RHEA:14381;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>A0A165U5V5</td>\n",
       "      <td>RHEA:33887;RHEA:10736;RHEA:10740;RHEA:25290;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:10736': 0.7777, 'RHEA:1074...</td>\n",
       "      <td>MNGCCSQDPTSKRLEGKVAVITGGASGIGACTVKLFVKHGAKVVIA...</td>\n",
       "      <td>RHEA:10736;RHEA:10740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>A0A2I7G3B3</td>\n",
       "      <td>RHEA:10740;RHEA:10736;RHEA:21284;RHEA:42052;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:10736': 0.998354, 'RHEA:10...</td>\n",
       "      <td>MSLNTPDVIICKAAVVRELGRSVMVEEIKVDPPKATEVRIKMLFAS...</td>\n",
       "      <td>RHEA:34347;RHEA:10664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>M2YJK1</td>\n",
       "      <td>RHEA:35475;RHEA:35479;-;RHEA:17389</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:17389': 0.17768, 'RHEA:354...</td>\n",
       "      <td>MEALNTNVDLAGLEGRSVLVTGGASGIGLATAKAWAAAGAYVTLAD...</td>\n",
       "      <td>RHEA:35479;RHEA:35475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>P74721</td>\n",
       "      <td>RHEA:15937;-;RHEA:19257</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:15937': 0.994628, 'RHEA:19...</td>\n",
       "      <td>MIKAYAALEANGKLQPFEYDPGALGANEVEIEVQYCGVCHSDLSMI...</td>\n",
       "      <td>RHEA:15937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>E5Y379</td>\n",
       "      <td>RHEA:23288;RHEA:29595;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:23288': 0.994144, 'RHEA:29...</td>\n",
       "      <td>MDVRQQDVERIVVEVLKKMMSDQPTAAATTVVAASGCDCGDFGLFD...</td>\n",
       "      <td>RHEA:23288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>A0A0H2ZM56</td>\n",
       "      <td>RHEA:10736;RHEA:10740;RHEA:16185;RHEA:23288;RH...</td>\n",
       "      <td>{'RHEA:10736': 0.995319, 'RHEA:10740': 0.99531...</td>\n",
       "      <td>MADKKTVTPEEKKLVAEKHVDELVQKALVALEEMRKLNQEQVDYIV...</td>\n",
       "      <td>RHEA:16185;RHEA:25290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>A0A0H2URT2</td>\n",
       "      <td>RHEA:10736;RHEA:10740;RHEA:16185;RHEA:23288;RH...</td>\n",
       "      <td>{'RHEA:10736': 0.995307, 'RHEA:10740': 0.99530...</td>\n",
       "      <td>MADKKTVTPEEKKLVAEKHVDELVQKALVALEEMRKLDQEQVDYIV...</td>\n",
       "      <td>RHEA:16185;RHEA:25290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>Q8GIX7</td>\n",
       "      <td>RHEA:10740;RHEA:10736;RHEA:25290</td>\n",
       "      <td>{'RHEA:10736': 0.999835, 'RHEA:10740': 0.99992...</td>\n",
       "      <td>MKAAVLHEFGQSLQIEEVDIPTPGAGEIVVKMQASGVCHTDLHAVE...</td>\n",
       "      <td>RHEA:10736;RHEA:50704;RHEA:41984;RHEA:25290;RH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>A0A0S1X9S7</td>\n",
       "      <td>RHEA:16981;RHEA:16985;RHEA:21412;RHEA:21416;RH...</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:10736': 0.7777, 'RHEA:1074...</td>\n",
       "      <td>MQFFSLKTRIVLGEGSLSYIKSVAKKHSRVLIFSSKSMRVHGFLNE...</td>\n",
       "      <td>RHEA:10736;RHEA:60972;RHEA:33199;RHEA:25290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>I6X235</td>\n",
       "      <td>RHEA:60876;RHEA:11800;RHEA:48868;RHEA:10412;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:10412': 0.948521, 'RHEA:11...</td>\n",
       "      <td>MAEHDFETISSETLHTGAIFALRRDQVRMPGGGIVTREVVEHFGAV...</td>\n",
       "      <td>RHEA:62356;RHEA:32063;RHEA:10412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>H3BCW1</td>\n",
       "      <td>RHEA:58112;RHEA:68792;RHEA:58256;RHEA:52216;-</td>\n",
       "      <td>{'-': 0.7777, 'RHEA:52216': 0.9944109916687012...</td>\n",
       "      <td>MSAVGRLAAVSLAQVRGALCGALLGDCMGAEFEGSDAVELPDVLEF...</td>\n",
       "      <td>RHEA:58256;RHEA:68792;RHEA:52216;RHEA:58112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     uniprot_id                                           RXNRECer  \\\n",
       "348      Q9VK50                              RHEA:15389;RHEA:13665   \n",
       "350      Q8LPL7                 RHEA:16001;RHEA:24408;RHEA:28190;-   \n",
       "353  A0A509ALD0  RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;-;...   \n",
       "354      K6UCV4  RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;RH...   \n",
       "355      Q8IJA9  RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;RH...   \n",
       "356  A0A0K1SC59  RHEA:12745;RHEA:52644;RHEA:52648;RHEA:24408;RH...   \n",
       "357  A0A1Y3DYH2                 RHEA:24408;RHEA:28190;RHEA:23688;-   \n",
       "358      A5KE01  RHEA:12745;RHEA:52644;RHEA:52648;RHEA:23688;RH...   \n",
       "359  A0A1S3YCW2                                         RHEA:17641   \n",
       "360  A0A1S3WYR7                                         RHEA:17641   \n",
       "362      P72539                              RHEA:21732;RHEA:11672   \n",
       "363      F2RB79                              RHEA:21732;RHEA:11672   \n",
       "367      C0P9J6                                       RHEA:15305;-   \n",
       "368      G5DDC2                                       RHEA:15305;-   \n",
       "369  A0A165U5V5      RHEA:33887;RHEA:10736;RHEA:10740;RHEA:25290;-   \n",
       "371  A0A2I7G3B3  RHEA:10740;RHEA:10736;RHEA:21284;RHEA:42052;RH...   \n",
       "373      M2YJK1                 RHEA:35475;RHEA:35479;-;RHEA:17389   \n",
       "374      P74721                            RHEA:15937;-;RHEA:19257   \n",
       "375      E5Y379                            RHEA:23288;RHEA:29595;-   \n",
       "376  A0A0H2ZM56  RHEA:10736;RHEA:10740;RHEA:16185;RHEA:23288;RH...   \n",
       "377  A0A0H2URT2  RHEA:10736;RHEA:10740;RHEA:16185;RHEA:23288;RH...   \n",
       "378      Q8GIX7                   RHEA:10740;RHEA:10736;RHEA:25290   \n",
       "379  A0A0S1X9S7  RHEA:16981;RHEA:16985;RHEA:21412;RHEA:21416;RH...   \n",
       "384      I6X235      RHEA:60876;RHEA:11800;RHEA:48868;RHEA:10412;-   \n",
       "385      H3BCW1      RHEA:58112;RHEA:68792;RHEA:58256;RHEA:52216;-   \n",
       "\n",
       "                                    RXNRECer_with_prob  \\\n",
       "348  {'RHEA:13665': 0.668856, 'RHEA:15389': 0.97908...   \n",
       "350  {'-': 0.7777, 'RHEA:16001': 0.9868509769439697...   \n",
       "353  {'-': 0.7777, 'RHEA:12745': 0.987500011920929,...   \n",
       "354  {'-': 0.7777, 'RHEA:12745': 0.9861199855804443...   \n",
       "355  {'-': 0.7777, 'RHEA:12745': 0.9844340085983276...   \n",
       "356  {'RHEA:12745': 0.9848179817199707, 'RHEA:23688...   \n",
       "357  {'-': 0.7777, 'RHEA:23688': 0.976769, 'RHEA:24...   \n",
       "358  {'-': 0.7777, 'RHEA:12745': 0.985480010509491,...   \n",
       "359                           {'RHEA:17641': 0.999725}   \n",
       "360                           {'RHEA:17641': 0.999783}   \n",
       "362   {'RHEA:11672': 0.994222, 'RHEA:21732': 0.996486}   \n",
       "363  {'RHEA:11672': 0.9908679723739624, 'RHEA:21732...   \n",
       "367              {'-': 0.7777, 'RHEA:15305': 0.999959}   \n",
       "368              {'-': 0.7777, 'RHEA:15305': 0.999959}   \n",
       "369  {'-': 0.7777, 'RHEA:10736': 0.7777, 'RHEA:1074...   \n",
       "371  {'-': 0.7777, 'RHEA:10736': 0.998354, 'RHEA:10...   \n",
       "373  {'-': 0.7777, 'RHEA:17389': 0.17768, 'RHEA:354...   \n",
       "374  {'-': 0.7777, 'RHEA:15937': 0.994628, 'RHEA:19...   \n",
       "375  {'-': 0.7777, 'RHEA:23288': 0.994144, 'RHEA:29...   \n",
       "376  {'RHEA:10736': 0.995319, 'RHEA:10740': 0.99531...   \n",
       "377  {'RHEA:10736': 0.995307, 'RHEA:10740': 0.99530...   \n",
       "378  {'RHEA:10736': 0.999835, 'RHEA:10740': 0.99992...   \n",
       "379  {'-': 0.7777, 'RHEA:10736': 0.7777, 'RHEA:1074...   \n",
       "384  {'-': 0.7777, 'RHEA:10412': 0.948521, 'RHEA:11...   \n",
       "385  {'-': 0.7777, 'RHEA:52216': 0.9944109916687012...   \n",
       "\n",
       "                                                   seq  \\\n",
       "348  MPRSLGNCQLNYSKERMWEPGYLKAKCAELRLESEFRLYRIRLWKS...   \n",
       "350  MEWIQSLPKIELHAHLNGSIRDSTLLELARVLGEKGVIVFADVEHV...   \n",
       "353  MEIPNEEIKFLKKEDIKNINLNGMNKKERYEIWKKIPKVELHCHLD...   \n",
       "354  MNILQEPIDFLKKDEIKNIDLSQMSKKERYKIWKRIPKCELHCHLD...   \n",
       "355  MNCKNMDTSYEIINYLTKDELDIDLSCMDKKERYKIWKRLPKCELH...   \n",
       "356  MTILHEEINFLKKDELNINLKCLDKKERYKIWRRIPKCELHCHLDL...   \n",
       "357  MNILQEPIDFLKKDELKNIDLSQMDKKERYKIWKRIPKCELHCHLD...   \n",
       "358  MNILQEPIDFLKKEELKNIDLSQMSKKERYKIWKRIPKCELHCHLD...   \n",
       "359  MPALGCCVDAAVSPPPGYSFLWDSSLPAPEIFPSGVPPSTNTAVAT...   \n",
       "360  MPALGCCVDAAVSPPPGYSFLWDSSLPAPEIFPSGVPLSTNTAATT...   \n",
       "362  MRTVRTLLIDNYDSFTYNLFQMLAEVNGAAPLVVRNDDTRTWQALA...   \n",
       "363  MRTLLIDNYDSFTHNLFQYIGEATGQPPVVVPNDADWSRLPLEDFD...   \n",
       "367  MASPAMVPLRQLFVDGEWRPPAQGRRLPVVNPTTEAHIGEIPAGTA...   \n",
       "368  MMASQAMVPLRQLFVDGEWRPPAQGRRLPVVNPTTEAHIGEIPAGT...   \n",
       "369  MNGCCSQDPTSKRLEGKVAVITGGASGIGACTVKLFVKHGAKVVIA...   \n",
       "371  MSLNTPDVIICKAAVVRELGRSVMVEEIKVDPPKATEVRIKMLFAS...   \n",
       "373  MEALNTNVDLAGLEGRSVLVTGGASGIGLATAKAWAAAGAYVTLAD...   \n",
       "374  MIKAYAALEANGKLQPFEYDPGALGANEVEIEVQYCGVCHSDLSMI...   \n",
       "375  MDVRQQDVERIVVEVLKKMMSDQPTAAATTVVAASGCDCGDFGLFD...   \n",
       "376  MADKKTVTPEEKKLVAEKHVDELVQKALVALEEMRKLNQEQVDYIV...   \n",
       "377  MADKKTVTPEEKKLVAEKHVDELVQKALVALEEMRKLDQEQVDYIV...   \n",
       "378  MKAAVLHEFGQSLQIEEVDIPTPGAGEIVVKMQASGVCHTDLHAVE...   \n",
       "379  MQFFSLKTRIVLGEGSLSYIKSVAKKHSRVLIFSSKSMRVHGFLNE...   \n",
       "384  MAEHDFETISSETLHTGAIFALRRDQVRMPGGGIVTREVVEHFGAV...   \n",
       "385  MSAVGRLAAVSLAQVRGALCGALLGDCMGAEFEGSDAVELPDVLEF...   \n",
       "\n",
       "                                       rxn_groundtruth  \n",
       "348                                         RHEA:15389  \n",
       "350                                         RHEA:16001  \n",
       "353                              RHEA:25025;RHEA:24408  \n",
       "354                              RHEA:25025;RHEA:24408  \n",
       "355                              RHEA:25025;RHEA:24408  \n",
       "356                                         RHEA:24408  \n",
       "357                              RHEA:25025;RHEA:24408  \n",
       "358                              RHEA:25025;RHEA:24408  \n",
       "359                                         RHEA:17641  \n",
       "360                                         RHEA:17641  \n",
       "362                                         RHEA:11672  \n",
       "363                                         RHEA:11672  \n",
       "367  RHEA:30695;RHEA:17985;RHEA:19105;RHEA:14381;RH...  \n",
       "368  RHEA:30695;RHEA:17985;RHEA:19105;RHEA:14381;RH...  \n",
       "369                              RHEA:10736;RHEA:10740  \n",
       "371                              RHEA:34347;RHEA:10664  \n",
       "373                              RHEA:35479;RHEA:35475  \n",
       "374                                         RHEA:15937  \n",
       "375                                         RHEA:23288  \n",
       "376                              RHEA:16185;RHEA:25290  \n",
       "377                              RHEA:16185;RHEA:25290  \n",
       "378  RHEA:10736;RHEA:50704;RHEA:41984;RHEA:25290;RH...  \n",
       "379        RHEA:10736;RHEA:60972;RHEA:33199;RHEA:25290  \n",
       "384                   RHEA:62356;RHEA:32063;RHEA:10412  \n",
       "385        RHEA:58256;RHEA:68792;RHEA:52216;RHEA:58112  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer[res_ecrecer.rxn_groundtruth!='-'].iloc[130:155]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"protein information\": {\n",
      "        \"uniprot id\": \"Q9VK50\",\n",
      "        \"protein amino acid sequence\": \"MPRSLGNCQLNYSKERMWEPGYLKAKCAELRLESEFRLYRIRLWKSYLLTFFMLHIFVTSVHCALLLATIERRSIIYFDVALSIGCALVLILVLSVNFCDEFIAKHTWYMYASSIFASLTLVFADLTESIYHTYAHSWILGTFYDTYIIYMIYMFLPIHFISGAVLLALLVSGLYILYFVIFIAQGFAQFASALFSVGGMSVDIVHYLCLNLVGIFYRVMNDTVVRSSFLDRHQYIKEKIWLRNARLQEKQLLDSILPPQISLPLQKDIQGRIVMAKQGIHSWTAMERTMAIQIHPDVSILYADVVNYTHLTTTLTVEMLVKVLHDLYGRFDLAAYRYKVQRIKFLGDCYYCVAGLSDPDPDHANNCVILGLSMINHIMEVRDIHGLDINMRIGVHSGNLFAGVIGEAKLQFDIWGLDVTIANVLESTGVPGCVHISGATLNNLDVNRFDIEDGPEEARDHPLLKKYRIRSYIIRQDLHMDDEDSDEFLGDLHSISLCNMGAQPRISDSANQSLRALFHEELREEFRKMPVSAFSPKRLLGICRFNTGKEVPAHQNLNICLTFTDPILERAYLKQTDYMYKYSIILSASVGCSLVYIELMDTQMICSSCFVLPASVATIQCILALIAWYKKYCWTRYGRNNVPHHYNGFSCFIFRIHDKILNSLPIRICIYLFLMISSFFVMCLIVMSCQREEFEMAYIEERLFHYEQEAHICFHPWVTTNMLSLMICLTFTFAHIPIMVKTAVAILETLAYLLLIFFQFDFVFHHSVTTNPYFKSEYAHALLICITFLIMFVKERQIEFTNKVNFNWRVDLRKEENAASLTNHSIIIILNNILPSHIVDVYLNSLAKHELYFENYRMVSVMFAMLINFEMDLRSLRVLNEIIAEFDTLLLFYKEYYTVEKIKIVGCTYMAACGLDLNFAGSTSTNRKESIPPTEFNEEQSRRILFQQSNEDLDEVVFVMTSYALDMMRTLAKSNEAYQSIAGDRNITDGTIAIGISSGEVMAGIVGASQPHYDIWGNPVNMASRMESTGLPGHIHVTEETSEILQQFGITCSYRGMTFVKGRGKIPTYLVGIDENLNFIPQKATRFPSHQERSTVISLQSTYTHAENNNSIASTSRHTLQSL\"\n",
      "    },\n",
      "    \"reaction information\": {\n",
      "        \"candidate reaction 1\": {\n",
      "            \"reaction id\": \"RHEA:15389\",\n",
      "            \"reaction equation\": \"ATP = 3',5'-cyclic AMP + diphosphate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:30616 = CHEBI:58165 + CHEBI:33019\",\n",
      "            \"reaction equation in SMILES format\": \"Nc1ncnc2n(cnc12)[C@@H]1O[C@H](COP([O-])(=O)OP([O-])(=O)OP([O-])([O-])=O)[C@@H](O)[C@H]1O>>Nc1ncnc2n(cnc12)[C@@H]1O[C@@H]2COP([O-])(=O)O[C@H]2[C@H]1O.OP([O-])(=O)OP([O-])([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:4.6.1.1\"\n",
      "        },\n",
      "        \"candidate reaction 2\": {\n",
      "            \"reaction id\": \"RHEA:13665\",\n",
      "            \"reaction equation\": \"GTP = 3',5'-cyclic GMP + diphosphate\",\n",
      "            \"reaction equation in ChEBI format\": \"CHEBI:37565 = CHEBI:57746 + CHEBI:33019\",\n",
      "            \"reaction equation in SMILES format\": \"Nc1nc2n(cnc2c(=O)[nH]1)[C@@H]1O[C@H](COP([O-])(=O)OP([O-])(=O)OP([O-])([O-])=O)[C@@H](O)[C@H]1O>>Nc1nc2n(cnc2c(=O)[nH]1)[C@@H]1O[C@@H]2COP([O-])(=O)O[C@H]2[C@H]1O.OP([O-])(=O)OP([O-])([O-])=O\",\n",
      "            \"reaction associated Enzyme Commission Number\": \"EC:4.6.1.2\"\n",
      "        }\n",
      "    }\n",
      "}\n",
      "ChatCompletion(id='gen-1747305103-uzo1G175cAPOIn6MnQdM', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\\n  \"results\": [\\n    {\\n      \"reaction_id\": \"RHEA:15389\",\\n      \"selected\": \"yes\",\\n      \"rank\": 1,\\n      \"confidence\": 0.95,\\n      \"reason\": \"The UniProt ID Q9VK50 corresponds to the Drosophila melanogaster protein \\'Adenylate cyclase type 3\\' (AC3), which catalyzes the conversion of ATP to 3\\',5\\'-cyclic AMP and diphosphate (EC 4.6.1.1). The protein sequence contains conserved domains characteristic of class III adenylyl cyclases, including the CYTH domain and key residues involved in ATP binding and catalysis. There is strong literature and database support for this function.\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:13665\",\\n      \"selected\": \"no\",\\n      \"confidence\": 0.2,\\n      \"reason\": \"The protein lacks the conserved motifs and active site residues required for guanylate cyclase activity (EC 4.6.1.2). There is no evidence from UniProt, BRENDA, or sequence analysis that Q9VK50 catalyzes the conversion of GTP to cGMP.\"\\n    }\\n  ]\\n}', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning=None), native_finish_reason='stop')], created=1747305103, model='openai/gpt-4.1', object='chat.completion', service_tier=None, system_fingerprint='fp_b38e740b47', usage=CompletionUsage(completion_tokens=265, prompt_tokens=1500, total_tokens=1765, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=0, rejected_prediction_tokens=None), prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0)), provider='OpenAI')\n",
      "{'results': [{'reaction_id': 'RHEA:15389', 'selected': 'yes', 'rank': 1, 'confidence': 0.95, 'reason': \"The UniProt ID Q9VK50 corresponds to the Drosophila melanogaster protein 'Adenylate cyclase type 3' (AC3), which catalyzes the conversion of ATP to 3',5'-cyclic AMP and diphosphate (EC 4.6.1.1). The protein sequence contains conserved domains characteristic of class III adenylyl cyclases, including the CYTH domain and key residues involved in ATP binding and catalysis. There is strong literature and database support for this function.\"}, {'reaction_id': 'RHEA:13665', 'selected': 'no', 'confidence': 0.2, 'reason': 'The protein lacks the conserved motifs and active site residues required for guanylate cyclase activity (EC 4.6.1.2). There is no evidence from UniProt, BRENDA, or sequence analysis that Q9VK50 catalyzes the conversion of GTP to cGMP.'}]}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(  reaction_id selected  rank  confidence  \\\n",
       " 0  RHEA:15389      yes   1.0        0.95   \n",
       " 1  RHEA:13665       no   inf        0.20   \n",
       " \n",
       "                                               reason  \n",
       " 0  The UniProt ID Q9VK50 corresponds to the Droso...  \n",
       " 1  The protein lacks the conserved motifs and act...  ,\n",
       " {'results': [{'reaction_id': 'RHEA:15389',\n",
       "    'selected': 'yes',\n",
       "    'rank': 1,\n",
       "    'confidence': 0.95,\n",
       "    'reason': \"The UniProt ID Q9VK50 corresponds to the Drosophila melanogaster protein 'Adenylate cyclase type 3' (AC3), which catalyzes the conversion of ATP to 3',5'-cyclic AMP and diphosphate (EC 4.6.1.1). The protein sequence contains conserved domains characteristic of class III adenylyl cyclases, including the CYTH domain and key residues involved in ATP binding and catalysis. There is strong literature and database support for this function.\"},\n",
       "   {'reaction_id': 'RHEA:13665',\n",
       "    'selected': 'no',\n",
       "    'confidence': 0.2,\n",
       "    'reason': 'The protein lacks the conserved motifs and active site residues required for guanylate cyclase activity (EC 4.6.1.2). There is no evidence from UniProt, BRENDA, or sequence analysis that Q9VK50 catalyzes the conversion of GTP to cGMP.'}]})"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session_chat(chat_bot=cli5, ix=348, system_prompt=system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rxn' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43msession_chat\u001b[49m\u001b[43m(\u001b[49m\u001b[43mchat_bot\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcli5\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m89\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msystem_prompt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msystem_prompt\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[12], line 32\u001b[0m, in \u001b[0;36msession_chat\u001b[0;34m(chat_bot, ix, system_prompt, debug)\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msession_chat\u001b[39m(chat_bot, ix, system_prompt, debug\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[0;32m---> 32\u001b[0m     query_user \u001b[38;5;241m=\u001b[39m make_make_user_query(uniprot_id\u001b[38;5;241m=\u001b[39mres_ecrecer\u001b[38;5;241m.\u001b[39miloc[ix]\u001b[38;5;241m.\u001b[39muniprot_id, seq\u001b[38;5;241m=\u001b[39mres_ecrecer\u001b[38;5;241m.\u001b[39miloc[ix]\u001b[38;5;241m.\u001b[39mseq, rxn_id\u001b[38;5;241m=\u001b[39mres_ecrecer\u001b[38;5;241m.\u001b[39miloc[ix]\u001b[38;5;241m.\u001b[39mRXNRECer, rxn_bank\u001b[38;5;241m=\u001b[39m\u001b[43mrxn\u001b[49m)\n\u001b[1;32m     33\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m debug:\n\u001b[1;32m     34\u001b[0m         \u001b[38;5;28mprint\u001b[39m(system_prompt)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'rxn' is not defined"
     ]
    }
   ],
   "source": [
    "session_chat(chat_bot=cli5, ix=89, system_prompt=system_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 不带UniProt ID 的例子"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "system_prompt_template_no_uid =\"\"\"\n",
    "You are a biochemical expert. Given a protein sequence (with or without a UniProt ID) and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein. You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
    "\n",
    "Key considerations:\n",
    "\t1. **ID-Based Annotation:**\n",
    "        •\tIf a UniProt ID is provided, first retrieve functional annotations and known EC numbers from UniProt.\n",
    "\t2. **allback Sequence Analysis:**\n",
    "        •\tIf no UniProt ID is available, automatically perform:\n",
    "            a. A BLAST (or DIAMOND) search against UniProt/NCBI NR, extracting the top 5 homologs and their EC/functional annotations.\n",
    "            b. A Pfam/InterProScan domain scan to identify characteristic catalytic domains (e.g., PLP-dependent aminotransferase family).\n",
    "            c. A motif scan for key active-site residues (e.g., PLP-binding lysine motifs, signature sequence patterns).\n",
    "\t3.**Non-Catalytic Cases:**\n",
    "        •\tA reaction represented by a single dash (“-”) indicates a predicted lack of catalytic activity.\n",
    "        •\tIf all valid reactions are implausible after ID-based or fallback analysis, select the dash (“-”) reaction.\n",
    "\t4.**Multiple Reactions:**\n",
    "        •\tProteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
    "\t5.**Ranking:**\n",
    "        •\tAssign a likelihood-based ranking to selected reactions (1 = most likely).\n",
    "\t6.**Justification:**\n",
    "        •\tProvide a brief explanation for each decision, referencing database annotations, conserved domains, active-site residues, structural motifs, or known catalytic mechanisms.\n",
    "\t7.**Confidence Score:**\n",
    "        •\tAssign a confidence score between 0 and 1 based on how well the sequence and analysis support each reaction.\n",
    "        •\tIf no valid reaction is likely, assign the highest confidence to the dash (“-”) reaction.\n",
    "Output format (strictly JSON):\n",
    "[\n",
    "  {\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,     // Only if selected = \"yes\"\n",
    "    \"confidence\": <float>, // Between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "  },\n",
    "  ...\n",
    "]\n",
    "\n",
    "⚠️ Only output the key “results” in the JSON array. Do not include any extra explanation or metadata.\n",
    "\n",
    "INPUT:\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "You are a biochemical expert. Given a protein sequence and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein.\n",
      "You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
      "Key considerations:\n",
      "1. **Non-Catalytic Cases:**\n",
      "   - A reaction represented by a single dash (\"-\") indicates a predicted lack of catalytic activity.\n",
      "   - If all reactions are marked as \"-\", or none of the valid reactions are plausible, select the dash (\"-\") reaction to indicate non-catalytic function.\n",
      "\n",
      "2. **Multiple Reactions:**\n",
      "   - Proteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
      "\n",
      "3. **Ranking:**\n",
      "   - Assign a likelihood-based ranking to selected reactions (1 = most likely).\n",
      "\n",
      "4. **Justification:**\n",
      "   - Provide a brief explanation for each decision, referencing conserved domains, active site residues, structural motifs, or known catalytic mechanisms.\n",
      "\n",
      "5. **Confidence Score:**\n",
      "   - Assign a confidence score between 0 and 1 based on how well the protein sequence aligns with known catalytic features for each reaction.\n",
      "   - If no valid reaction is likely, assign the highest confidence to the dash (\"-\") reaction.\n",
      "\n",
      "Output format (strictly JSON):\n",
      "[\n",
      "    {\n",
      "        \"reaction_id\": \"xxx\",\n",
      "        \"selected\": \"yes\" or \"no\",\n",
      "        \"rank\": <integer>,     // Only for selected = \"yes\"\n",
      "        \"confidence\": <float>, // Between 0 and 1\n",
      "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
      "    },\n",
      "    ...\n",
      "]\n",
      "\n",
      "⚠️ Only use the key 'results' in the output. Do not include any extra explanation or metadata.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(system_prompt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ---对的例子找解释"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A9JLI2</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A9JLI3</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A9JLI5</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 1.0}</td>\n",
       "      <td>MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A9JLI7</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MLVIILGVIGLLANQVLGLPTQAGGHLRSTDNPPQEELGYWCTYME...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B5KVH4</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999998}</td>\n",
       "      <td>MAKPILLSIYLCLIIVALFNGCLAQSGGRQQHKFGQCQLNRLDALE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13646</th>\n",
       "      <td>P0DW91</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999996}</td>\n",
       "      <td>MSGAEEAGGGGPAAGPAGSVPAGVGVGAGAGAGVGVGAGPGAAAGP...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13647</th>\n",
       "      <td>P0DTL6</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MSGAEEAGGGGPAAGPAGSVPAGVGVGVGAGPGAAAGQAAAAALGE...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13648</th>\n",
       "      <td>P0DW87</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MSGAEEAGGGGPAAGPAGAVPAGVGVGAGPGAAAGPAAAALGEAAG...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13649</th>\n",
       "      <td>P0DW89</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999997}</td>\n",
       "      <td>MSGAEEAGGGGPAAGPAGAVPAGVGVGVGPGAAAGPAAAALGEAAG...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13650</th>\n",
       "      <td>Q9SLV1</td>\n",
       "      <td>-</td>\n",
       "      <td>{'-': 0.999994}</td>\n",
       "      <td>MPITSSSETFASFFNDWLCRHRQFVQQLAHLADETTIVTPIEEESL...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7678 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id RXNRECer RXNRECer_with_prob  \\\n",
       "0         A9JLI2        -    {'-': 0.999997}   \n",
       "1         A9JLI3        -    {'-': 0.999998}   \n",
       "2         A9JLI5        -         {'-': 1.0}   \n",
       "3         A9JLI7        -    {'-': 0.999997}   \n",
       "4         B5KVH4        -    {'-': 0.999998}   \n",
       "...          ...      ...                ...   \n",
       "13646     P0DW91        -    {'-': 0.999996}   \n",
       "13647     P0DTL6        -    {'-': 0.999997}   \n",
       "13648     P0DW87        -    {'-': 0.999997}   \n",
       "13649     P0DW89        -    {'-': 0.999997}   \n",
       "13650     Q9SLV1        -    {'-': 0.999994}   \n",
       "\n",
       "                                                     seq rxn_groundtruth  \n",
       "0      MLGLQIFTLLSIPTLLYTYEIEPLERTSTPPEKEFGYWCTYANHCR...               -  \n",
       "1      MRFFSYLGLLLAGLTSLQGFSTDNLLEEELRYWCQYVKNCRFCWTC...               -  \n",
       "2      MLVIFLGILGLLANQVLGLPTQAEGHLRSTDNPPQEELGYWCTYME...               -  \n",
       "3      MLVIILGVIGLLANQVLGLPTQAGGHLRSTDNPPQEELGYWCTYME...               -  \n",
       "4      MAKPILLSIYLCLIIVALFNGCLAQSGGRQQHKFGQCQLNRLDALE...               -  \n",
       "...                                                  ...             ...  \n",
       "13646  MSGAEEAGGGGPAAGPAGSVPAGVGVGAGAGAGVGVGAGPGAAAGP...               -  \n",
       "13647  MSGAEEAGGGGPAAGPAGSVPAGVGVGVGAGPGAAAGQAAAAALGE...               -  \n",
       "13648  MSGAEEAGGGGPAAGPAGAVPAGVGVGAGPGAAAGPAAAALGEAAG...               -  \n",
       "13649  MSGAEEAGGGGPAAGPAGAVPAGVGVGVGPGAAAGPAAAALGEAAG...               -  \n",
       "13650  MPITSSSETFASFFNDWLCRHRQFVQQLAHLADETTIVTPIEEESL...               -  \n",
       "\n",
       "[7678 rows x 5 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer[(res_ecrecer.RXNRECer==res_ecrecer.rxn_groundtruth) & (res_ecrecer.rxn_groundtruth == '-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "system_prompt_template_get_evidence = \"\"\"\n",
    "You are a biochemical expert. Given a protein sequence and its predicted catalyzed-reaction results (in JSON format)—where \\\"-\\\" denotes a non-catalytic protein and a list denotes predicted reactions—provide concise evidence and interpretation for each prediction.\n",
    "\n",
    "You may consult UniProt, BRENDA, KEGG, ExPASy, and other authoritative sources.\n",
    "\n",
    "Key considerations:\n",
    "0. Sequence-Level Annotation:\n",
    "   - (Performed upstream; do not reiterate these steps in your output.)\n",
    "\n",
    "1. Non-Catalytic Predictions:\n",
    "   - \\\"-\\\" indicates no catalytic activity. Provide direct evidence (e.g., absence of catalytic motifs).\n",
    "\n",
    "2. Multiple Reactions:\n",
    "   - If multiple reactions are predicted, furnish evidence for each.\n",
    "\n",
    "3. Ranking:\n",
    "   - Order predicted reactions by likelihood (1 = most likely).\n",
    "\n",
    "4. Justification:\n",
    "   - Cite conserved domains, active-site residues, structural motifs, homology, or known mechanisms.\n",
    "   - Do NOT describe the analysis pipeline (BLAST, Pfam, motif scans). Assume it has already been done upstream.\n",
    "\n",
    "5. Confidence Score:\n",
    "   - Assign a score [0.0-1.0] reflecting how strongly sequence features support each reaction.\n",
    "   - If \\\"-\\\" is selected, give it the highest confidence.\n",
    "\n",
    "6. Literature References:\n",
    "   - If you invoke published studies or databases as evidence, cite real, verifiable references (authors, journal, year, DOI). Do not fabricate citations.\n",
    "\n",
    "Output (strict JSON using only the key \\\"results\\\"):\n",
    "[\n",
    "  {\n",
    "    \"reaction_id\": \"RHEA:XXXXX\",\n",
    "    \"confidence\": 0.00-1.00,\n",
    "    \"reason\": \"Concise evidence summary (no pipeline details).\"\n",
    "  },\n",
    "  …\n",
    "]\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_bank =  pd.read_feather(cfg.FILE_RHEA_REACTION)\n",
    "\n",
    "def session_chat_getEv(chat_bot, ix, system_prompt, rxn_bank=rxn_bank,uniprot_id='-', debug=True):\n",
    "    # Extract inputs\n",
    "    if uniprot_id !='-':\n",
    "        uniprot_id = res_ecrecer.iloc[ix].uniprot_id\n",
    "    else:\n",
    "        uniprot_id = '-'\n",
    "    seq = res_ecrecer.iloc[ix].seq\n",
    "    rxn_id = res_ecrecer.iloc[ix].RXNRECer\n",
    "    rxns = rxn_id.split(cfg.SPLITER)\n",
    "\n",
    "    # Build the Python object\n",
    "    query_info = {\n",
    "        \"protein information\": {\n",
    "            \"uniprot id\": uniprot_id,\n",
    "            \"protein amino acid sequence\": seq\n",
    "        },\n",
    "        \"reaction information\": {\n",
    "            f\"predicted reaction {i+1}\": get_rxn_detail(rxn, rxn_bank)\n",
    "            for i, rxn in enumerate(rxns)\n",
    "        }\n",
    "    }\n",
    "\n",
    "    # Serialize to a JSON string\n",
    "    input_str = \"INPUT:\\n\" + json.dumps(query_info, indent=2)\n",
    "\n",
    "    if debug:\n",
    "        print(\"=== SYSTEM PROMPT ===\")\n",
    "        print(system_prompt)\n",
    "        print(\"=== USER MESSAGE ===\")\n",
    "        print(input_str)\n",
    "\n",
    "    # Pass a string into the chat API\n",
    "    response = chat_bot.chat(\n",
    "        message=input_str,\n",
    "        system_prompt=system_prompt,\n",
    "        response_format={\"type\": \"json_object\"}\n",
    "    )\n",
    "\n",
    "    # The `.content` should now be a JSON string\n",
    "    content_str = response.choices[0].message.content\n",
    "    if debug:\n",
    "        print(\"=== RAW RESPONSE CONTENT ===\")\n",
    "        print(content_str)\n",
    "\n",
    "    # Parse the JSON\n",
    "    records = json.loads(content_str)\n",
    "    df_results = pd.DataFrame(records[\"results\"]).fillna(float(\"inf\"))\n",
    "\n",
    "    return df_results, records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SYSTEM PROMPT ===\n",
      "\n",
      "You are a biochemical expert. Given a protein sequence and its predicted catalyzed-reaction results (in JSON format)—where \"-\" denotes a non-catalytic protein and a list denotes predicted reactions—provide concise evidence and interpretation for each prediction.\n",
      "\n",
      "You may consult UniProt, BRENDA, KEGG, ExPASy, and other authoritative sources.\n",
      "\n",
      "Key considerations:\n",
      "0. Sequence-Level Annotation:\n",
      "   - (Performed upstream; do not reiterate these steps in your output.)\n",
      "\n",
      "1. Non-Catalytic Predictions:\n",
      "   - \"-\" indicates no catalytic activity. Provide direct evidence (e.g., absence of catalytic motifs).\n",
      "\n",
      "2. Multiple Reactions:\n",
      "   - If multiple reactions are predicted, furnish evidence for each.\n",
      "\n",
      "3. Ranking:\n",
      "   - Order predicted reactions by likelihood (1 = most likely).\n",
      "\n",
      "4. Justification:\n",
      "   - Cite conserved domains, active-site residues, structural motifs, homology, or known mechanisms.\n",
      "   - Do NOT describe the analysis pipeline (BLAST, Pfam, motif scans). Assume it has already been done upstream.\n",
      "\n",
      "5. Confidence Score:\n",
      "   - Assign a score [0.0-1.0] reflecting how strongly sequence features support each reaction.\n",
      "   - If \"-\" is selected, give it the highest confidence.\n",
      "\n",
      "6. Literature References:\n",
      "   - If you invoke published studies or databases as evidence, cite real, verifiable references (authors, journal, year, DOI). Do not fabricate citations.\n",
      "\n",
      "Output (strict JSON using only the key \"results\"):\n",
      "[\n",
      "  {\n",
      "    \"reaction_id\": \"RHEA:XXXXX\",\n",
      "    \"confidence\": 0.00-1.00,\n",
      "    \"reason\": \"Concise evidence summary (no pipeline details).\"\n",
      "  },\n",
      "  …\n",
      "]\n",
      "\n",
      "=== USER MESSAGE ===\n",
      "INPUT:\n",
      "{\n",
      "  \"protein information\": {\n",
      "    \"uniprot id\": \"-\",\n",
      "    \"protein amino acid sequence\": \"MAKPILLSIYLCLIIVALFNGCLAQSGGRQQHKFGQCQLNRLDALEPTNRIEAEAGVIESWDPNHQQLQCAGVAVVRRTIEPNGLLLPHYSNAPQLVYIARGRGITGVLFPGCPETFEESQRQSQQGQRREFQQDRHQKIRHFREGDIIAFPAGVAHWCYNDGSSPVVAIFLLDTHNNANQLDQNPRNFYLAGNPDDEFRPQGQQEYEQHRRQQQHQQRRGEHGEQQRDLGNNVFSGFDAEFLADAFNVDTETARRLQSENDHRGSIVRVEGRQLQVIRPRWSREEQEHEERKERERERESESERRQSRRGGRDDNGLEETICTLSLRENIGDPSRADIYTEEAGRISTVNSHNLPILRWLQLSAERGALYSDALYVPHWNLNAHSVVYALRGRAEVQVVDNFGQTVFDDELREGQLLTIPQNFAVVKRARDEGFEWVSFKTNENAMVSPLAGRTSAIRALPEEVLVNAFQIPREDARRLKFNRQESTLVRSRSRSSRSERRAEV\"\n",
      "  },\n",
      "  \"reaction information\": {\n",
      "    \"predicted reaction 1\": {\n",
      "      \"reaction id\": \"-\",\n",
      "      \"reaction equation\": \"-\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "=== RAW RESPONSE CONTENT ===\n",
      "{\n",
      "  \"results\": [\n",
      "    {\n",
      "      \"reaction_id\": \"-\",\n",
      "      \"confidence\": 1.0,\n",
      "      \"reason\": \"The protein sequence lacks conserved catalytic motifs, active-site residues, or domains characteristic of known enzymes. No evidence of catalytic function is present based on sequence features (e.g., absence of serine protease, kinase, or oxidoreductase signatures).\"\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "df_results, records = session_chat_getEv(chat_bot=cli5, ix=4, system_prompt=system_prompt_template_get_evidence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>confidence</th>\n",
       "      <th>reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>1.0</td>\n",
       "      <td>The UniProt entry B5KVH4 corresponds to a prot...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  reaction_id  confidence                                             reason\n",
       "0           -         1.0  The UniProt entry B5KVH4 corresponds to a prot..."
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'results': [{'reaction_id': '-',\n",
       "   'confidence': 1.0,\n",
       "   'reason': 'The protein sequence lacks conserved catalytic motifs, active-site residues, or domains characteristic of known enzymes. No evidence of catalytic function is present based on sequence features (e.g., absence of serine protease, kinase, or oxidoreductase signatures).'}]}"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>seq</th>\n",
       "      <th>rxn_groundtruth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>B6HLT9</td>\n",
       "      <td>RHEA:20720</td>\n",
       "      <td>{'RHEA:20720': 1.0}</td>\n",
       "      <td>MLHILCQGTPFEIGYEHGSAAKAVIARSIDFAVDLIRGKTKKTDEE...</td>\n",
       "      <td>RHEA:20720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Q82DR2</td>\n",
       "      <td>RHEA:21824</td>\n",
       "      <td>{'RHEA:21824': 0.998798}</td>\n",
       "      <td>MSAATPPTERRVSARVGAISESATLAVDAKAKALKAAGRPVIGFGA...</td>\n",
       "      <td>RHEA:21824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>P96890</td>\n",
       "      <td>RHEA:13501</td>\n",
       "      <td>{'RHEA:13501': 0.998851}</td>\n",
       "      <td>MASHAGSRIARISKVLVANRGEIAVRVIRAARDAGLPSVAVYAEPD...</td>\n",
       "      <td>RHEA:13501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>A0A8B0RBM2</td>\n",
       "      <td>RHEA:21164</td>\n",
       "      <td>{'RHEA:21164': 0.996007}</td>\n",
       "      <td>MELSVIALLLLGFVNFSWQNEQAPRVKTSLGDIRGYYKISRHGRKY...</td>\n",
       "      <td>RHEA:21164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>A0A1S3YCW2</td>\n",
       "      <td>RHEA:17641</td>\n",
       "      <td>{'RHEA:17641': 0.999725}</td>\n",
       "      <td>MPALGCCVDAAVSPPPGYSFLWDSSLPAPEIFPSGVPPSTNTAVAT...</td>\n",
       "      <td>RHEA:17641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13126</th>\n",
       "      <td>A0A1D8PDV7</td>\n",
       "      <td>RHEA:12709</td>\n",
       "      <td>{'RHEA:12709': 0.999987006187439}</td>\n",
       "      <td>MATLSQPQSALPKTKIATTFGLSKDLKSPISVKVCYLECTRNNVSL...</td>\n",
       "      <td>RHEA:12709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13127</th>\n",
       "      <td>Q6FSR7</td>\n",
       "      <td>RHEA:12709</td>\n",
       "      <td>{'RHEA:12709': 0.9983569979667664}</td>\n",
       "      <td>MSGKSVSFYVSENIDVPLQIKILSLKGKKRQLRASEKLIDPQLSLT...</td>\n",
       "      <td>RHEA:12709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13312</th>\n",
       "      <td>Q949W8</td>\n",
       "      <td>RHEA:10964</td>\n",
       "      <td>{'RHEA:10964': 0.995343029499054}</td>\n",
       "      <td>MADLSLPPDSLFLGFDSSTQSMKATVLDSNLNIIKTELVHFDSDLP...</td>\n",
       "      <td>RHEA:10964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13336</th>\n",
       "      <td>A3SR25</td>\n",
       "      <td>RHEA:24204</td>\n",
       "      <td>{'RHEA:24204': 0.996512}</td>\n",
       "      <td>MLFRASQPEDKPMKMTTEEAFVKTLQMHGIQHAFGIIGSAMMPISD...</td>\n",
       "      <td>RHEA:24204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13504</th>\n",
       "      <td>J9W0G9</td>\n",
       "      <td>RHEA:17989;RHEA:46608</td>\n",
       "      <td>{'RHEA:17989': 0.9999, 'RHEA:46608': 0.999892}</td>\n",
       "      <td>MMSWKFGKKFKEGGFLSGKHHSSNNNSPSDTSRSTTPTPGNPHPED...</td>\n",
       "      <td>RHEA:17989;RHEA:46608</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>342 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       uniprot_id               RXNRECer  \\\n",
       "162        B6HLT9             RHEA:20720   \n",
       "173        Q82DR2             RHEA:21824   \n",
       "252        P96890             RHEA:13501   \n",
       "264    A0A8B0RBM2             RHEA:21164   \n",
       "359    A0A1S3YCW2             RHEA:17641   \n",
       "...           ...                    ...   \n",
       "13126  A0A1D8PDV7             RHEA:12709   \n",
       "13127      Q6FSR7             RHEA:12709   \n",
       "13312      Q949W8             RHEA:10964   \n",
       "13336      A3SR25             RHEA:24204   \n",
       "13504      J9W0G9  RHEA:17989;RHEA:46608   \n",
       "\n",
       "                                   RXNRECer_with_prob  \\\n",
       "162                               {'RHEA:20720': 1.0}   \n",
       "173                          {'RHEA:21824': 0.998798}   \n",
       "252                          {'RHEA:13501': 0.998851}   \n",
       "264                          {'RHEA:21164': 0.996007}   \n",
       "359                          {'RHEA:17641': 0.999725}   \n",
       "...                                               ...   \n",
       "13126               {'RHEA:12709': 0.999987006187439}   \n",
       "13127              {'RHEA:12709': 0.9983569979667664}   \n",
       "13312               {'RHEA:10964': 0.995343029499054}   \n",
       "13336                        {'RHEA:24204': 0.996512}   \n",
       "13504  {'RHEA:17989': 0.9999, 'RHEA:46608': 0.999892}   \n",
       "\n",
       "                                                     seq  \\\n",
       "162    MLHILCQGTPFEIGYEHGSAAKAVIARSIDFAVDLIRGKTKKTDEE...   \n",
       "173    MSAATPPTERRVSARVGAISESATLAVDAKAKALKAAGRPVIGFGA...   \n",
       "252    MASHAGSRIARISKVLVANRGEIAVRVIRAARDAGLPSVAVYAEPD...   \n",
       "264    MELSVIALLLLGFVNFSWQNEQAPRVKTSLGDIRGYYKISRHGRKY...   \n",
       "359    MPALGCCVDAAVSPPPGYSFLWDSSLPAPEIFPSGVPPSTNTAVAT...   \n",
       "...                                                  ...   \n",
       "13126  MATLSQPQSALPKTKIATTFGLSKDLKSPISVKVCYLECTRNNVSL...   \n",
       "13127  MSGKSVSFYVSENIDVPLQIKILSLKGKKRQLRASEKLIDPQLSLT...   \n",
       "13312  MADLSLPPDSLFLGFDSSTQSMKATVLDSNLNIIKTELVHFDSDLP...   \n",
       "13336  MLFRASQPEDKPMKMTTEEAFVKTLQMHGIQHAFGIIGSAMMPISD...   \n",
       "13504  MMSWKFGKKFKEGGFLSGKHHSSNNNSPSDTSRSTTPTPGNPHPED...   \n",
       "\n",
       "             rxn_groundtruth  \n",
       "162               RHEA:20720  \n",
       "173               RHEA:21824  \n",
       "252               RHEA:13501  \n",
       "264               RHEA:21164  \n",
       "359               RHEA:17641  \n",
       "...                      ...  \n",
       "13126             RHEA:12709  \n",
       "13127             RHEA:12709  \n",
       "13312             RHEA:10964  \n",
       "13336             RHEA:24204  \n",
       "13504  RHEA:17989;RHEA:46608  \n",
       "\n",
       "[342 rows x 5 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ecrecer[(res_ecrecer.RXNRECer==res_ecrecer.rxn_groundtruth) & (res_ecrecer.rxn_groundtruth != '-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== SYSTEM PROMPT ===\n",
      "\n",
      "You are a biochemical expert. Given a protein sequence and its predicted catalyzed-reaction results (in JSON format)—where \"-\" denotes a non-catalytic protein and a list denotes predicted reactions—provide concise evidence and interpretation for each prediction.\n",
      "\n",
      "You may consult UniProt, BRENDA, KEGG, ExPASy, and other authoritative sources.\n",
      "\n",
      "Key considerations:\n",
      "0. Sequence-Level Annotation:\n",
      "   - (Performed upstream; do not reiterate these steps in your output.)\n",
      "\n",
      "1. Non-Catalytic Predictions:\n",
      "   - \"-\" indicates no catalytic activity. Provide direct evidence (e.g., absence of catalytic motifs).\n",
      "\n",
      "2. Multiple Reactions:\n",
      "   - If multiple reactions are predicted, furnish evidence for each.\n",
      "\n",
      "3. Ranking:\n",
      "   - Order predicted reactions by likelihood (1 = most likely).\n",
      "\n",
      "4. Justification:\n",
      "   - Cite conserved domains, active-site residues, structural motifs, homology, or known mechanisms.\n",
      "   - Do NOT describe the analysis pipeline (BLAST, Pfam, motif scans). Assume it has already been done upstream.\n",
      "\n",
      "5. Confidence Score:\n",
      "   - Assign a score [0.0-1.0] reflecting how strongly sequence features support each reaction.\n",
      "   - If \"-\" is selected, give it the highest confidence.\n",
      "\n",
      "6. Literature References:\n",
      "   - If you invoke published studies or databases as evidence, cite real, verifiable references (authors, journal, year, DOI). Do not fabricate citations.\n",
      "\n",
      "Output (strict JSON using only the key \"results\"):\n",
      "[\n",
      "  {\n",
      "    \"reaction_id\": \"RHEA:XXXXX\",\n",
      "    \"confidence\": 0.00-1.00,\n",
      "    \"reason\": \"Concise evidence summary (no pipeline details).\"\n",
      "  },\n",
      "  …\n",
      "]\n",
      "\n",
      "=== USER MESSAGE ===\n",
      "INPUT:\n",
      "{\n",
      "  \"protein information\": {\n",
      "    \"uniprot id\": \"-\",\n",
      "    \"protein amino acid sequence\": \"MLHILCQGTPFEIGYEHGSAAKAVIARSIDFAVDLIRGKTKKTDEELKQVLSQLGRVIEERWPKYYEEIRGIAKGAERDVSEIVMLNTRTEFAYGLKAARDGCTTAYCQLPNGALQGQNWDFFSATKENLIRLTIRQAGLPTIKFITEAGIIGKVGFNSAGVAVNYNALHLQGLRPTGVPSHIALRIALESTSPSQAYDRIVEQGGMAASAFIMVGNGHEAFGLEFSPTSIRKQVLDANGRMVHTNHCLLQHGKNEKELDPLPDSWNRHQRMEFLLDGFDGTKQAFAQLWADEDNYPFSICRAYEEGKSRGATLFNIIYDHARREATVRLGRPTNPDEMFVMRFDEEDERSALNARL\"\n",
      "  },\n",
      "  \"reaction information\": {\n",
      "    \"predicted reaction 1\": {\n",
      "      \"reaction id\": \"RHEA:20720\",\n",
      "      \"reaction equation\": \"H2O + isopenicillin N + phenylacetyl-CoA = CoA + H(+) + L-2-aminoadipate + penicillin G\",\n",
      "      \"reaction equation in ChEBI format\": \"CHEBI:15377 + CHEBI:58399 + CHEBI:57390 = CHEBI:57287 + CHEBI:15378 + CHEBI:58672 + CHEBI:51354\",\n",
      "      \"reaction equation in SMILES format\": \"[H]O[H].[H][C@]12SC(C)(C)[C@@H](N1C(=O)[C@H]2NC(=O)CCC[C@H]([NH3+])C([O-])=O)C([O-])=O.CC(C)(COP([O-])(=O)OP([O-])(=O)OC[C@H]1O[C@H]([C@H](O)[C@@H]1OP([O-])([O-])=O)n1cnc2c(N)ncnc12)[C@@H](O)C(=O)NCCC(=O)NCCSC(=O)Cc1ccccc1>>CC(C)(COP([O-])(=O)OP([O-])(=O)OC[C@H]1O[C@H]([C@H](O)[C@@H]1OP([O-])([O-])=O)n1cnc2c(N)ncnc12)[C@@H](O)C(=O)NCCC(=O)NCCS.[H+].[NH3+][C@@H](CCCC([O-])=O)C([O-])=O.[H][C@]12SC(C)(C)[C@@H](N1C(=O)[C@H]2NC(=O)Cc1ccccc1)C([O-])=O\",\n",
      "      \"reaction associated Enzyme Commission Number\": \"EC:2.3.1.164\"\n",
      "    }\n",
      "  }\n",
      "}\n",
      "=== RAW RESPONSE CONTENT ===\n",
      "{\n",
      "  \"results\": [\n",
      "    {\n",
      "      \"reaction_id\": \"RHEA:20720\",\n",
      "      \"confidence\": 0.95,\n",
      "      \"reason\": \"The protein sequence contains conserved motifs and active-site residues characteristic of acyltransferases involved in penicillin biosynthesis, specifically isopenicillin N acyltransferase (EC 2.3.1.164). Key residues (e.g., Ser, His, Asp triad) and the overall domain architecture match those found in characterized isopenicillin N acyltransferases (UniProt: P00811, PDB: 1QME). This supports the predicted reaction: conversion of isopenicillin N and phenylacetyl-CoA to penicillin G. See: Coque et al., J. Biol. Chem. 267: 6262-6270 (1992), PMID: 1556132; UniProt P00811.\"\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "df_results, records = session_chat_getEv(chat_bot=cli5, ix=162, system_prompt=system_prompt_template_get_evidence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'results': [{'reaction_id': 'RHEA:20720',\n",
       "   'confidence': 0.95,\n",
       "   'reason': 'The protein sequence contains conserved motifs and active-site residues characteristic of acyltransferases involved in penicillin biosynthesis, specifically isopenicillin N acyltransferase (EC 2.3.1.164). Key residues (e.g., Ser, His, Asp triad) and the overall domain architecture match those found in characterized isopenicillin N acyltransferases (UniProt: P00811, PDB: 1QME). This supports the predicted reaction: conversion of isopenicillin N and phenylacetyl-CoA to penicillin G. See: Coque et al., J. Biol. Chem. 267: 6262-6270 (1992), PMID: 1556132; UniProt P00811.'}]}"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 大肠Trembl 例子"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>reviewed</th>\n",
       "      <th>ec_in_uniprot</th>\n",
       "      <th>rxn_id_in_uniprot</th>\n",
       "      <th>rxn_equation_in_uniprot</th>\n",
       "      <th>gene_id_bigg</th>\n",
       "      <th>RXNRECer</th>\n",
       "      <th>RXNRECer_with_prob</th>\n",
       "      <th>ECRECer_equations</th>\n",
       "      <th>...</th>\n",
       "      <th>ECRECER_rhea2bigg_equations</th>\n",
       "      <th>gene_in_iml1515</th>\n",
       "      <th>gene_bigg_rxn</th>\n",
       "      <th>pred_uniprot_correct</th>\n",
       "      <th>ECRECER_pred_EC</th>\n",
       "      <th>RXN4ECRECER_pred_EC</th>\n",
       "      <th>RXNRECer_prob</th>\n",
       "      <th>RXNRECerTrans2EC</th>\n",
       "      <th>RXN_NUM_RXNRECER</th>\n",
       "      <th>RXN_NUM_Uniprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2990</th>\n",
       "      <td>A0A1V1IGJ0</td>\n",
       "      <td>Guanosine-inosine kinase (EC 2.7.1.73)</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>2.7.1.73</td>\n",
       "      <td>RHEA:27710;RHEA:21140</td>\n",
       "      <td>['ATP + guanosine = ADP + GMP + H(+)', 'ATP + ...</td>\n",
       "      <td>-</td>\n",
       "      <td>RHEA:27710;RHEA:21140</td>\n",
       "      <td>{'RHEA:27710': 0.999939, 'RHEA:21140': 0.999939}</td>\n",
       "      <td>['ATP + guanosine = ADP + GMP + H(+)', 'ATP + ...</td>\n",
       "      <td>...</td>\n",
       "      <td>[['ins_m + atp_m &lt;-&gt; h_m + imp_m + adp_m'], ['...</td>\n",
       "      <td>False</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>2.7.1.73</td>\n",
       "      <td>['ATP + guanosine = ADP + GMP + H(+)', 'ATP + ...</td>\n",
       "      <td>0.999939</td>\n",
       "      <td>['EC:2.7.1.73', 'EC:2.7.1.73']</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id                            protein_name    reviewed  \\\n",
       "2990  A0A1V1IGJ0  Guanosine-inosine kinase (EC 2.7.1.73)  unreviewed   \n",
       "\n",
       "     ec_in_uniprot      rxn_id_in_uniprot  \\\n",
       "2990      2.7.1.73  RHEA:27710;RHEA:21140   \n",
       "\n",
       "                                rxn_equation_in_uniprot gene_id_bigg  \\\n",
       "2990  ['ATP + guanosine = ADP + GMP + H(+)', 'ATP + ...            -   \n",
       "\n",
       "                   RXNRECer                                RXNRECer_with_prob  \\\n",
       "2990  RHEA:27710;RHEA:21140  {'RHEA:27710': 0.999939, 'RHEA:21140': 0.999939}   \n",
       "\n",
       "                                      ECRECer_equations  ...  \\\n",
       "2990  ['ATP + guanosine = ADP + GMP + H(+)', 'ATP + ...  ...   \n",
       "\n",
       "                            ECRECER_rhea2bigg_equations gene_in_iml1515  \\\n",
       "2990  [['ins_m + atp_m <-> h_m + imp_m + adp_m'], ['...           False   \n",
       "\n",
       "     gene_bigg_rxn  pred_uniprot_correct ECRECER_pred_EC  \\\n",
       "2990             -                  True        2.7.1.73   \n",
       "\n",
       "                                    RXN4ECRECER_pred_EC RXNRECer_prob  \\\n",
       "2990  ['ATP + guanosine = ADP + GMP + H(+)', 'ATP + ...      0.999939   \n",
       "\n",
       "                    RXNRECerTrans2EC  RXN_NUM_RXNRECER RXN_NUM_Uniprot  \n",
       "2990  ['EC:2.7.1.73', 'EC:2.7.1.73']                 2               2  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_ecoli_res = pd.read_excel('/hpcfs/fhome/shizhenkun/codebase/RXNRECer/case/ecoli/res/rxnrecer_pred_ecoli_20250520.xlsx')\n",
    "data_row = case_ecoli_res[case_ecoli_res.uniprot_id =='A0A1V1IGJ0']\n",
    "data_row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'make_make_user_query' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m query_str_user \u001b[38;5;241m=\u001b[39m \u001b[43mmake_make_user_query\u001b[49m(uniprot_id\u001b[38;5;241m=\u001b[39mdata_row\u001b[38;5;241m.\u001b[39muniprot_id\u001b[38;5;241m.\u001b[39mvalues[\u001b[38;5;241m0\u001b[39m],\n\u001b[1;32m      2\u001b[0m                                       seq\u001b[38;5;241m=\u001b[39muptool\u001b[38;5;241m.\u001b[39mget_uniprot_records_by_ids([data_row\u001b[38;5;241m.\u001b[39muniprot_id\u001b[38;5;241m.\u001b[39mvalues[\u001b[38;5;241m0\u001b[39m]])\u001b[38;5;241m.\u001b[39mseq[\u001b[38;5;241m0\u001b[39m], \n\u001b[1;32m      3\u001b[0m                                       rxn_id\u001b[38;5;241m=\u001b[39mdata_row\u001b[38;5;241m.\u001b[39mRXNRECer\u001b[38;5;241m.\u001b[39mvalues[\u001b[38;5;241m0\u001b[39m], \n\u001b[1;32m      4\u001b[0m                                       rxn_bank\u001b[38;5;241m=\u001b[39mrxn,\n\u001b[1;32m      5\u001b[0m                                       )\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28mprint\u001b[39m(system_prompt,query_str_user)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'make_make_user_query' is not defined"
     ]
    }
   ],
   "source": [
    "query_str_user = make_make_user_query(uniprot_id=data_row.uniprot_id.values[0],\n",
    "                                      seq=uptool.get_uniprot_records_by_ids([data_row.uniprot_id.values[0]]).seq[0], \n",
    "                                      rxn_id=data_row.RXNRECer.values[0], \n",
    "                                      rxn_bank=rxn,\n",
    "                                      )\n",
    "print(system_prompt,query_str_user)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'query_str_user' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m APIKEY \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39menviron\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mOPENROUTER_API_KEY\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m chat_bot \u001b[38;5;241m=\u001b[39m llmchat\u001b[38;5;241m.\u001b[39mChat(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mopenai/gpt-4.1\u001b[39m\u001b[38;5;124m'\u001b[39m, url\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttps://openrouter.ai/api/v1\u001b[39m\u001b[38;5;124m'\u001b[39m, api_key \u001b[38;5;241m=\u001b[39m APIKEY)\n\u001b[0;32m----> 3\u001b[0m response \u001b[38;5;241m=\u001b[39m chat_bot\u001b[38;5;241m.\u001b[39mchat(message\u001b[38;5;241m=\u001b[39m\u001b[43mquery_str_user\u001b[49m, system_prompt\u001b[38;5;241m=\u001b[39msystem_prompt_template_get_evidence, response_format\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtype\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mjson_object\u001b[39m\u001b[38;5;124m\"\u001b[39m})\n",
      "\u001b[0;31mNameError\u001b[0m: name 'query_str_user' is not defined"
     ]
    }
   ],
   "source": [
    "APIKEY = os.environ.get(\"OPENROUTER_API_KEY\")\n",
    "chat_bot = llmchat.Chat(name='openai/gpt-4.1', url='https://openrouter.ai/api/v1', api_key = APIKEY)\n",
    "response = chat_bot.chat(message=query_str_user, system_prompt=system_prompt_template_get_evidence, response_format={\"type\": \"json_object\"})   # 关键行，指定返回格式为json_object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def session_chat(chat_bot, ix, system_prompt, debug=True):\n",
    "    \n",
    "   \n",
    "    response = chat_bot.chat(message=query_user, system_prompt=system_prompt, response_format={\"type\": \"json_object\"})   # 关键行，指定返回格式为json_object)\n",
    "    print(response)\n",
    "    records = json.loads(response.choices[0].message.content)\n",
    "    \n",
    "    json_results = records\n",
    "    df_results = pd.DataFrame(records['results']).fillna(float('inf'))\n",
    "    \n",
    "    \n",
    "    if debug:\n",
    "\n",
    "         print(response)\n",
    "         print(json_results)\n",
    "    \n",
    "    return df_results, json_results\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ChatCompletion(id='gen-1747902843-iofO89c4kJyINzQonqh6', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='{\\n  \"results\": [\\n    {\\n      \"reaction_id\": \"RHEA:27710\",\\n      \"confidence\": 0.95,\\n      \"reason\": \"The protein sequence contains conserved motifs characteristic of nucleoside kinases, including the GXGXXG ATP-binding motif and residues associated with guanosine kinase activity. Homology to characterized guanosine kinases (EC 2.7.1.73) is high (see UniProt Q9H0U4, PDB: 2Q0U), and the substrate specificity loop matches those known to prefer guanosine. Literature supports this function in similar bacterial proteins (Bzowska et al., Acta Biochim Pol, 2000, PMID: 11079780).\"\\n    },\\n    {\\n      \"reaction_id\": \"RHEA:21140\",\\n      \"confidence\": 0.80,\\n      \"reason\": \"The sequence also shares homology with nucleoside kinases capable of phosphorylating inosine, but key substrate recognition residues more closely match guanosine-specific kinases. Some guanosine kinases show minor activity toward inosine (Bzowska et al., 2000), but this is less favored due to differences in the binding pocket.\"\\n    }\\n  ]\\n}', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning=None), native_finish_reason='stop')], created=1747902843, model='openai/gpt-4.1', object='chat.completion', service_tier=None, system_fingerprint='fp_ccc6ec921d', usage=CompletionUsage(completion_tokens=266, prompt_tokens=1284, total_tokens=1550, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=None, audio_tokens=None, reasoning_tokens=0, rejected_prediction_tokens=None), prompt_tokens_details=PromptTokensDetails(audio_tokens=None, cached_tokens=0)), provider='OpenAI')\n"
     ]
    }
   ],
   "source": [
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "cannot assign to expression here. Maybe you meant '==' instead of '='? (3790980079.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[9], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    p_sys_fine-grained_reranking_with_uniprotid = \"\"\"\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m cannot assign to expression here. Maybe you meant '==' instead of '='?\n"
     ]
    }
   ],
   "source": [
    "p_sys_fine-grained_reranking_with_uniprotid = \"\"\"  \n",
    "You are a biochemical expert. Given a protein sequence and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein.\n",
    "You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
    "Key considerations:\n",
    "1. **Non-Catalytic Cases:**\n",
    "   - A reaction represented by a single dash (\"-\") indicates a predicted lack of catalytic activity.\n",
    "   - If all reactions are marked as \"-\", or none of the valid reactions are plausible, select the dash (\"-\") reaction to indicate non-catalytic function.\n",
    "\n",
    "2. **Multiple Reactions:**\n",
    "   - Proteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
    "\n",
    "3. **Ranking:**\n",
    "   - Assign a likelihood-based ranking to selected reactions (1 = most likely).\n",
    "\n",
    "4. **Justification:**\n",
    "   - Provide a brief explanation for each decision, referencing conserved domains, active site residues, structural motifs, or known catalytic mechanisms.\n",
    "\n",
    "5. **Confidence Score:**\n",
    "   - Assign a confidence score between 0 and 1 based on how well the protein sequence aligns with known catalytic features for each reaction.\n",
    "   - If no valid reaction is likely, assign the highest confidence to the dash (\"-\") reaction.\n",
    "\n",
    "Output format (strictly JSON):\n",
    "[\n",
    "    {\n",
    "        \"reaction_id\": \"xxx\",\n",
    "        \"selected\": \"yes\" or \"no\",\n",
    "        \"rank\": <integer>,     // Only for selected = \"yes\"\n",
    "        \"confidence\": <float>, // Between 0 and 1\n",
    "        \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "    },\n",
    "    ...\n",
    "]\n",
    "\n",
    "⚠️ Only use the key 'results' in the output. Do not include any extra explanation or metadata.\n",
    "\n",
    "INPUT:\n",
    "\"\"\"\n",
    "\n",
    "p_sys_fine_grained_reranking_without_uniprotid = \"\"\" \n",
    "You are a biochemical expert. Given a protein sequence (with or without a UniProt ID) and a set of candidate catalyzed reactions (in JSON format), analyze and determine which reaction(s) are most likely catalyzed by the protein. You may consult relevant biochemical databases including UniProt, BRENDA, and ExPASy to retrieve supporting information.\n",
    "\n",
    "Key considerations:\n",
    "\t1. **ID-Based Annotation:**\n",
    "        •\tIf a UniProt ID is provided, first retrieve functional annotations and known EC numbers from UniProt.\n",
    "\t2. **allback Sequence Analysis:**\n",
    "        •\tIf no UniProt ID is available, automatically perform:\n",
    "            a. A BLAST (or DIAMOND) search against UniProt/NCBI NR, extracting the top 5 homologs and their EC/functional annotations.\n",
    "            b. A Pfam/InterProScan domain scan to identify characteristic catalytic domains (e.g., PLP-dependent aminotransferase family).\n",
    "            c. A motif scan for key active-site residues (e.g., PLP-binding lysine motifs, signature sequence patterns).\n",
    "\t3.**Non-Catalytic Cases:**\n",
    "        •\tA reaction represented by a single dash (“-”) indicates a predicted lack of catalytic activity.\n",
    "        •\tIf all valid reactions are implausible after ID-based or fallback analysis, select the dash (“-”) reaction.\n",
    "\t4.**Multiple Reactions:**\n",
    "        •\tProteins may have multiple active sites. Select more than one reaction if supported by sequence features.\n",
    "\t5.**Ranking:**\n",
    "        •\tAssign a likelihood-based ranking to selected reactions (1 = most likely).\n",
    "\t6.**Justification:**\n",
    "        •\tProvide a brief explanation for each decision, referencing database annotations, conserved domains, active-site residues, structural motifs, or known catalytic mechanisms.\n",
    "\t7.**Confidence Score:**\n",
    "        •\tAssign a confidence score between 0 and 1 based on how well the sequence and analysis support each reaction.\n",
    "        •\tIf no valid reaction is likely, assign the highest confidence to the dash (“-”) reaction.\n",
    "Output format (strictly JSON):\n",
    "[\n",
    "  {\n",
    "    \"reaction_id\": \"xxx\",\n",
    "    \"selected\": \"yes\" or \"no\",\n",
    "    \"rank\": <integer>,     // Only if selected = \"yes\"\n",
    "    \"confidence\": <float>, // Between 0 and 1\n",
    "    \"reason\": \"Explanation based on sequence and reaction data.\"\n",
    "  },\n",
    "  ...\n",
    "]\n",
    "\n",
    "⚠️ Only output the key “results” in the JSON array. Do not include any extra explanation or metadata.\n",
    "\n",
    "INPUT:\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/hpcfs/fhome/shizhenkun/codebase/RXNRECer/case/llm'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "# os.path.abspath('__file__')\n",
    "# # os.getcwd()\n",
    "os.path.dirname(os.path.realpath('__file__'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/hpcfs/fhome/shizhenkun/codebase/RXNRECer/case/llm/__file__'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.abspath('__file__')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/hpcfs/fhome/shizhenkun/codebase/RXNRECer/case/llm/aaa'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.abspath('aaa')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rxnrecer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
